{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Viral Genome Data from NCBI\n",
    "**[Work in progress]**\n",
    "\n",
    "This notebook downloads and standardizes viral genome data from NCBI for ingestion into the Knowledge Graph.\n",
    "\n",
    "Data source: [NCBI](https://www.ncbi.nlm.nih.gov)\n",
    "\n",
    "Author: Peter Rose (pwrose@ucsd.edu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import dateutil\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"covid19@mybinder.org\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_rows = None  # display all rows\n",
    "pd.options.display.max_columns = None  # display all columsns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/peter/Library/Application Support/Neo4j Desktop/Application/neo4jDatabases/database-4af96121-2328-4e2f-ba60-6d8b728a26d5/installation-4.0.3\n"
     ]
    }
   ],
   "source": [
    "NEO4J_HOME = Path(os.getenv('NEO4J_HOME'))\n",
    "print(NEO4J_HOME)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NCBI Data Source\n",
    "[NCBI Severe acute respiratory syndrome coronavirus 2 data hub](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&VirusLineage_ss=Severe%20acute%20respiratory%20syndrome%20coronavirus%202,%20taxid:2697049)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# constants used to generate fasta url: \n",
    "# \"https://www.ncbi.nlm.nih.gov/nuccore/{genbank_accession}/?report=fasta\"\n",
    "\n",
    "ncbi_nuccore_url = \"https://www.ncbi.nlm.nih.gov/nuccore/\"\n",
    "fasta_format = \"/?report=fasta\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download viral genome information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO loop over a list of taxonomy ids. \n",
    "\n",
    "# for now SARS-CoV-2 is hardcoded.\n",
    "tax_id = \"2697049\" # SARS-CoV-2, NC_045512\n",
    "#tax_id = \"694009\" # SARS, NC_004718\n",
    "#tax_id = \"1335626\" # MERS-CoV, NC_019843\n",
    "#tax_id = \"333387\" # Bat SARS coronavirus HKU3-1\n",
    "#tax_id = \"147711\" # Rhinovirus A\n",
    "#tax_id = \"147712\" # Rhinovirus B\n",
    "#tax_id = \"186538\" # Zaire ebolavirus, NC_002549\n",
    "\n",
    "#term = f\"txid{tax_id}[Organism]+refseq[filter]\" # format for reference sequences only"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download nucleotide accession numbers for the given taxonomy id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nucleotides: 1731\n"
     ]
    }
   ],
   "source": [
    "term = f\"txid{tax_id}[Organism]\"\n",
    "handle = Entrez.esearch(db=\"nucleotide\", retmax=5000, term=term, idtype=\"acc\")\n",
    "nuc_accessions = Entrez.read(handle)\n",
    "handle.close()\n",
    "print(\"Nucleotides:\", nuc_accessions['Count'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO refactor function into a separate .py file\n",
    "def extract_genbank_features(gb_record, feature_type, qualifier_list) :\n",
    "    answers = list()\n",
    "    for (index, feature) in enumerate(gb_record.features) :\n",
    "        answer = dict()\n",
    "        if feature.type==feature_type :\n",
    "            for qualifier in qualifier_list:\n",
    "                if qualifier in feature.qualifiers :\n",
    "                    #print(feature, feature.qualifiers)\n",
    "                    # keep NCBI \"1\"-based indexing\n",
    "                    answer['start'] = min(feature.location) + 1\n",
    "                    answer['end'] = max(feature.location) + 1\n",
    "                    #There should only be one locus_tag per feature, but there\n",
    "                    #are usually several db_xref entries\n",
    "                    for value in feature.qualifiers[qualifier] :\n",
    "                        if value in answer :\n",
    "                                print(\"WARNING - Duplicate key %s for %s features %i and %i\" \\\n",
    "                                % (value, feature_type, answer[value], index))\n",
    "                        else :\n",
    "                            answer[qualifier] = value\n",
    "                else:\n",
    "                    answer[qualifier] = ''\n",
    "            answers.append(answer)\n",
    "    return answers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_nucleotide_record(accession):\n",
    "    with Entrez.efetch(db=\"nucleotide\", rettype=\"gb\", retmode=\"text\", id=accession) as handle:\n",
    "        record = SeqIO.read(handle, \"gb\")\n",
    "    # TODO cache genbank files in a temporary directory, so they don't need to be downloaded from scratch\n",
    "    # every time this notebook is run\n",
    "    source = extract_genbank_features(record, 'source', ['db_xref','host','isolate','isolation_source','mol_type','collection_date','country'])   \n",
    "    source_df = pd.DataFrame(source)\n",
    "    source_df['genbank_accession'] = accession\n",
    "    source_df['complete'] = record.description.endswith('complete genome')\n",
    "    \n",
    "    return source_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concatenate all Genbank dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MT396266.1 MT394531.1 MT394530.1 MT394529.1 MT394528.1 MT394864.1 MT396241.1 MT396248.1 MT396247.1 MT396246.1 MT396245.1 MT396244.1 MT396243.1 MT396242.1 MT385497.1 MT385496.1 MT385495.1 MT385494.1 MT385493.1 MT385492.1 MT385491.1 MT385490.1 MT385489.1 MT385488.1 MT385487.1 MT385486.1 MT385485.1 MT385484.1 MT385483.1 MT385482.1 MT385481.1 MT385480.1 MT385479.1 MT385478.1 MT385477.1 MT385476.1 MT385475.1 MT385474.1 MT385473.1 MT385472.1 MT385471.1 MT385470.1 MT385469.1 MT385468.1 MT385467.1 MT385466.1 MT385465.1 MT385464.1 MT385463.1 MT385462.1 MT385461.1 MT385460.1 MT385459.1 MT385458.1 MT385457.1 MT385456.1 MT385455.1 MT385454.1 MT385453.1 MT385452.1 MT385451.1 MT385450.1 MT385449.1 MT385448.1 MT385447.1 MT385446.1 MT385445.1 MT385444.1 MT385443.1 MT385442.1 MT385441.1 MT385440.1 MT385439.1 MT385438.1 MT385437.1 MT385436.1 MT385435.1 MT385434.1 MT385433.1 MT385432.1 MT385431.1 MT385430.1 MT385429.1 MT385428.1 MT385427.1 MT385426.1 MT385425.1 MT385424.1 MT385423.1 MT385422.1 MT385421.1 MT385420.1 MT385419.1 MT385418.1 MT385417.1 MT385416.1 MT385415.1 MT385414.1 MT380734.1 MT380733.1 MT380732.1 MT380731.1 MT380730.1 MT380729.1 MT380728.1 MT380727.1 MT380726.1 MT263458.1 MT263457.1 MT263456.1 MT263455.1 MT263454.1 MT263453.1 MT263452.1 MT263451.1 MT263450.1 MT263449.1 MT263448.1 MT263447.1 MT263446.1 MT263445.1 MT263444.1 MT263443.1 MT263442.1 MT263441.1 MT263440.1 MT263439.1 MT263438.1 MT263437.1 MT263436.1 MT263435.1 MT263434.1 MT263433.1 MT263432.1 MT263431.1 MT263430.1 MT263429.1 MT263428.1 MT263427.1 MT263426.1 MT263425.1 MT263424.1 MT263423.1 MT263422.1 MT263421.1 MT263420.1 MT263419.1 MT263418.1 MT263417.1 MT263416.1 MT263415.1 MT263414.1 MT263413.1 MT263412.1 MT263411.1 MT263410.1 MT263409.1 MT263408.1 MT263407.1 MT263406.1 MT263405.1 MT263404.1 MT263403.1 MT263402.1 MT252824.1 MT252823.1 MT252822.1 MT252821.1 MT252811.1 MT252810.1 MT252809.1 MT252808.1 MT252807.1 MT252806.1 MT252805.1 MT252804.1 MT252803.1 MT252802.1 MT252801.1 MT252800.1 MT252799.1 MT252798.1 MT252797.1 MT252795.1 MT252794.1 MT252793.1 MT252792.1 MT252788.1 MT252787.1 MT252785.1 MT252784.1 MT252783.1 MT252782.1 MT252781.1 MT252780.1 MT252779.1 MT252778.1 MT252775.1 MT252774.1 MT252773.1 MT252772.1 MT252771.1 MT252770.1 MT252769.1 MT252768.1 MT252767.1 MT252766.1 MT252765.1 MT252764.1 MT252763.1 MT252761.1 MT252758.1 MT252757.1 MT252756.1 MT252749.1 MT252748.1 MT252747.1 MT252746.1 MT252745.1 MT252742.1 MT252741.1 MT252740.1 MT252739.1 MT252738.1 MT252737.1 MT252736.1 MT252735.1 MT252734.1 MT252733.1 MT252730.1 MT252729.1 MT252728.1 MT252726.1 MT252725.1 MT252723.1 MT252721.1 MT252719.1 MT252717.1 MT252716.1 MT252715.1 MT252713.1 MT252712.1 MT252711.1 MT252710.1 MT252709.1 MT252708.1 MT252707.1 MT252706.1 MT252705.1 MT252704.1 MT252703.1 MT252702.1 MT252701.1 MT252700.1 MT252699.1 MT252698.1 MT252697.1 MT252696.1 MT252695.1 MT252694.1 MT252693.1 MT252692.1 MT252691.1 MT252690.1 MT252689.1 MT252688.1 MT252687.1 MT252686.1 MT252685.1 MT252684.1 MT252683.1 MT252682.1 MT252681.1 MT252680.1 MT252679.1 MT252678.1 MT252677.1 MT339043.1 MT375483.1 MT375482.1 MT375481.1 MT375480.1 MT375479.1 MT375478.1 MT375477.1 MT375476.1 MT375475.1 MT375474.1 MT375473.1 MT375472.1 MT375471.1 MT375470.1 MT375469.1 MT375468.1 MT375467.1 MT375466.1 MT375465.1 MT375464.1 MT375463.1 MT375462.1 MT375461.1 MT375460.1 MT375459.1 MT375458.1 MT375457.1 MT375456.1 MT375455.1 MT375454.1 MT375453.1 MT375452.1 MT375451.1 MT375450.1 MT375449.1 MT375448.1 MT375447.1 MT375446.1 MT375445.1 MT375444.1 MT375443.1 MT375442.1 MT375441.1 MT375440.1 MT375439.1 MT375438.1 MT375437.1 MT375436.1 MT375435.1 MT375434.1 MT375433.1 MT375432.1 MT375431.1 MT375430.1 MT375429.1 MT375428.1 MT374116.1 MT374115.1 MT374114.1 MT374113.1 MT374112.1 MT374111.1 MT374110.1 MT374109.1 MT374108.1 MT374107.1 MT374106.1 MT374105.1 MT374104.1 MT374103.1 MT374102.1 MT374101.1 MT320538.2 MT365033.1 MT186683.1 MT371038.1 MT371037.1 MT371036.1 MT371035.1 MT371034.1 MT371033.1 MT371032.1 MT371031.1 MT371030.1 MT371029.1 MT371028.1 MT371027.1 MT371026.1 MT371025.1 MT371024.1 MT371023.1 MT371022.1 MT371021.1 MT371020.1 MT371019.1 MT371018.1 MT371017.1 MT371016.1 MT371015.1 MT371014.1 MT371013.1 MT371012.1 MT371011.1 MT371010.1 MT371009.1 MT371008.1 MT371007.1 MT371006.1 MT371005.1 MT371004.1 MT371003.1 MT371002.1 MT371001.1 MT371000.1 MT370999.1 MT370998.1 MT370997.1 MT370996.1 MT370995.1 MT370994.1 MT370993.1 MT370992.1 MT370991.1 MT370990.1 MT370989.1 MT370988.1 MT370987.1 MT370986.1 MT370985.1 MT370984.1 MT370983.1 MT370982.1 MT370981.1 MT370980.1 MT370979.1 MT370978.1 MT370977.1 MT370976.1 MT370975.1 MT370974.1 MT370973.1 MT370972.1 MT370971.1 MT370970.1 MT370969.1 MT370968.1 MT370967.1 MT370966.1 MT370965.1 MT370964.1 MT370963.1 MT370962.1 MT370961.1 MT370960.1 MT370959.1 MT370958.1 MT370957.1 MT370956.1 MT370955.1 MT370954.1 MT370953.1 MT370952.1 MT370951.1 MT370950.1 MT370949.1 MT370948.1 MT370947.1 MT370946.1 MT370945.1 MT370944.1 MT370943.1 MT370942.1 MT370941.1 MT370940.1 MT370939.1 MT370938.1 MT370937.1 MT370936.1 MT370935.1 MT370934.1 MT370933.1 MT370932.1 MT370931.1 MT370930.1 MT370929.1 MT370928.1 MT370927.1 MT370926.1 MT370925.1 MT370924.1 MT370923.1 MT370922.1 MT370921.1 MT370920.1 MT370919.1 MT370918.1 MT370917.1 MT370916.1 MT370915.1 MT370914.1 MT370913.1 MT370912.1 MT370911.1 MT370910.1 MT370909.1 MT370908.1 MT370907.1 MT370906.1 MT370905.1 MT370904.1 MT370903.1 MT370902.1 MT370901.1 MT370900.1 MT370899.1 MT370898.1 MT370897.1 MT370896.1 MT370895.1 MT370894.1 MT370893.1 MT370892.1 MT370891.1 MT370890.1 MT370889.1 MT370888.1 MT370887.1 MT370886.1 MT370885.1 MT370884.1 MT370883.1 MT370882.1 MT370881.1 MT370880.1 MT370879.1 MT370878.1 MT370877.1 MT370876.1 MT370875.1 MT370874.1 MT370873.1 MT370872.1 MT370871.1 MT370870.1 MT370869.1 MT370868.1 MT370867.1 MT370866.1 MT370865.1 MT370864.1 MT370863.1 MT370862.1 MT370861.1 MT370860.1 MT370859.1 MT370858.1 MT370857.1 MT370856.1 MT370855.1 MT370854.1 MT370853.1 MT370852.1 MT370851.1 MT370850.1 MT370849.1 MT370848.1 MT370847.1 MT370846.1 MT370845.1 MT370844.1 MT370843.1 MT370842.1 MT370841.1 MT370840.1 MT370839.1 MT370838.1 MT370837.1 MT370836.1 MT370835.1 MT370834.1 MT370833.1 MT370832.1 MT370831.1 MT372483.1 MT372482.1 MT372481.1 MT372480.1 MT371574.1 MT371573.1 MT371572.1 MT371571.1 MT371570.1 MT371569.1 MT371568.1 MT371050.1 MT371049.1 MT371048.1 MT371047.1 MT370518.1 MT370517.1 MT370516.1 LC542976.1 7BV2_T 7BV2_P LC542809.1 MT291835.2 MT359231.1 MT358402.1 MT358401.1 MT359866.1 MT359865.1 MT230904.1 MT358748.1 MT358747.1 MT358746.1 MT358745.1 MT358744.1 MT358743.1 MT358742.1 MT358741.1 MT358740.1 MT358739.1 MT358738.1 MT358737.1 MT358736.1 MT358735.1 MT358734.1 MT358733.1 MT358732.1 MT358731.1 MT358730.1 MT358729.1 MT358728.1 MT358727.1 MT358726.1 MT358725.1 MT358724.1 MT358723.1 MT358722.1 MT358721.1 MT358720.1 MT358719.1 MT358718.1 MT358717.1 MT358716.1 MT358715.1 MT358714.1 MT358713.1 MT358712.1 MT358711.1 MT358710.1 MT358709.1 MT358708.1 MT358707.1 MT358706.1 MT358705.1 MT358704.1 MT358703.1 MT358702.1 MT358701.1 MT358700.1 MT358699.1 MT358698.1 MT358697.1 MT358696.1 MT358695.1 MT358694.1 MT358693.1 MT358692.1 MT358691.1 MT358690.1 MT358689.1 MT358688.1 MT358687.1 MT358686.1 MT358685.1 MT358684.1 MT358683.1 MT358682.1 MT358681.1 MT358680.1 MT358679.1 MT358678.1 MT358677.1 MT358676.1 MT358675.1 MT358674.1 MT358673.1 MT358672.1 MT358671.1 MT358670.1 MT358669.1 MT358668.1 MT358667.1 MT358666.1 MT358665.1 MT358664.1 MT358663.1 MT358662.1 MT358661.1 MT358660.1 MT358659.1 MT358658.1 MT358657.1 MT358656.1 MT358655.1 MT358654.1 MT358653.1 MT358652.1 MT358651.1 MT358650.1 MT358649.1 MT358648.1 MT358647.1 MT358646.1 MT358645.1 MT358644.1 MT358637.1 MT308700.1 MT308699.1 MT308698.1 MT308697.1 MT308696.1 MT308695.1 MT308694.1 MT308693.1 MT308692.1 MT291836.1 MT291834.1 MT291833.1 MT291832.1 MT291831.1 MT291830.1 MT291829.1 MT291828.1 MT291827.1 MT291826.1 MT114419.1 MT114418.1 MT114417.1 MT114416.1 MT114415.1 MT114414.1 MT114413.1 MT114412.1 MT350257.1 MT350256.1 MT350255.1 MT350254.1 MT350253.1 MT350252.1 MT350251.1 MT350250.1 MT350249.1 MT350248.1 MT350247.1 MT350246.1 MT350245.1 MT350244.1 MT350243.1 MT350242.1 MT350241.1 MT350240.1 MT350239.1 MT350238.1 MT350237.1 MT350236.1 MT350234.1 MT350280.1 MT350279.1 MT350278.1 MT350277.1 MT350276.1 MT350275.1 MT350274.1 MT350273.1 MT350272.1 MT350271.1 MT350270.1 MT350269.1 MT350268.1 MT350267.1 MT350266.1 MT350265.1 MT350264.1 MT350263.1 MT350282.1 MT344963.1 MT344962.1 MT344961.1 MT344960.1 MT344959.1 MT344958.1 MT344957.1 MT344956.1 MT344955.1 MT344954.1 MT344953.1 MT344952.1 MT344951.1 MT344950.1 MT344949.1 MT344948.1 MT344947.1 MT344946.1 MT344945.1 MT344944.1 MT344135.1 MT345888.1 MT345887.1 MT345886.1 MT345885.1 MT345884.1 MT345883.1 MT345882.1 MT345881.1 MT345880.1 MT345879.1 MT345878.1 MT345877.1 MT345876.1 MT345875.1 MT345874.1 MT345873.1 MT345872.1 MT345871.1 MT345870.1 MT345869.1 MT345868.1 MT345867.1 MT345866.1 MT345865.1 MT345864.1 MT345863.1 MT345862.1 MT345861.1 MT345860.1 MT345859.1 MT345858.1 MT345857.1 MT345856.1 MT345855.1 MT345854.1 MT345853.1 MT345852.1 MT345851.1 MT345850.1 MT345849.1 MT345848.1 MT345847.1 MT345846.1 MT345845.1 MT345844.1 MT345843.1 MT345842.1 MT345841.1 MT345840.1 MT345839.1 MT345838.1 MT345837.1 MT345836.1 MT345835.1 MT345834.1 MT345833.1 MT345832.1 MT345831.1 MT345830.1 MT345829.1 MT345828.1 MT345827.1 MT345826.1 MT345825.1 MT345824.1 MT345823.1 MT345822.1 MT345821.1 MT345820.1 MT345819.1 MT345818.1 MT345817.1 MT345816.1 MT345815.1 MT345814.1 MT345813.1 MT345812.1 MT345811.1 MT345810.1 MT345809.1 MT345808.1 MT345807.1 MT345806.1 MT345805.1 MT345804.1 MT345803.1 MT345802.1 MT345801.1 MT345800.1 MT345799.1 MT345798.1 MT320891.2 MT281530.2 MT339041.1 MT339040.1 MT339039.1 MT293225.1 MT293224.1 MT293223.1 MT293222.1 MT293221.1 MT293220.1 MT293219.1 MT293218.1 MT293217.1 MT293216.1 MT293215.1 MT293214.1 MT293213.1 MT293212.1 MT293211.1 MT293210.1 MT293209.1 MT293208.1 MT293207.1 MT293206.1 MT293205.1 MT293204.1 MT293203.1 MT293202.1 MT293201.1 MT293200.1 MT293199.1 MT293198.1 MT293197.1 MT293196.1 MT293195.1 MT293194.1 MT293193.1 MT293192.1 MT293191.1 MT293190.1 MT293189.1 MT293188.1 MT293187.1 MT293186.1 MT293185.1 MT293184.1 MT293183.1 MT293182.1 MT293181.1 MT293180.1 MT293179.1 MT293178.1 MT293177.1 MT293176.1 MT293175.1 MT293174.1 MT293173.1 MT293172.1 MT293171.1 MT293170.1 MT293169.1 MT293168.1 MT293167.1 MT293166.1 MT293165.1 MT293164.1 MT293163.1 MT293162.1 MT293161.1 MT293160.1 MT293159.1 MT293158.1 MT293157.1 MT293156.1 MT263469.1 MT263468.1 MT263467.1 MT263466.1 MT263465.1 MT263464.1 MT263463.1 MT263462.1 MT263461.1 MT263460.1 MT263459.1 MT263401.1 MT263400.1 MT263399.1 MT263398.1 MT263397.1 MT263396.1 MT263395.1 MT263394.1 MT263393.1 MT263392.1 MT263391.1 MT263390.1 MT263389.1 MT263388.1 MT263387.1 MT263386.1 MT263385.1 MT263384.1 MT263383.1 MT263382.1 MT263381.1 MT259287.1 MT259286.1 MT259285.1 MT259284.1 MT259283.1 MT259282.1 MT259281.1 MT259280.1 MT259279.1 MT259278.1 MT259277.1 MT259276.1 MT259275.1 MT259274.1 MT259273.1 MT259272.1 MT259271.1 MT259270.1 MT259269.1 MT259268.1 MT259267.1 MT259266.1 MT259265.1 MT259264.1 MT259263.1 MT259262.1 MT259261.1 MT259260.1 MT259259.1 MT259258.1 MT259257.1 MT259256.1 MT259255.1 MT259254.1 MT259253.1 MT259252.1 MT259251.1 MT259250.1 MT259249.1 MT259248.1 MT259247.1 MT259246.1 MT259245.1 MT259244.1 MT259243.1 MT259242.1 MT259241.1 MT259240.1 MT259239.1 MT259238.1 MT259237.1 MT259236.1 MT259235.1 MT251980.1 MT251979.1 MT251978.1 MT251977.1 MT251976.1 MT251975.1 MT251974.1 MT251973.1 MT251972.1 MT334573.1 MT334572.1 MT334571.1 MT334570.1 MT334569.1 MT334568.1 MT334567.1 MT334566.1 MT334565.1 MT334564.1 MT334563.1 MT334562.1 MT334561.1 MT334560.1 MT334559.1 MT334558.1 MT334557.1 MT334556.1 MT334555.1 MT334554.1 MT334553.1 MT334552.1 MT334551.1 MT334550.1 MT334549.1 MT334548.1 MT334547.1 MT334546.1 MT334545.1 MT334544.1 MT334543.1 MT334542.1 MT334541.1 MT334540.1 MT334539.1 MT334538.1 MT334537.1 MT334536.1 MT334535.1 MT334534.1 MT334533.1 MT334532.1 MT334531.1 MT334530.1 MT334529.1 MT334528.1 MT334527.1 MT334526.1 MT334525.1 MT334524.1 MT334523.1 MT334522.1 MT328035.1 MT328034.1 MT328033.1 MT328032.1 MT327745.1 MT325640.1 MT325639.1 MT325638.1 MT325637.1 MT325636.1 MT325635.1 MT325634.1 MT325633.1 MT325632.1 MT325631.1 MT325630.1 MT325629.1 MT325628.1 MT325627.1 MT325626.1 MT325625.1 MT325624.1 MT325623.1 MT325622.1 MT325621.1 MT325620.1 MT325619.1 MT325618.1 MT325617.1 MT325616.1 MT325615.1 MT325614.1 MT325613.1 MT325612.1 MT325611.1 MT325610.1 MT325609.1 MT325608.1 MT325607.1 MT325606.1 MT325605.1 MT325604.1 MT325603.1 MT325602.1 MT325601.1 MT325600.1 MT325599.1 MT325598.1 MT325597.1 MT325596.1 MT325595.1 MT325594.1 MT325593.1 MT325592.1 MT325591.1 MT325590.1 MT325589.1 MT325588.1 MT325587.1 MT325586.1 MT325585.1 MT325584.1 MT325583.1 MT325582.1 MT325581.1 MT325580.1 MT325579.1 MT325578.1 MT325577.1 MT325576.1 MT325575.1 MT325574.1 MT325573.1 MT325572.1 MT325571.1 MT325570.1 MT325569.1 MT325568.1 MT325567.1 MT325566.1 MT325565.1 MT325564.1 MT325563.1 MT325562.1 MT325561.1 MT324062.1 MT326191.1 MT326190.1 MT326189.1 MT326188.1 MT326187.1 MT326186.1 MT326185.1 MT326184.1 MT326183.1 MT326182.1 MT326181.1 MT326180.1 MT326179.1 MT326178.1 MT326177.1 MT326176.1 MT326175.1 MT326174.1 MT326173.1 MT326172.1 MT326171.1 MT326170.1 MT326169.1 MT326168.1 MT326167.1 MT326166.1 MT326165.1 MT326164.1 MT326163.1 MT326162.1 MT326161.1 MT326160.1 MT326159.1 MT326158.1 MT326157.1 MT326156.1 MT326155.1 MT326154.1 MT326153.1 MT326152.1 MT326151.1 MT326150.1 MT326149.1 MT326148.1 MT326147.1 MT326146.1 MT326145.1 MT326144.1 MT326143.1 MT326142.1 MT326141.1 MT326140.1 MT326139.1 MT326138.1 MT326137.1 MT326136.1 MT326135.1 MT326134.1 MT326133.1 MT326132.1 MT326131.1 MT326130.1 MT326129.1 MT326128.1 MT326127.1 MT326126.1 MT326125.1 MT326124.1 MT326123.1 MT326122.1 MT326121.1 MT326120.1 MT326119.1 MT326118.1 MT326117.1 MT326116.1 MT326115.1 MT326114.1 MT326113.1 MT326112.1 MT326111.1 MT326110.1 MT326109.1 MT326108.1 MT326107.1 MT326106.1 MT326105.1 MT326104.1 MT326103.1 MT326102.1 MT326101.1 MT326100.1 MT326099.1 MT326098.1 MT326097.1 MT326096.1 MT326095.1 MT326094.1 MT326093.1 MT326092.1 MT326091.1 MT326090.1 MT326089.1 MT326088.1 MT326087.1 MT326086.1 MT326085.1 MT326084.1 MT326083.1 MT326082.1 MT326081.1 MT326080.1 MT326079.1 MT326078.1 MT326077.1 MT326076.1 MT326075.1 MT326074.1 MT326073.1 MT326072.1 MT326071.1 MT326070.1 MT326069.1 MT326068.1 MT326067.1 MT326066.1 MT326065.1 MT326064.1 MT326063.1 MT326062.1 MT326061.1 MT326060.1 MT326059.1 MT326058.1 MT326057.1 MT326056.1 MT326055.1 MT326054.1 MT326053.1 MT326052.1 MT326051.1 MT326050.1 MT326049.1 MT326048.1 MT326047.1 MT326046.1 MT326045.1 MT326044.1 MT326043.1 MT326042.1 MT326041.1 MT326040.1 MT326039.1 MT326038.1 MT326037.1 MT326036.1 MT326035.1 MT326034.1 MT326033.1 MT326032.1 MT326031.1 MT326030.1 MT326029.1 MT326028.1 MT326027.1 MT326026.1 MT326025.1 MT326024.1 MT326023.1 MT324683.1 MT324682.1 MT324681.1 MT324680.1 MT324679.1 MT324684.1 MT256924.2 MT066156.1 MT008023.1 MT008022.1 MT322424.1 MT322423.1 MT322422.1 MT322421.1 MT322420.1 MT322419.1 MT322418.1 MT322417.1 MT322416.1 MT322415.1 MT322414.1 MT322413.1 MT322412.1 MT322411.1 MT322410.1 MT322409.1 MT322408.1 MT322407.1 MT322406.1 MT322405.1 MT322404.1 MT322403.1 MT322402.1 MT322401.1 MT322400.1 MT322399.1 MT322398.1 MT322397.1 MT322396.1 MT322395.1 MT322394.1 MT077125.1 MT039874.1 CADDYA000000000.1 MT300186.2 MT308704.1 MT308703.1 MT308702.1 MT308701.1 MT304491.1 MT304490.1 MT304489.1 MT304488.1 MT304487.1 MT304486.1 MT304485.1 MT304484.1 MT304483.1 MT304482.1 MT304481.1 MT304480.1 MT304479.1 MT304478.1 MT304477.1 MT304476.1 MT304475.1 MT304474.1 MT293547.1 MT295465.1 MT295464.1 MT281577.1 MT273658.1 MT292582.1 MT292581.1 MT292580.1 MT292579.1 MT292578.1 MT292577.1 MT292576.1 MT292575.1 MT292574.1 MT292573.1 MT292572.1 MT292571.1 MT292570.1 MT292569.1 MT276598.1 MT276597.1 MT276331.1 MT276330.1 MT276329.1 MT276328.1 MT276327.1 MT276326.1 MT276325.1 MT276324.1 MT276323.1 MT262916.1 MT262915.1 MT262914.1 MT262913.1 MT262912.1 MT262911.1 MT262910.1 MT262909.1 MT262908.1 MT262907.1 MT262906.1 MT262905.1 MT262904.1 MT262903.1 MT262902.1 MT262901.1 MT262900.1 MT262899.1 MT262898.1 MT262897.1 MT262896.1 MT263074.1 MT262993.1 MT256918.1 MT256917.1 MT258383.1 MT258382.1 MT258381.1 MT258380.1 MT258379.1 MT258378.1 MT258377.1 MT259231.1 MT259230.1 MT259229.1 MT259228.1 MT259227.1 MT259226.1 MT198652.2 MT253710.1 MT253709.1 MT253708.1 MT253707.1 MT253706.1 MT253705.1 MT253704.1 MT253703.1 MT253702.1 MT253701.1 MT253700.1 MT253699.1 MT253698.1 MT253697.1 MT253696.1 MT246490.1 MT246489.1 MT246488.1 MT246487.1 MT246486.1 MT246485.1 MT246484.1 MT246483.1 MT246482.1 MT246481.1 MT246480.1 MT246479.1 MT246478.1 MT246477.1 MT246476.1 MT246475.1 MT246474.1 MT246473.1 MT246472.1 MT246471.1 MT246470.1 MT246469.1 MT246468.1 MT246467.1 MT246466.1 MT246465.1 MT246464.1 MT246463.1 MT246462.1 MT246461.1 MT246460.1 MT246459.1 MT246458.1 MT246457.1 MT246456.1 MT246455.1 MT246454.1 MT246453.1 MT246452.1 MT246451.1 MT246450.1 MT246449.1 MT240479.1 MT233523.1 MT233522.1 MT233521.1 MT233520.1 MT233519.1 MT232872.1 MT232871.1 MT232870.1 MT232869.1 MT226610.1 MT198653.1 MT198651.1 MT192773.1 MT192772.1 MT192765.1 MT192759.1 MT192758.1 MT186682.1 MT186681.1 MT186680.1 MT186679.1 MT186678.1 MT186677.1 MT186676.1 MT187977.1 MT123293.2 MT123292.2 MT123291.2 MT093631.2 MT163721.1 MT163720.1 MT163719.1 MT163718.1 MT163717.1 MT163716.1 MT163712.1 MT163738.1 MT163737.1 MT163715.1 MT163714.1 MT161607.1 MT159778.1 MT121215.1 MT152824.1 MT152900.1 MT135044.1 MT135043.1 MT135042.1 MT135041.1 MT126808.1 MT127116.1 MT127115.1 MT127114.1 MT127113.1 MT123290.1 MT111896.1 MT111895.1 MT093571.1 MT081068.1 MT081067.1 MT081066.1 MT081065.1 MT081064.1 MT081063.1 MT081062.1 MT081061.1 MT081060.1 MT081059.1 MT072667.1 MT072668.1 MT072688.1 MT066176.1 MT066175.1 MT066159.1 MT066158.1 MT066157.1 MT050493.1 MT050417.1 MT050416.1 MT050415.1 MT050414.1 MT049951.1 MT042778.1 MT042777.1 MT042776.1 MT042775.1 MT042774.1 MT042773.1 MT012098.1 MT246667.1 MT233526.1 NC_045512.2 LC534419.1 LC534418.1 MN985325.1 MT020781.2 MN996531.1 MN996530.1 MN996529.1 MN996528.1 MN996527.1 MN908947.3 MT188341.1 MT188340.1 MT188339.1 MT184913.1 MT184912.1 MT184911.1 MT184910.1 MT184909.1 MT184908.1 MT184907.1 LC529905.1 LR757998.1 LR757997.1 LR757996.1 LR757995.1 MT159722.1 MT159721.1 MT159720.1 MT159719.1 MT159718.1 MT159717.1 MT159716.1 MT159715.1 MT159714.1 MT159713.1 MT159712.1 MT159711.1 MT159710.1 MT159709.1 MT159708.1 MT159707.1 MT159706.1 MT159705.1 LC528233.1 LC528232.1 MT118835.1 MT106054.1 MT106053.1 MT106052.1 LC523809.1 LC523808.1 LC523807.1 MT044258.1 MT044257.1 MT039888.1 MT039887.1 MT039890.1 MT039873.1 LC522350.1 MT027064.1 MT027063.1 MT027062.1 MT020881.1 MT020880.1 MT019533.1 MT019532.1 MT019531.1 MT019530.1 MT019529.1 MT007544.1 MN997409.1 MN994468.1 MN994467.1 MN988713.1 MN988669.1 MN988668.1 MN975268.1 MN975267.1 MN975266.1 MN975265.1 MN975264.1 MN975263.1 MN975262.1 MN938390.1 MN938389.1 MN938388.1 MN938387.1 MN938386.1 MN938385.1 MN938384.1 MN970004.1 MN970003.1 NC_045512 "
     ]
    }
   ],
   "source": [
    "nuc_ids = nuc_accessions['IdList']\n",
    "nuc_ids.append('NC_045512') # SARS-CoV-2 refseq is missing from list (why??). Add it back.\n",
    "df_list = []\n",
    "\n",
    "for nuc in nuc_ids:\n",
    "    print(nuc, end=' ')\n",
    "    df_list.append(process_nucleotide_record(nuc))\n",
    "    \n",
    "df1 = pd.concat(df_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's work on a copy of the original data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>host</th>\n",
       "      <th>isolate</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>complete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29880</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Mustela lutreola</td>\n",
       "      <td>SARS-CoV-2/mink/NLD/1/2020</td>\n",
       "      <td>lung</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-04-24</td>\n",
       "      <td>Netherlands: Milheeze</td>\n",
       "      <td>MT396266.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29872</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X003/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-31</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394531.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29870</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X004/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-31</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394530.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29863</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X001/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394529.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29867</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X002/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394528.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start    end        db_xref              host  \\\n",
       "0      1  29880  taxon:2697049  Mustela lutreola   \n",
       "0      1  29872  taxon:2697049      Homo sapiens   \n",
       "0      1  29870  taxon:2697049      Homo sapiens   \n",
       "0      1  29863  taxon:2697049      Homo sapiens   \n",
       "0      1  29867  taxon:2697049      Homo sapiens   \n",
       "\n",
       "                                   isolate                   isolation_source  \\\n",
       "0               SARS-CoV-2/mink/NLD/1/2020                               lung   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X003/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X004/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X001/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X002/2020  nasopharyngeal/oropharyngeal swab   \n",
       "\n",
       "      mol_type collection_date                country genbank_accession  \\\n",
       "0  genomic RNA      2020-04-24  Netherlands: Milheeze        MT396266.1   \n",
       "0  genomic RNA      2020-03-31                USA: CA        MT394531.1   \n",
       "0  genomic RNA      2020-03-31                USA: CA        MT394530.1   \n",
       "0  genomic RNA      2020-03-24                USA: CA        MT394529.1   \n",
       "0  genomic RNA      2020-03-24                USA: CA        MT394528.1   \n",
       "\n",
       "   complete  \n",
       "0      True  \n",
       "0      True  \n",
       "0      True  \n",
       "0      True  \n",
       "0      True  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df1.copy()\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter out incomplete records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# incomplete genomes\n",
    "df.query('complete', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter out any records that don't match the tax id, \n",
    "# e.g., tax_ids from organisms that are lower in the taxonomy tree.\n",
    "df.query(f\"db_xref == 'taxon:{tax_id}'\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>host</th>\n",
       "      <th>isolate</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>complete</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29880</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Mustela lutreola</td>\n",
       "      <td>SARS-CoV-2/mink/NLD/1/2020</td>\n",
       "      <td>lung</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-04-24</td>\n",
       "      <td>Netherlands: Milheeze</td>\n",
       "      <td>MT396266.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29872</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X003/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-31</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394531.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29870</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X004/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-31</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394530.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29863</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X001/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394529.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29867</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X002/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394528.1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start    end        db_xref              host  \\\n",
       "0      1  29880  taxon:2697049  Mustela lutreola   \n",
       "0      1  29872  taxon:2697049      Homo sapiens   \n",
       "0      1  29870  taxon:2697049      Homo sapiens   \n",
       "0      1  29863  taxon:2697049      Homo sapiens   \n",
       "0      1  29867  taxon:2697049      Homo sapiens   \n",
       "\n",
       "                                   isolate                   isolation_source  \\\n",
       "0               SARS-CoV-2/mink/NLD/1/2020                               lung   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X003/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X004/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X001/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X002/2020  nasopharyngeal/oropharyngeal swab   \n",
       "\n",
       "      mol_type collection_date                country genbank_accession  \\\n",
       "0  genomic RNA      2020-04-24  Netherlands: Milheeze        MT396266.1   \n",
       "0  genomic RNA      2020-03-31                USA: CA        MT394531.1   \n",
       "0  genomic RNA      2020-03-31                USA: CA        MT394530.1   \n",
       "0  genomic RNA      2020-03-24                USA: CA        MT394529.1   \n",
       "0  genomic RNA      2020-03-24                USA: CA        MT394528.1   \n",
       "\n",
       "   complete  \n",
       "0      True  \n",
       "0      True  \n",
       "0      True  \n",
       "0      True  \n",
       "0      True  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create unique and interoperable identifiers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**id**: CURIE: [insdc](https://registry.identifiers.org/registry/insdc) (International Nucleotide Sequence Database Collaboration, INSDC)\n",
    "\n",
    "**id**: CURIE: [ncbiprotein](https://registry.identifiers.org/registry/ncbiprotein) (NCBI Reference Sequences, Refseq)\n",
    "\n",
    "**taxonomy_id**: CURIE: [taxonomy](https://registry.identifiers.org/registry/taxonomy) (NCBI Taxonomy)\n",
    "\n",
    "A [CURIE](https://en.wikipedia.org/wiki/CURIE) (Compact URI) is a compact abbreviation for Uniform Resource Identifiers (URIs) that can be resolved by [Identifiers.org](https://identifiers.org/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove version number from genbank accession to enable linking with other nodes\n",
    "df['id'] = \"insdc:\" + df['genbank_accession'].apply(lambda s: s.split('.')[0])\n",
    "# # NCBI reference sequences resolve through the ncbiprotein CURIE\n",
    "df['id'] = df['id'].str.replace('insdc:NC_', 'ncbiprotein:NC_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse taxonomy_id from db_xref record\n",
    "df['taxonomy_id'] = 'taxonomy:' + df['db_xref'].str.split(':', expand=True)[1]\n",
    "df.rename(columns={'isolate': 'name'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Standardize data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: date standardization introduces artifacts, e.g. Dec 2019 -> 2019-12-01\n",
    "# Add column that specifies time granularity: Y, M, D\n",
    "df['collection_date'] = df['collection_date'].apply(lambda d: dateutil.parser.parse(d) if len(d) > 0 else '')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Standardize host organism and demographics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO a find general solution to map host name to NCBI taxonomy**\n",
    "\n",
    "# some organism specifications are ambiguous, \n",
    "# they don't match a specificn NCBI taxonomy\n",
    "\n",
    "taxonomy_to_id = {'Human': '9606', \n",
    "                  'Homo sapiens': '9606',\n",
    "                  'Rhinolophus affinis': '59477', \n",
    "                  'Mustela lutreola': '9666',\n",
    "                  'Panthera tigris jacksoni': '419130',\n",
    "                  'Rhinolophus sp. (bat)': '49442', # ambiguous\n",
    "                  'bat': 'taxonomy:49442', # ambiguous\n",
    "                  'Manis javanica': '9974',\n",
    "                  'palm civet': '71116', # ambiguous\n",
    "                  'Canine': '9608' # ambiguous\n",
    "                 }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Split host record into host, sex, and age (e.g., Homo sapiens; female; age 40 -> Homo sapiens, female, 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add placeholders in case the host record is empty to enbable 3-way split\n",
    "df['host'] = df['host'].apply(lambda s: ' ; ; ' if len(s) < 1 else s)\n",
    "df['host'] = df['host'].str.replace('age','')\n",
    "df[['host','sex','age']] = df['host'].str.split(';', 2, expand=True)\n",
    "df['host'] = df['host'].str.strip()\n",
    "\n",
    "# assign taxonomy id to host\n",
    "df['host_taxonomy_id'] = 'taxonomy:' + df['host'].apply(lambda s: taxonomy_to_id.get(s, ''))\n",
    "df['sex'] = df['sex'].str.strip()\n",
    "df['age'] = df['age'].str.strip()\n",
    "df = df.fillna('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List entries where taxonomy cannot be assigned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>host</th>\n",
       "      <th>name</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>complete</th>\n",
       "      <th>id</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>host_taxonomy_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29900</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td></td>\n",
       "      <td>SARS-CoV-2/human/CHN/YN-0306-466/2020</td>\n",
       "      <td></td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-06</td>\n",
       "      <td>China</td>\n",
       "      <td>MT396241.1</td>\n",
       "      <td>True</td>\n",
       "      <td>insdc:MT396241</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>taxonomy:</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29897</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Panthera tigris jacksoni</td>\n",
       "      <td>SARS-CoV-2/tiger/NY/040420/2020</td>\n",
       "      <td>oral swab; nasal swab; tracheal wash</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-04-02</td>\n",
       "      <td>USA: New York</td>\n",
       "      <td>MT365033.1</td>\n",
       "      <td>True</td>\n",
       "      <td>insdc:MT365033</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>taxonomy:</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29889</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td></td>\n",
       "      <td>SARS-CoV-2/ENV/USA/UF-3/2020</td>\n",
       "      <td>breathing air using a VIVAS air sampler</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-25</td>\n",
       "      <td>USA</td>\n",
       "      <td>MT324684.1</td>\n",
       "      <td>True</td>\n",
       "      <td>insdc:MT324684</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>taxonomy:</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start    end        db_xref                      host  \\\n",
       "0      1  29900  taxon:2697049                             \n",
       "0      1  29897  taxon:2697049  Panthera tigris jacksoni   \n",
       "0      1  29889  taxon:2697049                             \n",
       "\n",
       "                                    name  \\\n",
       "0  SARS-CoV-2/human/CHN/YN-0306-466/2020   \n",
       "0        SARS-CoV-2/tiger/NY/040420/2020   \n",
       "0           SARS-CoV-2/ENV/USA/UF-3/2020   \n",
       "\n",
       "                          isolation_source     mol_type collection_date  \\\n",
       "0                                           genomic RNA      2020-03-06   \n",
       "0     oral swab; nasal swab; tracheal wash  genomic RNA      2020-04-02   \n",
       "0  breathing air using a VIVAS air sampler  genomic RNA      2020-03-25   \n",
       "\n",
       "         country genbank_accession  complete              id  \\\n",
       "0          China        MT396241.1      True  insdc:MT396241   \n",
       "0  USA: New York        MT365033.1      True  insdc:MT365033   \n",
       "0            USA        MT324684.1      True  insdc:MT324684   \n",
       "\n",
       "        taxonomy_id sex age host_taxonomy_id  \n",
       "0  taxonomy:2697049                taxonomy:  \n",
       "0  taxonomy:2697049                taxonomy:  \n",
       "0  taxonomy:2697049                taxonomy:  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query(\"host_taxonomy_id == 'taxonomy:'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>host</th>\n",
       "      <th>name</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>genbank_accession</th>\n",
       "      <th>complete</th>\n",
       "      <th>id</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>host_taxonomy_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29880</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Mustela lutreola</td>\n",
       "      <td>SARS-CoV-2/mink/NLD/1/2020</td>\n",
       "      <td>lung</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-04-24</td>\n",
       "      <td>Netherlands: Milheeze</td>\n",
       "      <td>MT396266.1</td>\n",
       "      <td>True</td>\n",
       "      <td>insdc:MT396266</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>taxonomy:</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29872</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X003/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-31</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394531.1</td>\n",
       "      <td>True</td>\n",
       "      <td>insdc:MT394531</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29870</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X004/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-31</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394530.1</td>\n",
       "      <td>True</td>\n",
       "      <td>insdc:MT394530</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29863</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X001/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394529.1</td>\n",
       "      <td>True</td>\n",
       "      <td>insdc:MT394529</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>29867</td>\n",
       "      <td>taxon:2697049</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X002/2020</td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>USA: CA</td>\n",
       "      <td>MT394528.1</td>\n",
       "      <td>True</td>\n",
       "      <td>insdc:MT394528</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start    end        db_xref              host  \\\n",
       "0      1  29880  taxon:2697049  Mustela lutreola   \n",
       "0      1  29872  taxon:2697049      Homo sapiens   \n",
       "0      1  29870  taxon:2697049      Homo sapiens   \n",
       "0      1  29863  taxon:2697049      Homo sapiens   \n",
       "0      1  29867  taxon:2697049      Homo sapiens   \n",
       "\n",
       "                                      name                   isolation_source  \\\n",
       "0               SARS-CoV-2/mink/NLD/1/2020                               lung   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X003/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X004/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X001/2020  nasopharyngeal/oropharyngeal swab   \n",
       "0  SARS-CoV-2/human/USA/CA-CZB-59X002/2020  nasopharyngeal/oropharyngeal swab   \n",
       "\n",
       "      mol_type collection_date                country genbank_accession  \\\n",
       "0  genomic RNA      2020-04-24  Netherlands: Milheeze        MT396266.1   \n",
       "0  genomic RNA      2020-03-31                USA: CA        MT394531.1   \n",
       "0  genomic RNA      2020-03-31                USA: CA        MT394530.1   \n",
       "0  genomic RNA      2020-03-24                USA: CA        MT394529.1   \n",
       "0  genomic RNA      2020-03-24                USA: CA        MT394528.1   \n",
       "\n",
       "   complete              id       taxonomy_id sex age host_taxonomy_id  \n",
       "0      True  insdc:MT396266  taxonomy:2697049                taxonomy:  \n",
       "0      True  insdc:MT394531  taxonomy:2697049            taxonomy:9606  \n",
       "0      True  insdc:MT394530  taxonomy:2697049            taxonomy:9606  \n",
       "0      True  insdc:MT394529  taxonomy:2697049            taxonomy:9606  \n",
       "0      True  insdc:MT394528  taxonomy:2697049            taxonomy:9606  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assign locations\n",
    "**TODO replace this section with a general geoparsing solution**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fix inconsistencies and add missing location data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "locations = {\n",
    "    'India: Kerala State': 'India: Kerala',\n",
    "    'USA: Snohomish County, WA': 'USA: WA, Snohomish County',\n",
    "    'USA: San Francisco, CA': 'USA: CA, San Francisco',\n",
    "    'USA: Nashville, TN': 'USA: TN, Nashville',\n",
    "    'Canada: Toronto': 'Canada: Ontario, Toronto',\n",
    "    'China: Shenzhen': 'China: Guangdong, Shenzhen',\n",
    "    'China: Hangzhou': 'China: Zhejiang, Hangzhou',\n",
    "    'China: Wuhan': 'China: Hubei, Wuhan',\n",
    "    'Italy: Cagliari': 'Italy: Sardinia, Cagliari',\n",
    "    'Viet Nam: Ho Chi Minh city': 'Vietnam: Ho Chi Minh City'\n",
    "}\n",
    "df['country'] = df['country'].apply(lambda s: locations.get(s, s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "states = {\n",
    "        # US\n",
    "        'AK': 'Alaska',\n",
    "        'AL': 'Alabama',\n",
    "        'AR': 'Arkansas',\n",
    "        'AS': 'American Samoa',\n",
    "        'AZ': 'Arizona',\n",
    "        'CA': 'California',\n",
    "        'CO': 'Colorado',\n",
    "        'CT': 'Connecticut',\n",
    "        'DC': 'District of Columbia',\n",
    "        'DE': 'Delaware',\n",
    "        'FL': 'Florida',\n",
    "        'GA': 'Georgia',\n",
    "        'GU': 'Guam',\n",
    "        'HI': 'Hawaii',\n",
    "        'IA': 'Iowa',\n",
    "        'ID': 'Idaho',\n",
    "        'IL': 'Illinois',\n",
    "        'IN': 'Indiana',\n",
    "        'KS': 'Kansas',\n",
    "        'KY': 'Kentucky',\n",
    "        'LA': 'Louisiana',\n",
    "        'MA': 'Massachusetts',\n",
    "        'MD': 'Maryland',\n",
    "        'ME': 'Maine',\n",
    "        'MI': 'Michigan',\n",
    "        'MN': 'Minnesota',\n",
    "        'MO': 'Missouri',\n",
    "        'MP': 'Northern Mariana Islands',\n",
    "        'MS': 'Mississippi',\n",
    "        'MT': 'Montana',\n",
    "        'NA': 'National',\n",
    "        'NC': 'North Carolina',\n",
    "        'ND': 'North Dakota',\n",
    "        'NE': 'Nebraska',\n",
    "        'NH': 'New Hampshire',\n",
    "        'NJ': 'New Jersey',\n",
    "        'NM': 'New Mexico',\n",
    "        'NV': 'Nevada',\n",
    "        'NY': 'New York',\n",
    "        'OH': 'Ohio',\n",
    "        'OK': 'Oklahoma',\n",
    "        'OR': 'Oregon',\n",
    "        'PA': 'Pennsylvania',\n",
    "        'PR': 'Puerto Rico',\n",
    "        'RI': 'Rhode Island',\n",
    "        'SC': 'South Carolina',\n",
    "        'SD': 'South Dakota',\n",
    "        'TN': 'Tennessee',\n",
    "        'TX': 'Texas',\n",
    "        'UT': 'Utah',\n",
    "        'VA': 'Virginia',\n",
    "        'VI': 'Virgin Islands',\n",
    "        'VT': 'Vermont',\n",
    "        'WA': 'Washington',\n",
    "        'WI': 'Wisconsin',\n",
    "        'WV': 'West Virginia',\n",
    "        'WY': 'Wyoming',\n",
    "        # Canada\n",
    "        'AB': 'Alberta',\n",
    "        'BC': 'British Columbia',\n",
    "        'MB': 'Manitoba',\n",
    "        'NB': 'New Brunswick',\n",
    "        'NL': 'Newfoundland and Labrador',\n",
    "        'NT': 'Northwest Territories',\n",
    "        'NS': 'Nova Scotia',\n",
    "        'NU': 'Nunavut',\n",
    "        'ON': 'Ontario',\n",
    "        'PE': 'Prince Edward Island',\n",
    "        'QC': 'Quebec',\n",
    "        'SK': 'Saskatchewan',\n",
    "        'YT': 'Yukon'\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Split country records into administrative areas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add placeholders in case the country record is incomplete to enable 3-way split below\n",
    "df['country'] = df['country'].apply(lambda s: s if ':' in s else s + ': , ')\n",
    "df[['country','admin1','admin2']] = df['country'].str.split(':|,', 2, expand=True)\n",
    "df['country'] = df['country'].str.strip()\n",
    "df['admin2'] = df['admin2'].str.strip()\n",
    "df['admin1'] = df['admin1'].str.strip()\n",
    "\n",
    "# # expand states to full name\n",
    "df['admin1'] = df['admin1'].apply(lambda s: states.get(s, s))\n",
    "# create a unique ids\n",
    "# TODO use unique ids from Geonames.org\n",
    "df['country_id'] = df['country']\n",
    "df['admin1_id'] = df['country'] + '-' + df['admin1']\n",
    "df['admin2_id'] = df['country'] + '-' + df['admin1'] + '-' + df['admin2']\n",
    "\n",
    "df.fillna(value='', inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save data for Knowledge Graph Import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>host_taxonomy_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>isolation_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT396266</td>\n",
       "      <td>SARS-CoV-2/mink/NLD/1/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-04-24</td>\n",
       "      <td>taxonomy:</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT394531</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X003/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-03-31</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT394530</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X004/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-03-31</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT394529</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X001/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT394528</td>\n",
       "      <td>SARS-CoV-2/human/USA/CA-CZB-59X002/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>nasopharyngeal/oropharyngeal swab</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id                                     name       taxonomy_id  \\\n",
       "0  insdc:MT396266               SARS-CoV-2/mink/NLD/1/2020  taxonomy:2697049   \n",
       "0  insdc:MT394531  SARS-CoV-2/human/USA/CA-CZB-59X003/2020  taxonomy:2697049   \n",
       "0  insdc:MT394530  SARS-CoV-2/human/USA/CA-CZB-59X004/2020  taxonomy:2697049   \n",
       "0  insdc:MT394529  SARS-CoV-2/human/USA/CA-CZB-59X001/2020  taxonomy:2697049   \n",
       "0  insdc:MT394528  SARS-CoV-2/human/USA/CA-CZB-59X002/2020  taxonomy:2697049   \n",
       "\n",
       "  collection_date host_taxonomy_id sex age                   isolation_source  \n",
       "0      2020-04-24        taxonomy:                                       lung  \n",
       "0      2020-03-31    taxonomy:9606          nasopharyngeal/oropharyngeal swab  \n",
       "0      2020-03-31    taxonomy:9606          nasopharyngeal/oropharyngeal swab  \n",
       "0      2020-03-24    taxonomy:9606          nasopharyngeal/oropharyngeal swab  \n",
       "0      2020-03-24    taxonomy:9606          nasopharyngeal/oropharyngeal swab  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[['id','name', 'taxonomy_id', 'collection_date', 'host_taxonomy_id',\n",
    "         'sex', 'age', 'isolation_source']]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>host_taxonomy_id</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>isolation_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT396266</td>\n",
       "      <td>SARS-CoV-2/mink/NLD/1/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-04-24</td>\n",
       "      <td>taxonomy:</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT396241</td>\n",
       "      <td>SARS-CoV-2/human/CHN/YN-0306-466/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-03-06</td>\n",
       "      <td>taxonomy:</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT365033</td>\n",
       "      <td>SARS-CoV-2/tiger/NY/040420/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-04-02</td>\n",
       "      <td>taxonomy:</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>oral swab; nasal swab; tracheal wash</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>insdc:MT324684</td>\n",
       "      <td>SARS-CoV-2/ENV/USA/UF-3/2020</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>2020-03-25</td>\n",
       "      <td>taxonomy:</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>breathing air using a VIVAS air sampler</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               id                                   name       taxonomy_id  \\\n",
       "0  insdc:MT396266             SARS-CoV-2/mink/NLD/1/2020  taxonomy:2697049   \n",
       "0  insdc:MT396241  SARS-CoV-2/human/CHN/YN-0306-466/2020  taxonomy:2697049   \n",
       "0  insdc:MT365033        SARS-CoV-2/tiger/NY/040420/2020  taxonomy:2697049   \n",
       "0  insdc:MT324684           SARS-CoV-2/ENV/USA/UF-3/2020  taxonomy:2697049   \n",
       "\n",
       "  collection_date host_taxonomy_id sex age  \\\n",
       "0      2020-04-24        taxonomy:           \n",
       "0      2020-03-06        taxonomy:           \n",
       "0      2020-04-02        taxonomy:           \n",
       "0      2020-03-25        taxonomy:           \n",
       "\n",
       "                          isolation_source  \n",
       "0                                     lung  \n",
       "0                                           \n",
       "0     oral swab; nasal swab; tracheal wash  \n",
       "0  breathing air using a VIVAS air sampler  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(NEO4J_HOME / \"import/01a-NCBIStrain.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
