{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load SARS-CoV-2 Virus Strain Metadata from CNCB\n",
    "**[Work in progress]**\n",
    "\n",
    "This notebook downloads and standardizes viral strain data from CNCB for ingestion into a Knowledge Graph.\n",
    "\n",
    "Data source: [China National Center for Bioinformation, 2019 Novel Coronavirus Resource (2019nCoVR)](https://bigd.big.ac.cn/ncov/release_genome)\n",
    "\n",
    "Author: Peter Rose (pwrose@ucsd.edu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import glob\n",
    "import ftplib\n",
    "import re\n",
    "import requests\n",
    "import json\n",
    "import dateutil\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_rows = None  # display all rows\n",
    "pd.options.display.max_columns = None  # display all columsns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_url = \"https://bigd.big.ac.cn/ncov/genome/export/meta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/peter/Library/Application Support/com.Neo4j.Relate/data/dbmss/dbms-8bf637fc-0d20-4d9f-9c6f-f7e72e92a4da/import\n"
     ]
    }
   ],
   "source": [
    "# Path will take care of handling operating system differences.\n",
    "NEO4J_IMPORT = Path(os.getenv('NEO4J_IMPORT'))\n",
    "print(NEO4J_IMPORT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Download data on the fly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this values should be >= then the current number of isolates in CNCB\n",
    "# max_entries = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# url = f'https://bigd.big.ac.cn/ncov/genome/query?&columns%5B0%5D.data=accession&columns%5B0%5D.name=&columns%5B0%5D.searchable=false&columns%5B0%5D.orderable=false&columns%5B0%5D.search.value=&columns%5B0%5D.search.regex=false&columns%5B1%5D.data=name&columns%5B1%5D.name=&columns%5B1%5D.searchable=true&columns%5B1%5D.orderable=true&columns%5B1%5D.search.value=&columns%5B1%5D.search.regex=false&columns%5B2%5D.data=accession&columns%5B2%5D.name=&columns%5B2%5D.searchable=true&columns%5B2%5D.orderable=true&columns%5B2%5D.search.value=&columns%5B2%5D.search.regex=false&columns%5B3%5D.data=gender&columns%5B3%5D.name=&columns%5B3%5D.searchable=true&columns%5B3%5D.orderable=true&columns%5B3%5D.search.value=&columns%5B3%5D.search.regex=false&columns%5B4%5D.data=age&columns%5B4%5D.name=&columns%5B4%5D.searchable=true&columns%5B4%5D.orderable=true&columns%5B4%5D.search.value=&columns%5B4%5D.search.regex=false&columns%5B5%5D.data=source&columns%5B5%5D.name=&columns%5B5%5D.searchable=true&columns%5B5%5D.orderable=true&columns%5B5%5D.search.value=&columns%5B5%5D.search.regex=false&columns%5B6%5D.data=relatedAccession&columns%5B6%5D.name=&columns%5B6%5D.searchable=true&columns%5B6%5D.orderable=true&columns%5B6%5D.search.value=&columns%5B6%5D.search.regex=false&columns%5B7%5D.data=genomeLineage.lineage&columns%5B7%5D.name=&columns%5B7%5D.searchable=true&columns%5B7%5D.orderable=true&columns%5B7%5D.search.value=&columns%5B7%5D.search.regex=false&columns%5B8%5D.data=completeness&columns%5B8%5D.name=&columns%5B8%5D.searchable=true&columns%5B8%5D.orderable=true&columns%5B8%5D.search.value=&columns%5B8%5D.search.regex=false&columns%5B9%5D.data=genomeQuality&columns%5B9%5D.name=&columns%5B9%5D.searchable=false&columns%5B9%5D.orderable=true&columns%5B9%5D.search.value=&columns%5B9%5D.search.regex=false&columns%5B10%5D.data=genomeQuality&columns%5B10%5D.name=&columns%5B10%5D.searchable=false&columns%5B10%5D.orderable=true&columns%5B10%5D.search.value=&columns%5B10%5D.search.regex=false&columns%5B11%5D.data=genomeQuality&columns%5B11%5D.name=&columns%5B11%5D.searchable=false&columns%5B11%5D.orderable=true&columns%5B11%5D.search.value=&columns%5B11%5D.search.regex=false&columns%5B12%5D.data=host&columns%5B12%5D.name=&columns%5B12%5D.searchable=true&columns%5B12%5D.orderable=true&columns%5B12%5D.search.value=&columns%5B12%5D.search.regex=false&columns%5B13%5D.data=collectDate&columns%5B13%5D.name=&columns%5B13%5D.searchable=true&columns%5B13%5D.orderable=true&columns%5B13%5D.search.value=&columns%5B13%5D.search.regex=false&columns%5B14%5D.data=location&columns%5B14%5D.name=&columns%5B14%5D.searchable=true&columns%5B14%5D.orderable=true&columns%5B14%5D.search.value=&columns%5B14%5D.search.regex=false&columns%5B15%5D.data=dataProvider&columns%5B15%5D.name=&columns%5B15%5D.searchable=true&columns%5B15%5D.orderable=true&columns%5B15%5D.search.value=&columns%5B15%5D.search.regex=false&columns%5B16%5D.data=submitDate&columns%5B16%5D.name=&columns%5B16%5D.searchable=true&columns%5B16%5D.orderable=true&columns%5B16%5D.search.value=&columns%5B16%5D.search.regex=false&columns%5B17%5D.data=dataSubmitter&columns%5B17%5D.name=&columns%5B17%5D.searchable=true&columns%5B17%5D.orderable=true&columns%5B17%5D.search.value=&columns%5B17%5D.search.regex=false&columns%5B18%5D.data=createDate&columns%5B18%5D.name=&columns%5B18%5D.searchable=true&columns%5B18%5D.orderable=true&columns%5B18%5D.search.value=&columns%5B18%5D.search.regex=false&columns%5B19%5D.data=country&columns%5B19%5D.name=&columns%5B19%5D.searchable=true&columns%5B19%5D.orderable=true&columns%5B19%5D.search.value=&columns%5B19%5D.search.regex=false&columns%5B20%5D.data=province&columns%5B20%5D.name=&columns%5B20%5D.searchable=true&columns%5B20%5D.orderable=true&columns%5B20%5D.search.value=&columns%5B20%5D.search.regex=false&columns%5B21%5D.data=lastModified&columns%5B21%5D.name=&columns%5B21%5D.searchable=true&columns%5B21%5D.orderable=true&columns%5B21%5D.search.value=&columns%5B21%5D.search.regex=false&order%5B0%5D.column=15&order%5B0%5D.dir=desc&start=0&length={max_entries}&search.value=&search.regex=false&'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# response = json.loads(requests.get(url).text)\n",
    "# data = response['data']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use predownloaded data file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "STRAIN_DATA = Path(NEO4J_IMPORT / 'cache/cncb_strain/cncbstrain.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(STRAIN_DATA) as f:\n",
    "    response = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = response['data']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://stackoverflow.com/questions/52795561/flattening-nested-json-in-pandas-data-frame\n",
    "def flatten_json(nested_json, exclude=['']):\n",
    "    \"\"\"Flatten json object with nested keys into a single level.\n",
    "        Args:\n",
    "            nested_json: A nested json object.\n",
    "            exclude: Keys to exclude from output.\n",
    "        Returns:\n",
    "            The flattened json object if successful, None otherwise.\n",
    "    \"\"\"\n",
    "    out = {}\n",
    "\n",
    "    def flatten(x, name='', exclude=exclude):\n",
    "        if type(x) is dict:\n",
    "            for a in x:\n",
    "                if a not in exclude: flatten(x[a], name + a + '_')\n",
    "        elif type(x) is list:\n",
    "            i = 0\n",
    "            for a in x:\n",
    "                flatten(a, name + str(i) + '_')\n",
    "                i += 1\n",
    "        else:\n",
    "            out[name[:-1]] = x\n",
    "\n",
    "    flatten(nested_json)\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame([flatten_json(x, exclude=['variationInfoList']) for x in data])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.fillna('', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 405128 entries, 0 to 405127\n",
      "Data columns (total 50 columns):\n",
      " #   Column                         Non-Null Count   Dtype \n",
      "---  ------                         --------------   ----- \n",
      " 0   id                             405128 non-null  int64 \n",
      " 1   name                           405128 non-null  object\n",
      " 2   accession                      405128 non-null  object\n",
      " 3   source                         405128 non-null  object\n",
      " 4   link                           405128 non-null  object\n",
      " 5   location                       405128 non-null  object\n",
      " 6   collectDate                    405128 non-null  object\n",
      " 7   dataProvider                   405128 non-null  object\n",
      " 8   submitDate                     405128 non-null  object\n",
      " 9   dataSubmitter                  405128 non-null  object\n",
      " 10  host                           405128 non-null  object\n",
      " 11  country                        405128 non-null  object\n",
      " 12  province                       405128 non-null  object\n",
      " 13  city                           405128 non-null  object\n",
      " 14  relatedAccession               405128 non-null  object\n",
      " 15  completeness                   405128 non-null  object\n",
      " 16  gender                         405128 non-null  object\n",
      " 17  age                            405128 non-null  object\n",
      " 18  lastModified                   405128 non-null  object\n",
      " 19  createDate                     405128 non-null  object\n",
      " 20  genomeQuality_id               405128 non-null  int64 \n",
      " 21  genomeQuality_completeness     405128 non-null  object\n",
      " 22  genomeQuality_virusLength      405128 non-null  int64 \n",
      " 23  genomeQuality_nratio           405128 non-null  object\n",
      " 24  genomeQuality_nbase            405128 non-null  int64 \n",
      " 25  genomeQuality_mutation         405128 non-null  int64 \n",
      " 26  genomeQuality_degenerateBase   405128 non-null  int64 \n",
      " 27  genomeQuality_gap              405128 non-null  int64 \n",
      " 28  genomeQuality_maxGapLength     405128 non-null  int64 \n",
      " 29  genomeQuality_highDensity      405128 non-null  object\n",
      " 30  genomeQuality_accession        405128 non-null  object\n",
      " 31  genomeQuality_virusName        405128 non-null  object\n",
      " 32  genomeQuality_createTime       405128 non-null  object\n",
      " 33  genomeQuality_lastModified     405128 non-null  object\n",
      " 34  genomeLineage_id               405128 non-null  object\n",
      " 35  genomeLineage_accession        405128 non-null  object\n",
      " 36  genomeLineage_lineage          405128 non-null  object\n",
      " 37  genomeLineage_lineagesVersion  405128 non-null  object\n",
      " 38  genomeLineage_status           405128 non-null  object\n",
      " 39  genomeLineage_note             405128 non-null  object\n",
      " 40  genomeLineage_createTime       405128 non-null  object\n",
      " 41  genomeLineage_lastModified     405128 non-null  object\n",
      " 42  genomeVariants_id              405128 non-null  object\n",
      " 43  genomeVariants_accession       405128 non-null  object\n",
      " 44  genomeVariants_variants        405128 non-null  object\n",
      " 45  genomeVariants_count           405128 non-null  object\n",
      " 46  genomeVariants_lastModified    405128 non-null  object\n",
      " 47  genomeVariants_createDate      405128 non-null  object\n",
      " 48  genomeVariants                 405128 non-null  object\n",
      " 49  genomeLineage                  405128 non-null  object\n",
      "dtypes: int64(8), object(42)\n",
      "memory usage: 154.5+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download strain metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df = pd.read_excel(metadata_url, dtype='str')\n",
    "#df.fillna('', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of strains: 405128\n"
     ]
    }
   ],
   "source": [
    "print(\"Total number of strains:\", df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>source</th>\n",
       "      <th>link</th>\n",
       "      <th>location</th>\n",
       "      <th>collectDate</th>\n",
       "      <th>dataProvider</th>\n",
       "      <th>submitDate</th>\n",
       "      <th>dataSubmitter</th>\n",
       "      <th>host</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>relatedAccession</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>lastModified</th>\n",
       "      <th>createDate</th>\n",
       "      <th>genomeQuality_id</th>\n",
       "      <th>genomeQuality_completeness</th>\n",
       "      <th>genomeQuality_virusLength</th>\n",
       "      <th>genomeQuality_nratio</th>\n",
       "      <th>genomeQuality_nbase</th>\n",
       "      <th>genomeQuality_mutation</th>\n",
       "      <th>genomeQuality_degenerateBase</th>\n",
       "      <th>genomeQuality_gap</th>\n",
       "      <th>genomeQuality_maxGapLength</th>\n",
       "      <th>genomeQuality_highDensity</th>\n",
       "      <th>genomeQuality_accession</th>\n",
       "      <th>genomeQuality_virusName</th>\n",
       "      <th>genomeQuality_createTime</th>\n",
       "      <th>genomeQuality_lastModified</th>\n",
       "      <th>genomeLineage_id</th>\n",
       "      <th>genomeLineage_accession</th>\n",
       "      <th>genomeLineage_lineage</th>\n",
       "      <th>genomeLineage_lineagesVersion</th>\n",
       "      <th>genomeLineage_status</th>\n",
       "      <th>genomeLineage_note</th>\n",
       "      <th>genomeLineage_createTime</th>\n",
       "      <th>genomeLineage_lastModified</th>\n",
       "      <th>genomeVariants_id</th>\n",
       "      <th>genomeVariants_accession</th>\n",
       "      <th>genomeVariants_variants</th>\n",
       "      <th>genomeVariants_count</th>\n",
       "      <th>genomeVariants_lastModified</th>\n",
       "      <th>genomeVariants_createDate</th>\n",
       "      <th>genomeVariants</th>\n",
       "      <th>genomeLineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43584</td>\n",
       "      <td>hCoV-19/Brazil/AP162966-IEC/2020</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-05</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>37</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41660</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>hCoV-19/Brazil/AP162966-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>109137.0</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>B.1.1.314</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td>2021-01-15 14:12:12</td>\n",
       "      <td>12993.0</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>241,29148,6433,3037,28883,9286,14408,28882,234...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100674</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-164827/2020</td>\n",
       "      <td>EPI_ISL_524788</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>43</td>\n",
       "      <td>Brazil / Maranhao</td>\n",
       "      <td>2020-04-24</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Maranhao</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>43</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100019</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524788</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-164827/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>194626.0</td>\n",
       "      <td>EPI_ISL_524788</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2021-01-15 14:12:19</td>\n",
       "      <td>32683.0</td>\n",
       "      <td>EPI_ISL_524788</td>\n",
       "      <td>28883,14408,27750,25338,23403,28881,3037,28882...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100682</td>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-164920/2020</td>\n",
       "      <td>EPI_ISL_524796</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>33</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-23</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>33</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100027</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524796</td>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-164920/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>41990.0</td>\n",
       "      <td>EPI_ISL_524796</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td>2021-01-15 14:12:07</td>\n",
       "      <td>32691.0</td>\n",
       "      <td>EPI_ISL_524796</td>\n",
       "      <td>28883,14408,23403,28881,3037,28882,241,29148,2...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>43585</td>\n",
       "      <td>hCoV-19/Brazil/AP164082-IEC/2020</td>\n",
       "      <td>EPI_ISL_458143</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-15</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>44</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41661</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458143</td>\n",
       "      <td>hCoV-19/Brazil/AP164082-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>77380.0</td>\n",
       "      <td>EPI_ISL_458143</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td>2021-01-15 14:12:19</td>\n",
       "      <td>12994.0</td>\n",
       "      <td>EPI_ISL_458143</td>\n",
       "      <td>28882,23403,28881,27299,241,29148,3037,28883,1...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100675</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-166716/2020</td>\n",
       "      <td>EPI_ISL_524789</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>32</td>\n",
       "      <td>Brazil / Maranhao</td>\n",
       "      <td>2020-05-05</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Maranhao</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>32</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100020</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>27297~27299(3-2-0.67,SNP:27297; SNP:27299) 288...</td>\n",
       "      <td>EPI_ISL_524789</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-166716/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>194640.0</td>\n",
       "      <td>EPI_ISL_524789</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2021-01-15 14:12:19</td>\n",
       "      <td>32684.0</td>\n",
       "      <td>EPI_ISL_524789</td>\n",
       "      <td>241,29148,27299,28883,14408,22224,27297,23403,...</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>100683</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-162157/2020</td>\n",
       "      <td>EPI_ISL_524797</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>55</td>\n",
       "      <td>Brazil / Maranhao</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Maranhao</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>55</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100028</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524797</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-162157/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>194613.0</td>\n",
       "      <td>EPI_ISL_524797</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2021-01-15 14:12:19</td>\n",
       "      <td>32692.0</td>\n",
       "      <td>EPI_ISL_524797</td>\n",
       "      <td>241,29148,27299,28883,14408,23403,28881,3037,2...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>43586</td>\n",
       "      <td>hCoV-19/Brazil/AP163972-IEC/2020</td>\n",
       "      <td>EPI_ISL_458144</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-15</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>44</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41662</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458144</td>\n",
       "      <td>hCoV-19/Brazil/AP163972-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>77367.0</td>\n",
       "      <td>EPI_ISL_458144</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td>2021-01-15 14:12:19</td>\n",
       "      <td>12995.0</td>\n",
       "      <td>EPI_ISL_458144</td>\n",
       "      <td>14408,28882,23403,28881,27299,241,29148,3037,2...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>100676</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-165398/2020</td>\n",
       "      <td>EPI_ISL_524790</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>37</td>\n",
       "      <td>Brazil / Maranhao</td>\n",
       "      <td>2020-04-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Maranhao</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>37</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100021</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524790</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-165398/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>42003.0</td>\n",
       "      <td>EPI_ISL_524790</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td>2021-01-15 14:12:07</td>\n",
       "      <td>32685.0</td>\n",
       "      <td>EPI_ISL_524790</td>\n",
       "      <td>241,29148,27299,28883,11083,14408,21147,23403,...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>100684</td>\n",
       "      <td>hCoV-19/Brazil/RN-IEC-162277/2020</td>\n",
       "      <td>EPI_ISL_524798</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>49</td>\n",
       "      <td>Brazil / Rio Grande do Norte</td>\n",
       "      <td>2020-03-14</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Rio Grande do Norte</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>49</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100029</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524798</td>\n",
       "      <td>hCoV-19/Brazil/RN-IEC-162277/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>41918.0</td>\n",
       "      <td>EPI_ISL_524798</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td>2021-01-15 14:12:07</td>\n",
       "      <td>32693.0</td>\n",
       "      <td>EPI_ISL_524798</td>\n",
       "      <td>241,29148,27299,28883,14408,23403,28881,3037,2...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>43587</td>\n",
       "      <td>hCoV-19/Brazil/AP164346-IEC/2020</td>\n",
       "      <td>EPI_ISL_458145</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-20</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>62</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41663</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458145</td>\n",
       "      <td>hCoV-19/Brazil/AP164346-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>109110.0</td>\n",
       "      <td>EPI_ISL_458145</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td>2021-01-15 14:12:06</td>\n",
       "      <td>12996.0</td>\n",
       "      <td>EPI_ISL_458145</td>\n",
       "      <td>28883,14408,28882,27047,23403,28881,27299,2572...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                               name       accession  source  \\\n",
       "0   43584   hCoV-19/Brazil/AP162966-IEC/2020  EPI_ISL_458142  GISAID   \n",
       "1  100674  hCoV-19/Brazil/MA-IEC-164827/2020  EPI_ISL_524788  GISAID   \n",
       "2  100682  hCoV-19/Brazil/AP-IEC-164920/2020  EPI_ISL_524796  GISAID   \n",
       "3   43585   hCoV-19/Brazil/AP164082-IEC/2020  EPI_ISL_458143  GISAID   \n",
       "4  100675  hCoV-19/Brazil/MA-IEC-166716/2020  EPI_ISL_524789  GISAID   \n",
       "5  100683  hCoV-19/Brazil/MA-IEC-162157/2020  EPI_ISL_524797  GISAID   \n",
       "6   43586   hCoV-19/Brazil/AP163972-IEC/2020  EPI_ISL_458144  GISAID   \n",
       "7  100676  hCoV-19/Brazil/MA-IEC-165398/2020  EPI_ISL_524790  GISAID   \n",
       "8  100684  hCoV-19/Brazil/RN-IEC-162277/2020  EPI_ISL_524798  GISAID   \n",
       "9   43587   hCoV-19/Brazil/AP164346-IEC/2020  EPI_ISL_458145  GISAID   \n",
       "\n",
       "                  link                      location collectDate  \\\n",
       "0  https://gisaid.org/                Brazil / Amapa  2020-04-05   \n",
       "1                   43             Brazil / Maranhao  2020-04-24   \n",
       "2                   33                Brazil / Amapa  2020-04-23   \n",
       "3  https://gisaid.org/                Brazil / Amapa  2020-04-15   \n",
       "4                   32             Brazil / Maranhao  2020-05-05   \n",
       "5                   55             Brazil / Maranhao  2020-03-24   \n",
       "6  https://gisaid.org/                Brazil / Amapa  2020-04-15   \n",
       "7                   37             Brazil / Maranhao  2020-04-28   \n",
       "8                   49  Brazil / Rio Grande do Norte  2020-03-14   \n",
       "9  https://gisaid.org/                Brazil / Amapa  2020-04-20   \n",
       "\n",
       "                dataProvider  submitDate              dataSubmitter  \\\n",
       "0  ﻿Evandro Chagas Institute  2020-06-03  ﻿Evandro Chagas Institute   \n",
       "1  ﻿Evandro Chagas Institute  2020-08-28  ﻿Evandro Chagas Institute   \n",
       "2  ﻿Evandro Chagas Institute  2020-08-28  ﻿Evandro Chagas Institute   \n",
       "3  ﻿Evandro Chagas Institute  2020-06-03  ﻿Evandro Chagas Institute   \n",
       "4  ﻿Evandro Chagas Institute  2020-08-28  ﻿Evandro Chagas Institute   \n",
       "5  ﻿Evandro Chagas Institute  2020-08-28  ﻿Evandro Chagas Institute   \n",
       "6  ﻿Evandro Chagas Institute  2020-06-03  ﻿Evandro Chagas Institute   \n",
       "7  ﻿Evandro Chagas Institute  2020-08-28  ﻿Evandro Chagas Institute   \n",
       "8  ﻿Evandro Chagas Institute  2020-08-28  ﻿Evandro Chagas Institute   \n",
       "9  ﻿Evandro Chagas Institute  2020-06-03  ﻿Evandro Chagas Institute   \n",
       "\n",
       "           host country             province city relatedAccession  \\\n",
       "0  Homo Sapiens  Brazil                Amapa                         \n",
       "1  Homo Sapiens  Brazil             Maranhao                         \n",
       "2  Homo Sapiens  Brazil                Amapa                         \n",
       "3  Homo Sapiens  Brazil                Amapa                         \n",
       "4  Homo Sapiens  Brazil             Maranhao                         \n",
       "5  Homo Sapiens  Brazil             Maranhao                         \n",
       "6  Homo Sapiens  Brazil                Amapa                         \n",
       "7  Homo Sapiens  Brazil             Maranhao                         \n",
       "8  Homo Sapiens  Brazil  Rio Grande do Norte                         \n",
       "9  Homo Sapiens  Brazil                Amapa                         \n",
       "\n",
       "  completeness  gender age         lastModified           createDate  \\\n",
       "0     Complete  Female  37  2020-09-09 11:31:17  2020-06-04 12:19:36   \n",
       "1     Complete    Male  43  2020-09-09 11:31:17  2020-08-29 18:49:44   \n",
       "2     Complete  Female  33  2020-09-09 11:31:17  2020-08-29 18:49:44   \n",
       "3     Complete    Male  44  2020-09-09 11:31:17  2020-06-04 12:19:36   \n",
       "4     Complete    Male  32  2020-09-09 11:31:17  2020-08-29 18:49:44   \n",
       "5     Complete  Female  55  2020-09-09 11:31:17  2020-08-29 18:49:44   \n",
       "6     Complete    Male  44  2020-09-09 11:31:17  2020-06-04 12:19:36   \n",
       "7     Complete  Female  37  2020-09-09 11:31:17  2020-08-29 18:49:44   \n",
       "8     Complete  Female  49  2020-09-09 11:31:17  2020-08-29 18:49:44   \n",
       "9     Complete  Female  62  2020-09-09 11:31:17  2020-06-04 12:19:36   \n",
       "\n",
       "   genomeQuality_id genomeQuality_completeness  genomeQuality_virusLength  \\\n",
       "0             41660                   Complete                      29903   \n",
       "1            100019                   Complete                      29903   \n",
       "2            100027                   Complete                      29903   \n",
       "3             41661                   Complete                      29903   \n",
       "4            100020                   Complete                      29903   \n",
       "5            100028                   Complete                      29903   \n",
       "6             41662                   Complete                      29903   \n",
       "7            100021                   Complete                      29903   \n",
       "8            100029                   Complete                      29903   \n",
       "9             41663                   Complete                      29903   \n",
       "\n",
       "  genomeQuality_nratio  genomeQuality_nbase  genomeQuality_mutation  \\\n",
       "0                0.00%                    0                      11   \n",
       "1                0.00%                    0                      11   \n",
       "2                0.00%                    0                       9   \n",
       "3                0.00%                    0                       9   \n",
       "4                0.00%                    0                      12   \n",
       "5                0.00%                    0                       9   \n",
       "6                0.00%                    0                       9   \n",
       "7                0.00%                    0                      11   \n",
       "8                0.00%                    0                       9   \n",
       "9                0.00%                    0                      11   \n",
       "\n",
       "   genomeQuality_degenerateBase  genomeQuality_gap  \\\n",
       "0                             0                  0   \n",
       "1                             0                  0   \n",
       "2                             0                  0   \n",
       "3                             0                  0   \n",
       "4                             0                  0   \n",
       "5                             0                  0   \n",
       "6                             0                  0   \n",
       "7                             0                  0   \n",
       "8                             0                  0   \n",
       "9                             0                  0   \n",
       "\n",
       "   genomeQuality_maxGapLength  \\\n",
       "0                          -1   \n",
       "1                          -1   \n",
       "2                          -1   \n",
       "3                          -1   \n",
       "4                          -1   \n",
       "5                          -1   \n",
       "6                          -1   \n",
       "7                          -1   \n",
       "8                          -1   \n",
       "9                          -1   \n",
       "\n",
       "                           genomeQuality_highDensity genomeQuality_accession  \\\n",
       "0  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458142   \n",
       "1  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524788   \n",
       "2  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524796   \n",
       "3  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458143   \n",
       "4  27297~27299(3-2-0.67,SNP:27297; SNP:27299) 288...          EPI_ISL_524789   \n",
       "5  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524797   \n",
       "6  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458144   \n",
       "7  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524790   \n",
       "8  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524798   \n",
       "9  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458145   \n",
       "\n",
       "             genomeQuality_virusName genomeQuality_createTime  \\\n",
       "0   hCoV-19/Brazil/AP162966-IEC/2020      2020-06-04 12:20:14   \n",
       "1  hCoV-19/Brazil/MA-IEC-164827/2020      2020-08-29 18:52:11   \n",
       "2  hCoV-19/Brazil/AP-IEC-164920/2020      2020-08-29 18:52:11   \n",
       "3   hCoV-19/Brazil/AP164082-IEC/2020      2020-06-04 12:20:14   \n",
       "4  hCoV-19/Brazil/MA-IEC-166716/2020      2020-08-29 18:52:11   \n",
       "5  hCoV-19/Brazil/MA-IEC-162157/2020      2020-08-29 18:52:11   \n",
       "6   hCoV-19/Brazil/AP163972-IEC/2020      2020-06-04 12:20:14   \n",
       "7  hCoV-19/Brazil/MA-IEC-165398/2020      2020-08-29 18:52:11   \n",
       "8  hCoV-19/Brazil/RN-IEC-162277/2020      2020-08-29 18:52:11   \n",
       "9   hCoV-19/Brazil/AP164346-IEC/2020      2020-06-04 12:20:14   \n",
       "\n",
       "  genomeQuality_lastModified genomeLineage_id genomeLineage_accession  \\\n",
       "0        2020-06-04 12:20:14         109137.0          EPI_ISL_458142   \n",
       "1        2020-08-29 18:52:11         194626.0          EPI_ISL_524788   \n",
       "2        2020-08-29 18:52:11          41990.0          EPI_ISL_524796   \n",
       "3        2020-06-04 12:20:14          77380.0          EPI_ISL_458143   \n",
       "4        2020-08-29 18:52:11         194640.0          EPI_ISL_524789   \n",
       "5        2020-08-29 18:52:11         194613.0          EPI_ISL_524797   \n",
       "6        2020-06-04 12:20:14          77367.0          EPI_ISL_458144   \n",
       "7        2020-08-29 18:52:11          42003.0          EPI_ISL_524790   \n",
       "8        2020-08-29 18:52:11          41918.0          EPI_ISL_524798   \n",
       "9        2020-06-04 12:20:14         109110.0          EPI_ISL_458145   \n",
       "\n",
       "  genomeLineage_lineage genomeLineage_lineagesVersion genomeLineage_status  \\\n",
       "0             B.1.1.314                     2021/1/11            passed_qc   \n",
       "1              B.1.1.33                     2021/1/11            passed_qc   \n",
       "2              B.1.1.33                     2021/1/11            passed_qc   \n",
       "3              B.1.1.33                     2021/1/11            passed_qc   \n",
       "4              B.1.1.33                     2021/1/11            passed_qc   \n",
       "5              B.1.1.33                     2021/1/11            passed_qc   \n",
       "6              B.1.1.33                     2021/1/11            passed_qc   \n",
       "7              B.1.1.33                     2021/1/11            passed_qc   \n",
       "8              B.1.1.33                     2021/1/11            passed_qc   \n",
       "9              B.1.1.33                     2021/1/11            passed_qc   \n",
       "\n",
       "  genomeLineage_note genomeLineage_createTime genomeLineage_lastModified  \\\n",
       "0                         2020-11-30 15:44:38        2021-01-15 14:12:12   \n",
       "1                         2020-11-30 15:44:43        2021-01-15 14:12:19   \n",
       "2                         2020-11-30 15:44:32        2021-01-15 14:12:07   \n",
       "3                         2020-11-30 15:44:35        2021-01-15 14:12:19   \n",
       "4                         2020-11-30 15:44:43        2021-01-15 14:12:19   \n",
       "5                         2020-11-30 15:44:43        2021-01-15 14:12:19   \n",
       "6                         2020-11-30 15:44:35        2021-01-15 14:12:19   \n",
       "7                         2020-11-30 15:44:32        2021-01-15 14:12:07   \n",
       "8                         2020-11-30 15:44:32        2021-01-15 14:12:07   \n",
       "9                         2020-11-30 15:44:38        2021-01-15 14:12:06   \n",
       "\n",
       "  genomeVariants_id genomeVariants_accession  \\\n",
       "0           12993.0           EPI_ISL_458142   \n",
       "1           32683.0           EPI_ISL_524788   \n",
       "2           32691.0           EPI_ISL_524796   \n",
       "3           12994.0           EPI_ISL_458143   \n",
       "4           32684.0           EPI_ISL_524789   \n",
       "5           32692.0           EPI_ISL_524797   \n",
       "6           12995.0           EPI_ISL_458144   \n",
       "7           32685.0           EPI_ISL_524790   \n",
       "8           32693.0           EPI_ISL_524798   \n",
       "9           12996.0           EPI_ISL_458145   \n",
       "\n",
       "                             genomeVariants_variants genomeVariants_count  \\\n",
       "0  241,29148,6433,3037,28883,9286,14408,28882,234...                 11.0   \n",
       "1  28883,14408,27750,25338,23403,28881,3037,28882...                 11.0   \n",
       "2  28883,14408,23403,28881,3037,28882,241,29148,2...                  9.0   \n",
       "3  28882,23403,28881,27299,241,29148,3037,28883,1...                  9.0   \n",
       "4  241,29148,27299,28883,14408,22224,27297,23403,...                 12.0   \n",
       "5  241,29148,27299,28883,14408,23403,28881,3037,2...                  9.0   \n",
       "6  14408,28882,23403,28881,27299,241,29148,3037,2...                  9.0   \n",
       "7  241,29148,27299,28883,11083,14408,21147,23403,...                 11.0   \n",
       "8  241,29148,27299,28883,14408,23403,28881,3037,2...                  9.0   \n",
       "9  28883,14408,28882,27047,23403,28881,27299,2572...                 11.0   \n",
       "\n",
       "  genomeVariants_lastModified genomeVariants_createDate genomeVariants  \\\n",
       "0         2021-01-20 19:03:19       2021-01-20 19:03:19                  \n",
       "1         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "2         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "3         2021-01-20 19:03:19       2021-01-20 19:03:19                  \n",
       "4         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "5         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "6         2021-01-20 19:03:19       2021-01-20 19:03:19                  \n",
       "7         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "8         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "9         2021-01-20 19:03:19       2021-01-20 19:03:19                  \n",
       "\n",
       "  genomeLineage  \n",
       "0                \n",
       "1                \n",
       "2                \n",
       "3                \n",
       "4                \n",
       "5                \n",
       "6                \n",
       "7                \n",
       "8                \n",
       "9                "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Asssign unique identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://registry.identifiers.org/registry/insdc\n",
    "insdc_pattern = re.compile('^([A-Z]\\d{5}|[A-Z]{2}\\d{6}|[A-Z]{4}\\d{8}|[A-J][A-Z]{2}\\d{5})(\\.\\d+)?$')\n",
    "# https://registry.identifiers.org/registry/refseq\n",
    "refseq_pattern = re.compile('^(((AC|AP|NC|NG|NM|NP|NR|NT|NW|XM|XP|XR|YP|ZP)_\\d+)|(NZ\\_[A-Z]{2,4}\\d+))(\\.\\d+)?$')\n",
    "epi_pattern = re.compile('^EPI_ISL_\\d+$')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_curie(id):\n",
    "    id = id.strip()\n",
    "    if len(id) > 0:\n",
    "        if id.startswith('EPI'):\n",
    "            return 'https://www.gisaid.org/' + id\n",
    "        elif refseq_pattern.match(id) != None:\n",
    "            return 'refseq:' + id\n",
    "        elif insdc_pattern.match(id) != None:\n",
    "            return 'insdc:' + id\n",
    "        else:\n",
    "            # TODO are URIs available for these cases?\n",
    "            return id\n",
    "    else:\n",
    "        return id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_curies(ids):\n",
    "    return [assign_curie(id) for id in ids.split(',')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gisaid_id(ids):\n",
    "    for id in ids:\n",
    "        if id.startswith('https://www.gisaid.org/'):\n",
    "            return id\n",
    "        \n",
    "    return ''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rename and concatenate fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df['Accession ID'] = df['Accession ID'].str.strip()\n",
    "#df['Related ID'] = df['Related ID'].str.strip()\n",
    "\n",
    "# combine all ids into an accession column and assign curies\n",
    "df['accessions'] = df['accession'] + df['relatedAccession'].apply(lambda s: ',' + s if len(s) > 0 else s)\n",
    "df['accessions'] = df['accessions'].apply(assign_curies)\n",
    "df['gisaidId'] = df['accessions'].apply(get_gisaid_id)\n",
    "df['accessions'] = df['accessions'].apply(lambda x: ';'.join(x))\n",
    "\n",
    "df['accession'] = df['accession'].apply(lambda s: assign_curie(s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.rename(columns={'genomeQuality_virusLength': 'sequenceLength'}, inplace=True)\n",
    "#df.rename(columns={'Sequence Quality': 'sequenceQuality'}, inplace=True) # get from original spreadsheet\n",
    "#df.rename(columns={'Quality Assessment': 'qualityAssessment'}, inplace=True) # get from original spreadsheet\n",
    "df.rename(columns={'dataProvider': 'originatingLab'}, inplace=True)\n",
    "#df.rename(columns={'Virus Strain Name': 'name'}, inplace=True)\n",
    "df.rename(columns={'collectDate':'collectionDate'},inplace=True)\n",
    "df.rename(columns={'Location':'location'}, inplace=True)\n",
    "df.rename(columns={'genomeLineage_lineage': 'lineage'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove invalid collection date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>source</th>\n",
       "      <th>link</th>\n",
       "      <th>location</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>originatingLab</th>\n",
       "      <th>submitDate</th>\n",
       "      <th>dataSubmitter</th>\n",
       "      <th>host</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>relatedAccession</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>lastModified</th>\n",
       "      <th>createDate</th>\n",
       "      <th>genomeQuality_id</th>\n",
       "      <th>genomeQuality_completeness</th>\n",
       "      <th>sequenceLength</th>\n",
       "      <th>genomeQuality_nratio</th>\n",
       "      <th>genomeQuality_nbase</th>\n",
       "      <th>genomeQuality_mutation</th>\n",
       "      <th>genomeQuality_degenerateBase</th>\n",
       "      <th>genomeQuality_gap</th>\n",
       "      <th>genomeQuality_maxGapLength</th>\n",
       "      <th>genomeQuality_highDensity</th>\n",
       "      <th>genomeQuality_accession</th>\n",
       "      <th>genomeQuality_virusName</th>\n",
       "      <th>genomeQuality_createTime</th>\n",
       "      <th>genomeQuality_lastModified</th>\n",
       "      <th>genomeLineage_id</th>\n",
       "      <th>genomeLineage_accession</th>\n",
       "      <th>lineage</th>\n",
       "      <th>genomeLineage_lineagesVersion</th>\n",
       "      <th>genomeLineage_status</th>\n",
       "      <th>genomeLineage_note</th>\n",
       "      <th>genomeLineage_createTime</th>\n",
       "      <th>genomeLineage_lastModified</th>\n",
       "      <th>genomeVariants_id</th>\n",
       "      <th>genomeVariants_accession</th>\n",
       "      <th>genomeVariants_variants</th>\n",
       "      <th>genomeVariants_count</th>\n",
       "      <th>genomeVariants_lastModified</th>\n",
       "      <th>genomeVariants_createDate</th>\n",
       "      <th>genomeVariants</th>\n",
       "      <th>genomeLineage</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gisaidId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [id, name, accession, source, link, location, collectionDate, originatingLab, submitDate, dataSubmitter, host, country, province, city, relatedAccession, completeness, gender, age, lastModified, createDate, genomeQuality_id, genomeQuality_completeness, sequenceLength, genomeQuality_nratio, genomeQuality_nbase, genomeQuality_mutation, genomeQuality_degenerateBase, genomeQuality_gap, genomeQuality_maxGapLength, genomeQuality_highDensity, genomeQuality_accession, genomeQuality_virusName, genomeQuality_createTime, genomeQuality_lastModified, genomeLineage_id, genomeLineage_accession, lineage, genomeLineage_lineagesVersion, genomeLineage_status, genomeLineage_note, genomeLineage_createTime, genomeLineage_lastModified, genomeVariants_id, genomeVariants_accession, genomeVariants_variants, genomeVariants_count, genomeVariants_lastModified, genomeVariants_createDate, genomeVariants, genomeLineage, accessions, gisaidId]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query(\"collectionDate == '2020-00-00'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['collectionDate'] = df['collectionDate'].apply(lambda d: '' if d == '2020-00-00' else d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['collectionDate'] = df['collectionDate'].apply(lambda d: dateutil.parser.parse(d) if len(d) > 0 else '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>source</th>\n",
       "      <th>link</th>\n",
       "      <th>location</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>originatingLab</th>\n",
       "      <th>submitDate</th>\n",
       "      <th>dataSubmitter</th>\n",
       "      <th>host</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>relatedAccession</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>lastModified</th>\n",
       "      <th>createDate</th>\n",
       "      <th>genomeQuality_id</th>\n",
       "      <th>genomeQuality_completeness</th>\n",
       "      <th>sequenceLength</th>\n",
       "      <th>genomeQuality_nratio</th>\n",
       "      <th>genomeQuality_nbase</th>\n",
       "      <th>genomeQuality_mutation</th>\n",
       "      <th>genomeQuality_degenerateBase</th>\n",
       "      <th>genomeQuality_gap</th>\n",
       "      <th>genomeQuality_maxGapLength</th>\n",
       "      <th>genomeQuality_highDensity</th>\n",
       "      <th>genomeQuality_accession</th>\n",
       "      <th>genomeQuality_virusName</th>\n",
       "      <th>genomeQuality_createTime</th>\n",
       "      <th>genomeQuality_lastModified</th>\n",
       "      <th>genomeLineage_id</th>\n",
       "      <th>genomeLineage_accession</th>\n",
       "      <th>lineage</th>\n",
       "      <th>genomeLineage_lineagesVersion</th>\n",
       "      <th>genomeLineage_status</th>\n",
       "      <th>genomeLineage_note</th>\n",
       "      <th>genomeLineage_createTime</th>\n",
       "      <th>genomeLineage_lastModified</th>\n",
       "      <th>genomeVariants_id</th>\n",
       "      <th>genomeVariants_accession</th>\n",
       "      <th>genomeVariants_variants</th>\n",
       "      <th>genomeVariants_count</th>\n",
       "      <th>genomeVariants_lastModified</th>\n",
       "      <th>genomeVariants_createDate</th>\n",
       "      <th>genomeVariants</th>\n",
       "      <th>genomeLineage</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gisaidId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>71298</th>\n",
       "      <td>14</td>\n",
       "      <td>Wuhan-Hu-1</td>\n",
       "      <td>insdc:MN908947</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MN908947</td>\n",
       "      <td>China / Hubei / Wuhan</td>\n",
       "      <td>2019-12-31</td>\n",
       "      <td>Shanghai Public Health Clinical Center &amp; Schoo...</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>Shanghai Public Health Clinical Center &amp; Schoo...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>NC_045512,EPI_ISL_402125</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>41</td>\n",
       "      <td>2020-05-20 11:14:12</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>629</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>MN908947</td>\n",
       "      <td>Wuhan-Hu-1</td>\n",
       "      <td>2020-03-17 14:27:13</td>\n",
       "      <td>2020-03-17 14:27:13</td>\n",
       "      <td>195185.0</td>\n",
       "      <td>MN908947</td>\n",
       "      <td>B</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>insdc:MN908947;refseq:NC_045512;https://www.gi...</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_402125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id        name       accession   source  \\\n",
       "71298  14  Wuhan-Hu-1  insdc:MN908947  GenBank   \n",
       "\n",
       "                                                link               location  \\\n",
       "71298  https://www.ncbi.nlm.nih.gov/nuccore/MN908947  China / Hubei / Wuhan   \n",
       "\n",
       "      collectionDate                                     originatingLab  \\\n",
       "71298     2019-12-31  Shanghai Public Health Clinical Center & Schoo...   \n",
       "\n",
       "       submitDate                                      dataSubmitter  \\\n",
       "71298  2020-01-17  Shanghai Public Health Clinical Center & Schoo...   \n",
       "\n",
       "               host country province   city          relatedAccession  \\\n",
       "71298  Homo sapiens   China    Hubei  Wuhan  NC_045512,EPI_ISL_402125   \n",
       "\n",
       "      completeness gender age         lastModified           createDate  \\\n",
       "71298     Complete   Male  41  2020-05-20 11:14:12  2020-01-20 20:04:48   \n",
       "\n",
       "       genomeQuality_id genomeQuality_completeness  sequenceLength  \\\n",
       "71298               629                   Complete           29903   \n",
       "\n",
       "      genomeQuality_nratio  genomeQuality_nbase  genomeQuality_mutation  \\\n",
       "71298                    0                    0                       0   \n",
       "\n",
       "       genomeQuality_degenerateBase  genomeQuality_gap  \\\n",
       "71298                             0                  0   \n",
       "\n",
       "       genomeQuality_maxGapLength genomeQuality_highDensity  \\\n",
       "71298                          -1                        NO   \n",
       "\n",
       "      genomeQuality_accession genomeQuality_virusName  \\\n",
       "71298                MN908947              Wuhan-Hu-1   \n",
       "\n",
       "      genomeQuality_createTime genomeQuality_lastModified genomeLineage_id  \\\n",
       "71298      2020-03-17 14:27:13        2020-03-17 14:27:13         195185.0   \n",
       "\n",
       "      genomeLineage_accession lineage genomeLineage_lineagesVersion  \\\n",
       "71298                MN908947       B                    2020/10/30   \n",
       "\n",
       "      genomeLineage_status genomeLineage_note genomeLineage_createTime  \\\n",
       "71298            passed_qc                         2020-11-30 15:44:43   \n",
       "\n",
       "      genomeLineage_lastModified genomeVariants_id genomeVariants_accession  \\\n",
       "71298        2020-11-30 15:44:43                                              \n",
       "\n",
       "      genomeVariants_variants genomeVariants_count  \\\n",
       "71298                                                \n",
       "\n",
       "      genomeVariants_lastModified genomeVariants_createDate genomeVariants  \\\n",
       "71298                                                                        \n",
       "\n",
       "      genomeLineage                                         accessions  \\\n",
       "71298                insdc:MN908947;refseq:NC_045512;https://www.gi...   \n",
       "\n",
       "                                    gisaidId  \n",
       "71298  https://www.gisaid.org/EPI_ISL_402125  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['accessions'].str.contains('refseq:NC_045512')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>source</th>\n",
       "      <th>link</th>\n",
       "      <th>location</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>originatingLab</th>\n",
       "      <th>submitDate</th>\n",
       "      <th>dataSubmitter</th>\n",
       "      <th>host</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>relatedAccession</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>lastModified</th>\n",
       "      <th>createDate</th>\n",
       "      <th>genomeQuality_id</th>\n",
       "      <th>genomeQuality_completeness</th>\n",
       "      <th>sequenceLength</th>\n",
       "      <th>genomeQuality_nratio</th>\n",
       "      <th>genomeQuality_nbase</th>\n",
       "      <th>genomeQuality_mutation</th>\n",
       "      <th>genomeQuality_degenerateBase</th>\n",
       "      <th>genomeQuality_gap</th>\n",
       "      <th>genomeQuality_maxGapLength</th>\n",
       "      <th>genomeQuality_highDensity</th>\n",
       "      <th>genomeQuality_accession</th>\n",
       "      <th>genomeQuality_virusName</th>\n",
       "      <th>genomeQuality_createTime</th>\n",
       "      <th>genomeQuality_lastModified</th>\n",
       "      <th>genomeLineage_id</th>\n",
       "      <th>genomeLineage_accession</th>\n",
       "      <th>lineage</th>\n",
       "      <th>genomeLineage_lineagesVersion</th>\n",
       "      <th>genomeLineage_status</th>\n",
       "      <th>genomeLineage_note</th>\n",
       "      <th>genomeLineage_createTime</th>\n",
       "      <th>genomeLineage_lastModified</th>\n",
       "      <th>genomeVariants_id</th>\n",
       "      <th>genomeVariants_accession</th>\n",
       "      <th>genomeVariants_variants</th>\n",
       "      <th>genomeVariants_count</th>\n",
       "      <th>genomeVariants_lastModified</th>\n",
       "      <th>genomeVariants_createDate</th>\n",
       "      <th>genomeVariants</th>\n",
       "      <th>genomeLineage</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gisaidId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43584</td>\n",
       "      <td>hCoV-19/Brazil/AP162966-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458142</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-05</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>37</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41660</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>hCoV-19/Brazil/AP162966-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>109137.0</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>B.1.1.314</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td>2021-01-15 14:12:12</td>\n",
       "      <td>12993.0</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>241,29148,6433,3037,28883,9286,14408,28882,234...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458142;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100674</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-164827/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524788</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>43</td>\n",
       "      <td>Brazil / Maranhao</td>\n",
       "      <td>2020-04-24</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Maranhao</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>43</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100019</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524788</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-164827/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>194626.0</td>\n",
       "      <td>EPI_ISL_524788</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2021-01-15 14:12:19</td>\n",
       "      <td>32683.0</td>\n",
       "      <td>EPI_ISL_524788</td>\n",
       "      <td>28883,14408,27750,25338,23403,28881,3037,28882...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524788;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100682</td>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-164920/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524796</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>33</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-23</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>33</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100027</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524796</td>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-164920/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>41990.0</td>\n",
       "      <td>EPI_ISL_524796</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td>2021-01-15 14:12:07</td>\n",
       "      <td>32691.0</td>\n",
       "      <td>EPI_ISL_524796</td>\n",
       "      <td>28883,14408,23403,28881,3037,28882,241,29148,2...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524796;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>43585</td>\n",
       "      <td>hCoV-19/Brazil/AP164082-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458143</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-15</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>44</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41661</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458143</td>\n",
       "      <td>hCoV-19/Brazil/AP164082-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>77380.0</td>\n",
       "      <td>EPI_ISL_458143</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td>2021-01-15 14:12:19</td>\n",
       "      <td>12994.0</td>\n",
       "      <td>EPI_ISL_458143</td>\n",
       "      <td>28882,23403,28881,27299,241,29148,3037,28883,1...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td>2021-01-20 19:03:19</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458143;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100675</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-166716/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524789</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>32</td>\n",
       "      <td>Brazil / Maranhao</td>\n",
       "      <td>2020-05-05</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Maranhao</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>32</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100020</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>27297~27299(3-2-0.67,SNP:27297; SNP:27299) 288...</td>\n",
       "      <td>EPI_ISL_524789</td>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-166716/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>194640.0</td>\n",
       "      <td>EPI_ISL_524789</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2021/1/11</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2021-01-15 14:12:19</td>\n",
       "      <td>32684.0</td>\n",
       "      <td>EPI_ISL_524789</td>\n",
       "      <td>241,29148,27299,28883,14408,22224,27297,23403,...</td>\n",
       "      <td>12.0</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td>2021-01-20 19:03:20</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524789;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524789</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                               name  \\\n",
       "0   43584   hCoV-19/Brazil/AP162966-IEC/2020   \n",
       "1  100674  hCoV-19/Brazil/MA-IEC-164827/2020   \n",
       "2  100682  hCoV-19/Brazil/AP-IEC-164920/2020   \n",
       "3   43585   hCoV-19/Brazil/AP164082-IEC/2020   \n",
       "4  100675  hCoV-19/Brazil/MA-IEC-166716/2020   \n",
       "\n",
       "                               accession  source                 link  \\\n",
       "0  https://www.gisaid.org/EPI_ISL_458142  GISAID  https://gisaid.org/   \n",
       "1  https://www.gisaid.org/EPI_ISL_524788  GISAID                   43   \n",
       "2  https://www.gisaid.org/EPI_ISL_524796  GISAID                   33   \n",
       "3  https://www.gisaid.org/EPI_ISL_458143  GISAID  https://gisaid.org/   \n",
       "4  https://www.gisaid.org/EPI_ISL_524789  GISAID                   32   \n",
       "\n",
       "            location collectionDate             originatingLab  submitDate  \\\n",
       "0     Brazil / Amapa     2020-04-05  ﻿Evandro Chagas Institute  2020-06-03   \n",
       "1  Brazil / Maranhao     2020-04-24  ﻿Evandro Chagas Institute  2020-08-28   \n",
       "2     Brazil / Amapa     2020-04-23  ﻿Evandro Chagas Institute  2020-08-28   \n",
       "3     Brazil / Amapa     2020-04-15  ﻿Evandro Chagas Institute  2020-06-03   \n",
       "4  Brazil / Maranhao     2020-05-05  ﻿Evandro Chagas Institute  2020-08-28   \n",
       "\n",
       "               dataSubmitter          host country  province city  \\\n",
       "0  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Amapa        \n",
       "1  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil  Maranhao        \n",
       "2  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Amapa        \n",
       "3  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Amapa        \n",
       "4  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil  Maranhao        \n",
       "\n",
       "  relatedAccession completeness  gender age         lastModified  \\\n",
       "0                      Complete  Female  37  2020-09-09 11:31:17   \n",
       "1                      Complete    Male  43  2020-09-09 11:31:17   \n",
       "2                      Complete  Female  33  2020-09-09 11:31:17   \n",
       "3                      Complete    Male  44  2020-09-09 11:31:17   \n",
       "4                      Complete    Male  32  2020-09-09 11:31:17   \n",
       "\n",
       "            createDate  genomeQuality_id genomeQuality_completeness  \\\n",
       "0  2020-06-04 12:19:36             41660                   Complete   \n",
       "1  2020-08-29 18:49:44            100019                   Complete   \n",
       "2  2020-08-29 18:49:44            100027                   Complete   \n",
       "3  2020-06-04 12:19:36             41661                   Complete   \n",
       "4  2020-08-29 18:49:44            100020                   Complete   \n",
       "\n",
       "   sequenceLength genomeQuality_nratio  genomeQuality_nbase  \\\n",
       "0           29903                0.00%                    0   \n",
       "1           29903                0.00%                    0   \n",
       "2           29903                0.00%                    0   \n",
       "3           29903                0.00%                    0   \n",
       "4           29903                0.00%                    0   \n",
       "\n",
       "   genomeQuality_mutation  genomeQuality_degenerateBase  genomeQuality_gap  \\\n",
       "0                      11                             0                  0   \n",
       "1                      11                             0                  0   \n",
       "2                       9                             0                  0   \n",
       "3                       9                             0                  0   \n",
       "4                      12                             0                  0   \n",
       "\n",
       "   genomeQuality_maxGapLength  \\\n",
       "0                          -1   \n",
       "1                          -1   \n",
       "2                          -1   \n",
       "3                          -1   \n",
       "4                          -1   \n",
       "\n",
       "                           genomeQuality_highDensity genomeQuality_accession  \\\n",
       "0  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458142   \n",
       "1  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524788   \n",
       "2  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524796   \n",
       "3  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458143   \n",
       "4  27297~27299(3-2-0.67,SNP:27297; SNP:27299) 288...          EPI_ISL_524789   \n",
       "\n",
       "             genomeQuality_virusName genomeQuality_createTime  \\\n",
       "0   hCoV-19/Brazil/AP162966-IEC/2020      2020-06-04 12:20:14   \n",
       "1  hCoV-19/Brazil/MA-IEC-164827/2020      2020-08-29 18:52:11   \n",
       "2  hCoV-19/Brazil/AP-IEC-164920/2020      2020-08-29 18:52:11   \n",
       "3   hCoV-19/Brazil/AP164082-IEC/2020      2020-06-04 12:20:14   \n",
       "4  hCoV-19/Brazil/MA-IEC-166716/2020      2020-08-29 18:52:11   \n",
       "\n",
       "  genomeQuality_lastModified genomeLineage_id genomeLineage_accession  \\\n",
       "0        2020-06-04 12:20:14         109137.0          EPI_ISL_458142   \n",
       "1        2020-08-29 18:52:11         194626.0          EPI_ISL_524788   \n",
       "2        2020-08-29 18:52:11          41990.0          EPI_ISL_524796   \n",
       "3        2020-06-04 12:20:14          77380.0          EPI_ISL_458143   \n",
       "4        2020-08-29 18:52:11         194640.0          EPI_ISL_524789   \n",
       "\n",
       "     lineage genomeLineage_lineagesVersion genomeLineage_status  \\\n",
       "0  B.1.1.314                     2021/1/11            passed_qc   \n",
       "1   B.1.1.33                     2021/1/11            passed_qc   \n",
       "2   B.1.1.33                     2021/1/11            passed_qc   \n",
       "3   B.1.1.33                     2021/1/11            passed_qc   \n",
       "4   B.1.1.33                     2021/1/11            passed_qc   \n",
       "\n",
       "  genomeLineage_note genomeLineage_createTime genomeLineage_lastModified  \\\n",
       "0                         2020-11-30 15:44:38        2021-01-15 14:12:12   \n",
       "1                         2020-11-30 15:44:43        2021-01-15 14:12:19   \n",
       "2                         2020-11-30 15:44:32        2021-01-15 14:12:07   \n",
       "3                         2020-11-30 15:44:35        2021-01-15 14:12:19   \n",
       "4                         2020-11-30 15:44:43        2021-01-15 14:12:19   \n",
       "\n",
       "  genomeVariants_id genomeVariants_accession  \\\n",
       "0           12993.0           EPI_ISL_458142   \n",
       "1           32683.0           EPI_ISL_524788   \n",
       "2           32691.0           EPI_ISL_524796   \n",
       "3           12994.0           EPI_ISL_458143   \n",
       "4           32684.0           EPI_ISL_524789   \n",
       "\n",
       "                             genomeVariants_variants genomeVariants_count  \\\n",
       "0  241,29148,6433,3037,28883,9286,14408,28882,234...                 11.0   \n",
       "1  28883,14408,27750,25338,23403,28881,3037,28882...                 11.0   \n",
       "2  28883,14408,23403,28881,3037,28882,241,29148,2...                  9.0   \n",
       "3  28882,23403,28881,27299,241,29148,3037,28883,1...                  9.0   \n",
       "4  241,29148,27299,28883,14408,22224,27297,23403,...                 12.0   \n",
       "\n",
       "  genomeVariants_lastModified genomeVariants_createDate genomeVariants  \\\n",
       "0         2021-01-20 19:03:19       2021-01-20 19:03:19                  \n",
       "1         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "2         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "3         2021-01-20 19:03:19       2021-01-20 19:03:19                  \n",
       "4         2021-01-20 19:03:20       2021-01-20 19:03:20                  \n",
       "\n",
       "  genomeLineage                              accessions  \\\n",
       "0                https://www.gisaid.org/EPI_ISL_458142;   \n",
       "1                https://www.gisaid.org/EPI_ISL_524788;   \n",
       "2                https://www.gisaid.org/EPI_ISL_524796;   \n",
       "3                https://www.gisaid.org/EPI_ISL_458143;   \n",
       "4                https://www.gisaid.org/EPI_ISL_524789;   \n",
       "\n",
       "                                gisaidId  \n",
       "0  https://www.gisaid.org/EPI_ISL_458142  \n",
       "1  https://www.gisaid.org/EPI_ISL_524788  \n",
       "2  https://www.gisaid.org/EPI_ISL_524796  \n",
       "3  https://www.gisaid.org/EPI_ISL_458143  \n",
       "4  https://www.gisaid.org/EPI_ISL_524789  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assign taxonomy ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read Organism reference dictionary\n",
    "organism_to_id = dict()\n",
    "data = pd.read_csv(\"../../reference_data/OrganismDictionary.csv\", comment='#')\n",
    "for index, row in data.iterrows():\n",
    "    organism_to_id[row['organism']] = row['taxonomyId']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'human': 'taxonomy:9606', 'homo sapiens': 'taxonomy:9606', 'chlorocebus sabaeus': 'taxonomy:60711', 'mus musculus': 'taxonomy:10090', 'rhinolophus affinis': 'taxonomy:59477 ', 'rhinolophus malayanus': 'taxonomy:608659', 'rhinolophus shameli': 'taxonomy:608708', 'mustela lutreola': 'taxonomy:9666', 'mink': 'taxonomy:9666', 'panthera tigris jacksoni': 'taxonomy:419130', 'rhinolophus sp. (bat)': 'taxonomy:49442', 'bat': 'taxonomy:49442', 'manis javanica': 'taxonomy:9974', 'manis pentadactyla': 'taxonomy:143292', 'palm civet': 'taxonomy:71116', 'canine': 'taxonomy:9608', 'canis lupus familiaris': 'taxonomy:9615', 'felis catus': 'taxonomy:9685', 'neovison vison': 'taxonomy:452646', 'mesocricetus auratus': 'taxonomy:10036', 'panthera leo': 'taxonomy:9689', 'panthera tigris': 'taxonomy:9694', 'environment': 'taxonomy:151659', 'environmental': 'taxonomy:151659'}\n"
     ]
    }
   ],
   "source": [
    "print(organism_to_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['taxonomy:9606', 'taxonomy:10090', 'taxonomy:151659',\n",
       "       'taxonomy:9685', 'taxonomy:9666', 'taxonomy:608708',\n",
       "       'taxonomy:9615', 'taxonomy:143292', 'taxonomy:9974', 'unknown',\n",
       "       'taxonomy:608659', 'taxonomy:452646', 'taxonomy:419130',\n",
       "       'taxonomy:60711', 'taxonomy:59477 ', 'Vero cell culture',\n",
       "       'taxonomy:9694', 'taxonomy:9689', 'taxonomy:10036'], dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign taxonomy id to host\n",
    "df['host'] = df['host'].str.strip()\n",
    "df['hostTaxonomyId'] = df['host'].apply(lambda s: organism_to_id.get(s.lower(), s))\n",
    "df['hostTaxonomyId'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['taxonomyId'] = 'taxonomy:2697049' # SARS-CoV-2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Standardize location information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['loc0', 'loc1', 'loc2', 'loc3']] = df['location'].str.split('/', n=3, expand=True)\n",
    "# strip white space\n",
    "df = df.apply(lambda x: x.str.strip() if x.dtype == \"object\" else x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['origLocation'] = df[['loc0', 'loc1', 'loc2', 'loc3']].apply(\n",
    "    lambda x: ','.join(x.dropna().astype(str)),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save strain metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# strains = df[['name', 'accession', 'accessions', 'gisaidId', 'source', 'taxonomyId', 'hostTaxonomyId', \n",
    "#               'sequenceLength', 'sequenceQuality', 'qualityAssessment', 'collectionDate', 'location', \n",
    "#               'origLocation', 'originatingLab']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "strains = df[['name', 'accession', 'accessions', 'gisaidId', 'source', 'taxonomyId', 'hostTaxonomyId', 'lineage',\n",
    "              'sequenceLength', 'completeness', 'gender', 'age', 'collectionDate', 'location', \n",
    "              'origLocation', 'originatingLab']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gisaidId</th>\n",
       "      <th>source</th>\n",
       "      <th>taxonomyId</th>\n",
       "      <th>hostTaxonomyId</th>\n",
       "      <th>lineage</th>\n",
       "      <th>sequenceLength</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>location</th>\n",
       "      <th>origLocation</th>\n",
       "      <th>originatingLab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/Brazil/AP162966-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458142</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458142;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458142</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.314</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>37</td>\n",
       "      <td>2020-04-05</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>Brazil,Amapa</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-164827/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524788</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524788;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524788</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>43</td>\n",
       "      <td>2020-04-24</td>\n",
       "      <td>Brazil / Maranhao</td>\n",
       "      <td>Brazil,Maranhao</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-164920/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524796</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524796;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524796</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>33</td>\n",
       "      <td>2020-04-23</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>Brazil,Amapa</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCoV-19/Brazil/AP164082-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458143</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458143;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458143</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>44</td>\n",
       "      <td>2020-04-15</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>Brazil,Amapa</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCoV-19/Brazil/MA-IEC-166716/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524789</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524789;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524789</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>32</td>\n",
       "      <td>2020-05-05</td>\n",
       "      <td>Brazil / Maranhao</td>\n",
       "      <td>Brazil,Maranhao</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                name                              accession  \\\n",
       "0   hCoV-19/Brazil/AP162966-IEC/2020  https://www.gisaid.org/EPI_ISL_458142   \n",
       "1  hCoV-19/Brazil/MA-IEC-164827/2020  https://www.gisaid.org/EPI_ISL_524788   \n",
       "2  hCoV-19/Brazil/AP-IEC-164920/2020  https://www.gisaid.org/EPI_ISL_524796   \n",
       "3   hCoV-19/Brazil/AP164082-IEC/2020  https://www.gisaid.org/EPI_ISL_458143   \n",
       "4  hCoV-19/Brazil/MA-IEC-166716/2020  https://www.gisaid.org/EPI_ISL_524789   \n",
       "\n",
       "                               accessions  \\\n",
       "0  https://www.gisaid.org/EPI_ISL_458142;   \n",
       "1  https://www.gisaid.org/EPI_ISL_524788;   \n",
       "2  https://www.gisaid.org/EPI_ISL_524796;   \n",
       "3  https://www.gisaid.org/EPI_ISL_458143;   \n",
       "4  https://www.gisaid.org/EPI_ISL_524789;   \n",
       "\n",
       "                                gisaidId  source        taxonomyId  \\\n",
       "0  https://www.gisaid.org/EPI_ISL_458142  GISAID  taxonomy:2697049   \n",
       "1  https://www.gisaid.org/EPI_ISL_524788  GISAID  taxonomy:2697049   \n",
       "2  https://www.gisaid.org/EPI_ISL_524796  GISAID  taxonomy:2697049   \n",
       "3  https://www.gisaid.org/EPI_ISL_458143  GISAID  taxonomy:2697049   \n",
       "4  https://www.gisaid.org/EPI_ISL_524789  GISAID  taxonomy:2697049   \n",
       "\n",
       "  hostTaxonomyId    lineage  sequenceLength completeness  gender age  \\\n",
       "0  taxonomy:9606  B.1.1.314           29903     Complete  Female  37   \n",
       "1  taxonomy:9606   B.1.1.33           29903     Complete    Male  43   \n",
       "2  taxonomy:9606   B.1.1.33           29903     Complete  Female  33   \n",
       "3  taxonomy:9606   B.1.1.33           29903     Complete    Male  44   \n",
       "4  taxonomy:9606   B.1.1.33           29903     Complete    Male  32   \n",
       "\n",
       "  collectionDate           location     origLocation  \\\n",
       "0     2020-04-05     Brazil / Amapa     Brazil,Amapa   \n",
       "1     2020-04-24  Brazil / Maranhao  Brazil,Maranhao   \n",
       "2     2020-04-23     Brazil / Amapa     Brazil,Amapa   \n",
       "3     2020-04-15     Brazil / Amapa     Brazil,Amapa   \n",
       "4     2020-05-05  Brazil / Maranhao  Brazil,Maranhao   \n",
       "\n",
       "              originatingLab  \n",
       "0  ﻿Evandro Chagas Institute  \n",
       "1  ﻿Evandro Chagas Institute  \n",
       "2  ﻿Evandro Chagas Institute  \n",
       "3  ﻿Evandro Chagas Institute  \n",
       "4  ﻿Evandro Chagas Institute  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strains.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of strains: 405128\n"
     ]
    }
   ],
   "source": [
    "print('Number of strains:',strains.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "strains.to_csv(NEO4J_IMPORT / \"01c-CNCBStrainPre.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
