{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load SARS-CoV-2 Virus Strain Metadata from CNCB\n",
    "**[Work in progress]**\n",
    "\n",
    "This notebook downloads and standardizes viral strain data from CNCB for ingestion into a Knowledge Graph.\n",
    "\n",
    "Data source: [China National Center for Bioinformation, 2019 Novel Coronavirus Resource (2019nCoVR)](https://bigd.big.ac.cn/ncov/release_genome)\n",
    "\n",
    "Author: Peter Rose (pwrose@ucsd.edu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import glob\n",
    "import ftplib\n",
    "import re\n",
    "import requests\n",
    "import json\n",
    "import dateutil\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_rows = None  # display all rows\n",
    "pd.options.display.max_columns = None  # display all columsns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_url = \"https://bigd.big.ac.cn/ncov/genome/export/meta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/peter/Library/Application Support/com.Neo4j.Relate/data/dbmss/dbms-8bf637fc-0d20-4d9f-9c6f-f7e72e92a4da/import\n"
     ]
    }
   ],
   "source": [
    "# Path will take care of handling operating system differences.\n",
    "NEO4J_IMPORT = Path(os.getenv('NEO4J_IMPORT'))\n",
    "print(NEO4J_IMPORT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this values should be >= then the current number of isolates in CNCB\n",
    "max_entries = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = f'https://bigd.big.ac.cn/ncov/genome/query?&columns%5B0%5D.data=accession&columns%5B0%5D.name=&columns%5B0%5D.searchable=false&columns%5B0%5D.orderable=false&columns%5B0%5D.search.value=&columns%5B0%5D.search.regex=false&columns%5B1%5D.data=name&columns%5B1%5D.name=&columns%5B1%5D.searchable=true&columns%5B1%5D.orderable=true&columns%5B1%5D.search.value=&columns%5B1%5D.search.regex=false&columns%5B2%5D.data=accession&columns%5B2%5D.name=&columns%5B2%5D.searchable=true&columns%5B2%5D.orderable=true&columns%5B2%5D.search.value=&columns%5B2%5D.search.regex=false&columns%5B3%5D.data=gender&columns%5B3%5D.name=&columns%5B3%5D.searchable=true&columns%5B3%5D.orderable=true&columns%5B3%5D.search.value=&columns%5B3%5D.search.regex=false&columns%5B4%5D.data=age&columns%5B4%5D.name=&columns%5B4%5D.searchable=true&columns%5B4%5D.orderable=true&columns%5B4%5D.search.value=&columns%5B4%5D.search.regex=false&columns%5B5%5D.data=source&columns%5B5%5D.name=&columns%5B5%5D.searchable=true&columns%5B5%5D.orderable=true&columns%5B5%5D.search.value=&columns%5B5%5D.search.regex=false&columns%5B6%5D.data=relatedAccession&columns%5B6%5D.name=&columns%5B6%5D.searchable=true&columns%5B6%5D.orderable=true&columns%5B6%5D.search.value=&columns%5B6%5D.search.regex=false&columns%5B7%5D.data=genomeLineage.lineage&columns%5B7%5D.name=&columns%5B7%5D.searchable=true&columns%5B7%5D.orderable=true&columns%5B7%5D.search.value=&columns%5B7%5D.search.regex=false&columns%5B8%5D.data=completeness&columns%5B8%5D.name=&columns%5B8%5D.searchable=true&columns%5B8%5D.orderable=true&columns%5B8%5D.search.value=&columns%5B8%5D.search.regex=false&columns%5B9%5D.data=genomeQuality&columns%5B9%5D.name=&columns%5B9%5D.searchable=false&columns%5B9%5D.orderable=true&columns%5B9%5D.search.value=&columns%5B9%5D.search.regex=false&columns%5B10%5D.data=genomeQuality&columns%5B10%5D.name=&columns%5B10%5D.searchable=false&columns%5B10%5D.orderable=true&columns%5B10%5D.search.value=&columns%5B10%5D.search.regex=false&columns%5B11%5D.data=genomeQuality&columns%5B11%5D.name=&columns%5B11%5D.searchable=false&columns%5B11%5D.orderable=true&columns%5B11%5D.search.value=&columns%5B11%5D.search.regex=false&columns%5B12%5D.data=host&columns%5B12%5D.name=&columns%5B12%5D.searchable=true&columns%5B12%5D.orderable=true&columns%5B12%5D.search.value=&columns%5B12%5D.search.regex=false&columns%5B13%5D.data=collectDate&columns%5B13%5D.name=&columns%5B13%5D.searchable=true&columns%5B13%5D.orderable=true&columns%5B13%5D.search.value=&columns%5B13%5D.search.regex=false&columns%5B14%5D.data=location&columns%5B14%5D.name=&columns%5B14%5D.searchable=true&columns%5B14%5D.orderable=true&columns%5B14%5D.search.value=&columns%5B14%5D.search.regex=false&columns%5B15%5D.data=dataProvider&columns%5B15%5D.name=&columns%5B15%5D.searchable=true&columns%5B15%5D.orderable=true&columns%5B15%5D.search.value=&columns%5B15%5D.search.regex=false&columns%5B16%5D.data=submitDate&columns%5B16%5D.name=&columns%5B16%5D.searchable=true&columns%5B16%5D.orderable=true&columns%5B16%5D.search.value=&columns%5B16%5D.search.regex=false&columns%5B17%5D.data=dataSubmitter&columns%5B17%5D.name=&columns%5B17%5D.searchable=true&columns%5B17%5D.orderable=true&columns%5B17%5D.search.value=&columns%5B17%5D.search.regex=false&columns%5B18%5D.data=createDate&columns%5B18%5D.name=&columns%5B18%5D.searchable=true&columns%5B18%5D.orderable=true&columns%5B18%5D.search.value=&columns%5B18%5D.search.regex=false&columns%5B19%5D.data=country&columns%5B19%5D.name=&columns%5B19%5D.searchable=true&columns%5B19%5D.orderable=true&columns%5B19%5D.search.value=&columns%5B19%5D.search.regex=false&columns%5B20%5D.data=province&columns%5B20%5D.name=&columns%5B20%5D.searchable=true&columns%5B20%5D.orderable=true&columns%5B20%5D.search.value=&columns%5B20%5D.search.regex=false&columns%5B21%5D.data=lastModified&columns%5B21%5D.name=&columns%5B21%5D.searchable=true&columns%5B21%5D.orderable=true&columns%5B21%5D.search.value=&columns%5B21%5D.search.regex=false&order%5B0%5D.column=15&order%5B0%5D.dir=desc&start=0&length={max_entries}&search.value=&search.regex=false&'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "response = json.loads(requests.get(url).text)\n",
    "data = response['data']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://stackoverflow.com/questions/52795561/flattening-nested-json-in-pandas-data-frame\n",
    "def flatten_json(nested_json, exclude=['']):\n",
    "    \"\"\"Flatten json object with nested keys into a single level.\n",
    "        Args:\n",
    "            nested_json: A nested json object.\n",
    "            exclude: Keys to exclude from output.\n",
    "        Returns:\n",
    "            The flattened json object if successful, None otherwise.\n",
    "    \"\"\"\n",
    "    out = {}\n",
    "\n",
    "    def flatten(x, name='', exclude=exclude):\n",
    "        if type(x) is dict:\n",
    "            for a in x:\n",
    "                if a not in exclude: flatten(x[a], name + a + '_')\n",
    "        elif type(x) is list:\n",
    "            i = 0\n",
    "            for a in x:\n",
    "                flatten(a, name + str(i) + '_')\n",
    "                i += 1\n",
    "        else:\n",
    "            out[name[:-1]] = x\n",
    "\n",
    "    flatten(nested_json)\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame([flatten_json(x) for x in data])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.fillna('', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 294247 entries, 0 to 294246\n",
      "Data columns (total 43 columns):\n",
      " #   Column                         Non-Null Count   Dtype \n",
      "---  ------                         --------------   ----- \n",
      " 0   id                             294247 non-null  int64 \n",
      " 1   name                           294247 non-null  object\n",
      " 2   accession                      294247 non-null  object\n",
      " 3   source                         294247 non-null  object\n",
      " 4   link                           294247 non-null  object\n",
      " 5   location                       294247 non-null  object\n",
      " 6   collectDate                    294247 non-null  object\n",
      " 7   dataProvider                   294247 non-null  object\n",
      " 8   submitDate                     294247 non-null  object\n",
      " 9   dataSubmitter                  294247 non-null  object\n",
      " 10  host                           294247 non-null  object\n",
      " 11  country                        294247 non-null  object\n",
      " 12  province                       294247 non-null  object\n",
      " 13  city                           294247 non-null  object\n",
      " 14  relatedAccession               294247 non-null  object\n",
      " 15  completeness                   294247 non-null  object\n",
      " 16  gender                         294247 non-null  object\n",
      " 17  age                            294247 non-null  object\n",
      " 18  lastModified                   294247 non-null  object\n",
      " 19  createDate                     294247 non-null  object\n",
      " 20  genomeQuality_id               294247 non-null  int64 \n",
      " 21  genomeQuality_completeness     294247 non-null  object\n",
      " 22  genomeQuality_virusLength      294247 non-null  int64 \n",
      " 23  genomeQuality_nratio           294247 non-null  object\n",
      " 24  genomeQuality_nbase            294247 non-null  int64 \n",
      " 25  genomeQuality_mutation         294247 non-null  int64 \n",
      " 26  genomeQuality_degenerateBase   294247 non-null  int64 \n",
      " 27  genomeQuality_gap              294247 non-null  int64 \n",
      " 28  genomeQuality_maxGapLength     294247 non-null  int64 \n",
      " 29  genomeQuality_highDensity      294247 non-null  object\n",
      " 30  genomeQuality_accession        294247 non-null  object\n",
      " 31  genomeQuality_virusName        294247 non-null  object\n",
      " 32  genomeQuality_createTime       294247 non-null  object\n",
      " 33  genomeQuality_lastModified     294247 non-null  object\n",
      " 34  genomeLineage_id               294247 non-null  object\n",
      " 35  genomeLineage_accession        294247 non-null  object\n",
      " 36  genomeLineage_lineage          294247 non-null  object\n",
      " 37  genomeLineage_lineagesVersion  294247 non-null  object\n",
      " 38  genomeLineage_status           294247 non-null  object\n",
      " 39  genomeLineage_note             294247 non-null  object\n",
      " 40  genomeLineage_createTime       294247 non-null  object\n",
      " 41  genomeLineage_lastModified     294247 non-null  object\n",
      " 42  genomeLineage                  294247 non-null  object\n",
      "dtypes: int64(8), object(35)\n",
      "memory usage: 96.5+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download strain metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df = pd.read_excel(metadata_url, dtype='str')\n",
    "#df.fillna('', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of strains: 294247\n"
     ]
    }
   ],
   "source": [
    "print(\"Total number of strains:\", df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>source</th>\n",
       "      <th>link</th>\n",
       "      <th>location</th>\n",
       "      <th>collectDate</th>\n",
       "      <th>dataProvider</th>\n",
       "      <th>submitDate</th>\n",
       "      <th>dataSubmitter</th>\n",
       "      <th>host</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>relatedAccession</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>lastModified</th>\n",
       "      <th>createDate</th>\n",
       "      <th>genomeQuality_id</th>\n",
       "      <th>genomeQuality_completeness</th>\n",
       "      <th>genomeQuality_virusLength</th>\n",
       "      <th>genomeQuality_nratio</th>\n",
       "      <th>genomeQuality_nbase</th>\n",
       "      <th>genomeQuality_mutation</th>\n",
       "      <th>genomeQuality_degenerateBase</th>\n",
       "      <th>genomeQuality_gap</th>\n",
       "      <th>genomeQuality_maxGapLength</th>\n",
       "      <th>genomeQuality_highDensity</th>\n",
       "      <th>genomeQuality_accession</th>\n",
       "      <th>genomeQuality_virusName</th>\n",
       "      <th>genomeQuality_createTime</th>\n",
       "      <th>genomeQuality_lastModified</th>\n",
       "      <th>genomeLineage_id</th>\n",
       "      <th>genomeLineage_accession</th>\n",
       "      <th>genomeLineage_lineage</th>\n",
       "      <th>genomeLineage_lineagesVersion</th>\n",
       "      <th>genomeLineage_status</th>\n",
       "      <th>genomeLineage_note</th>\n",
       "      <th>genomeLineage_createTime</th>\n",
       "      <th>genomeLineage_lastModified</th>\n",
       "      <th>genomeLineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43580</td>\n",
       "      <td>hCoV-19/Brazil/AP162741-IEC/2020</td>\n",
       "      <td>EPI_ISL_458138</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>37</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41656</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458138</td>\n",
       "      <td>hCoV-19/Brazil/AP162741-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2.62809e+07</td>\n",
       "      <td>EPI_ISL_458138</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100669</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-164747/2020</td>\n",
       "      <td>EPI_ISL_524783</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>25</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>2020-04-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Para</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>25</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100014</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524783</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-164747/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2.62457e+07</td>\n",
       "      <td>EPI_ISL_524783</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>43581</td>\n",
       "      <td>hCoV-19/Brazil/AC162535-IEC/2020</td>\n",
       "      <td>EPI_ISL_458139</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Acre</td>\n",
       "      <td>2020-03-18</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Acre</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>81</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41657</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458139</td>\n",
       "      <td>hCoV-19/Brazil/AC162535-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2.62809e+07</td>\n",
       "      <td>EPI_ISL_458139</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100670</td>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-165513/2020</td>\n",
       "      <td>EPI_ISL_524784</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>49</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>49</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100015</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524784</td>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-165513/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2.63981e+07</td>\n",
       "      <td>EPI_ISL_524784</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>43582</td>\n",
       "      <td>hCoV-19/Brazil/PA162802-IEC/2020</td>\n",
       "      <td>EPI_ISL_458140</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>2020-04-07</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Para</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>63</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41658</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458140</td>\n",
       "      <td>hCoV-19/Brazil/PA162802-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2.63127e+07</td>\n",
       "      <td>EPI_ISL_458140</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>100671</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-165313/2020</td>\n",
       "      <td>EPI_ISL_524785</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>34</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>2020-05-05</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Para</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>34</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100016</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524785</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-165313/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2.63984e+07</td>\n",
       "      <td>EPI_ISL_524785</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>43583</td>\n",
       "      <td>hCoV-19/Brazil/PA164239-IEC/2020</td>\n",
       "      <td>EPI_ISL_458141</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>2020-04-26</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Para</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>38</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41659</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458141</td>\n",
       "      <td>hCoV-19/Brazil/PA164239-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2.62809e+07</td>\n",
       "      <td>EPI_ISL_458141</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>100672</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-165302/2020</td>\n",
       "      <td>EPI_ISL_524786</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>58</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>2020-05-01</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Para</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>58</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100017</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524786</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-165302/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2.63983e+07</td>\n",
       "      <td>EPI_ISL_524786</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>43584</td>\n",
       "      <td>hCoV-19/Brazil/AP162966-IEC/2020</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-05</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>37</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41660</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>hCoV-19/Brazil/AP162966-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2.63127e+07</td>\n",
       "      <td>EPI_ISL_458142</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>100673</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-163469/2020</td>\n",
       "      <td>EPI_ISL_524787</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>32</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>2020-04-15</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Para</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>32</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100018</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524787</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-163469/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2.62457e+07</td>\n",
       "      <td>EPI_ISL_524787</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                               name       accession  source  \\\n",
       "0   43580   hCoV-19/Brazil/AP162741-IEC/2020  EPI_ISL_458138  GISAID   \n",
       "1  100669  hCoV-19/Brazil/PA-IEC-164747/2020  EPI_ISL_524783  GISAID   \n",
       "2   43581   hCoV-19/Brazil/AC162535-IEC/2020  EPI_ISL_458139  GISAID   \n",
       "3  100670  hCoV-19/Brazil/AP-IEC-165513/2020  EPI_ISL_524784  GISAID   \n",
       "4   43582   hCoV-19/Brazil/PA162802-IEC/2020  EPI_ISL_458140  GISAID   \n",
       "5  100671  hCoV-19/Brazil/PA-IEC-165313/2020  EPI_ISL_524785  GISAID   \n",
       "6   43583   hCoV-19/Brazil/PA164239-IEC/2020  EPI_ISL_458141  GISAID   \n",
       "7  100672  hCoV-19/Brazil/PA-IEC-165302/2020  EPI_ISL_524786  GISAID   \n",
       "8   43584   hCoV-19/Brazil/AP162966-IEC/2020  EPI_ISL_458142  GISAID   \n",
       "9  100673  hCoV-19/Brazil/PA-IEC-163469/2020  EPI_ISL_524787  GISAID   \n",
       "\n",
       "                  link        location collectDate               dataProvider  \\\n",
       "0  https://gisaid.org/  Brazil / Amapa  2020-04-03  ﻿Evandro Chagas Institute   \n",
       "1                   25   Brazil / Para  2020-04-28  ﻿Evandro Chagas Institute   \n",
       "2  https://gisaid.org/   Brazil / Acre  2020-03-18  ﻿Evandro Chagas Institute   \n",
       "3                   49  Brazil / Amapa  2020-04-29  ﻿Evandro Chagas Institute   \n",
       "4  https://gisaid.org/   Brazil / Para  2020-04-07  ﻿Evandro Chagas Institute   \n",
       "5                   34   Brazil / Para  2020-05-05  ﻿Evandro Chagas Institute   \n",
       "6  https://gisaid.org/   Brazil / Para  2020-04-26  ﻿Evandro Chagas Institute   \n",
       "7                   58   Brazil / Para  2020-05-01  ﻿Evandro Chagas Institute   \n",
       "8  https://gisaid.org/  Brazil / Amapa  2020-04-05  ﻿Evandro Chagas Institute   \n",
       "9                   32   Brazil / Para  2020-04-15  ﻿Evandro Chagas Institute   \n",
       "\n",
       "   submitDate              dataSubmitter          host country province city  \\\n",
       "0  2020-06-03  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil    Amapa        \n",
       "1  2020-08-28  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Para        \n",
       "2  2020-06-03  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Acre        \n",
       "3  2020-08-28  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil    Amapa        \n",
       "4  2020-06-03  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Para        \n",
       "5  2020-08-28  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Para        \n",
       "6  2020-06-03  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Para        \n",
       "7  2020-08-28  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Para        \n",
       "8  2020-06-03  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil    Amapa        \n",
       "9  2020-08-28  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Para        \n",
       "\n",
       "  relatedAccession completeness  gender age         lastModified  \\\n",
       "0                      Complete  Female  37  2020-09-09 11:31:17   \n",
       "1                      Complete  Female  25  2020-09-09 11:31:17   \n",
       "2                      Complete    Male  81  2020-09-09 11:31:17   \n",
       "3                      Complete    Male  49  2020-09-09 11:31:17   \n",
       "4                      Complete    Male  63  2020-09-09 11:31:17   \n",
       "5                      Complete    Male  34  2020-09-09 11:31:17   \n",
       "6                      Complete    Male  38  2020-09-09 11:31:17   \n",
       "7                      Complete    Male  58  2020-09-09 11:31:17   \n",
       "8                      Complete  Female  37  2020-09-09 11:31:17   \n",
       "9                      Complete    Male  32  2020-09-09 11:31:17   \n",
       "\n",
       "            createDate  genomeQuality_id genomeQuality_completeness  \\\n",
       "0  2020-06-04 12:19:36             41656                   Complete   \n",
       "1  2020-08-29 18:49:44            100014                   Complete   \n",
       "2  2020-06-04 12:19:36             41657                   Complete   \n",
       "3  2020-08-29 18:49:44            100015                   Complete   \n",
       "4  2020-06-04 12:19:36             41658                   Complete   \n",
       "5  2020-08-29 18:49:44            100016                   Complete   \n",
       "6  2020-06-04 12:19:36             41659                   Complete   \n",
       "7  2020-08-29 18:49:44            100017                   Complete   \n",
       "8  2020-06-04 12:19:36             41660                   Complete   \n",
       "9  2020-08-29 18:49:44            100018                   Complete   \n",
       "\n",
       "   genomeQuality_virusLength genomeQuality_nratio  genomeQuality_nbase  \\\n",
       "0                      29903                0.00%                    0   \n",
       "1                      29903                0.00%                    0   \n",
       "2                      29903                0.00%                    0   \n",
       "3                      29903                0.00%                    0   \n",
       "4                      29903                0.00%                    0   \n",
       "5                      29903                0.00%                    0   \n",
       "6                      29903                0.00%                    0   \n",
       "7                      29903                0.00%                    0   \n",
       "8                      29903                0.00%                    0   \n",
       "9                      29903                0.00%                    0   \n",
       "\n",
       "   genomeQuality_mutation  genomeQuality_degenerateBase  genomeQuality_gap  \\\n",
       "0                      11                             0                  0   \n",
       "1                      11                             0                  0   \n",
       "2                      11                             0                  0   \n",
       "3                      13                             0                  0   \n",
       "4                      10                             0                  0   \n",
       "5                      13                             0                  0   \n",
       "6                      12                             0                  0   \n",
       "7                      11                             0                  0   \n",
       "8                      11                             0                  0   \n",
       "9                      10                             0                  0   \n",
       "\n",
       "   genomeQuality_maxGapLength  \\\n",
       "0                          -1   \n",
       "1                          -1   \n",
       "2                          -1   \n",
       "3                          -1   \n",
       "4                          -1   \n",
       "5                          -1   \n",
       "6                          -1   \n",
       "7                          -1   \n",
       "8                          -1   \n",
       "9                          -1   \n",
       "\n",
       "                           genomeQuality_highDensity genomeQuality_accession  \\\n",
       "0  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458138   \n",
       "1  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524783   \n",
       "2  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458139   \n",
       "3  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524784   \n",
       "4  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458140   \n",
       "5  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524785   \n",
       "6  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458141   \n",
       "7  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524786   \n",
       "8  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458142   \n",
       "9  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524787   \n",
       "\n",
       "             genomeQuality_virusName genomeQuality_createTime  \\\n",
       "0   hCoV-19/Brazil/AP162741-IEC/2020      2020-06-04 12:20:14   \n",
       "1  hCoV-19/Brazil/PA-IEC-164747/2020      2020-08-29 18:52:11   \n",
       "2   hCoV-19/Brazil/AC162535-IEC/2020      2020-06-04 12:20:14   \n",
       "3  hCoV-19/Brazil/AP-IEC-165513/2020      2020-08-29 18:52:11   \n",
       "4   hCoV-19/Brazil/PA162802-IEC/2020      2020-06-04 12:20:14   \n",
       "5  hCoV-19/Brazil/PA-IEC-165313/2020      2020-08-29 18:52:11   \n",
       "6   hCoV-19/Brazil/PA164239-IEC/2020      2020-06-04 12:20:14   \n",
       "7  hCoV-19/Brazil/PA-IEC-165302/2020      2020-08-29 18:52:11   \n",
       "8   hCoV-19/Brazil/AP162966-IEC/2020      2020-06-04 12:20:14   \n",
       "9  hCoV-19/Brazil/PA-IEC-163469/2020      2020-08-29 18:52:11   \n",
       "\n",
       "  genomeQuality_lastModified genomeLineage_id genomeLineage_accession  \\\n",
       "0        2020-06-04 12:20:14      2.62809e+07          EPI_ISL_458138   \n",
       "1        2020-08-29 18:52:11      2.62457e+07          EPI_ISL_524783   \n",
       "2        2020-06-04 12:20:14      2.62809e+07          EPI_ISL_458139   \n",
       "3        2020-08-29 18:52:11      2.63981e+07          EPI_ISL_524784   \n",
       "4        2020-06-04 12:20:14      2.63127e+07          EPI_ISL_458140   \n",
       "5        2020-08-29 18:52:11      2.63984e+07          EPI_ISL_524785   \n",
       "6        2020-06-04 12:20:14      2.62809e+07          EPI_ISL_458141   \n",
       "7        2020-08-29 18:52:11      2.63983e+07          EPI_ISL_524786   \n",
       "8        2020-06-04 12:20:14      2.63127e+07          EPI_ISL_458142   \n",
       "9        2020-08-29 18:52:11      2.62457e+07          EPI_ISL_524787   \n",
       "\n",
       "  genomeLineage_lineage genomeLineage_lineagesVersion genomeLineage_status  \\\n",
       "0              B.1.1.33                    2020/10/30            passed_qc   \n",
       "1              B.1.1.28                    2020/10/30            passed_qc   \n",
       "2              B.1.1.33                    2020/10/30            passed_qc   \n",
       "3              B.1.1.33                    2020/10/30            passed_qc   \n",
       "4              B.1.1.28                    2020/10/30            passed_qc   \n",
       "5              B.1.1.28                    2020/10/30            passed_qc   \n",
       "6              B.1.1.28                    2020/10/30            passed_qc   \n",
       "7              B.1.1.28                    2020/10/30            passed_qc   \n",
       "8              B.1.1.33                    2020/10/30            passed_qc   \n",
       "9              B.1.1.28                    2020/10/30            passed_qc   \n",
       "\n",
       "  genomeLineage_note genomeLineage_createTime genomeLineage_lastModified  \\\n",
       "0                         2020-11-30 15:44:35        2020-11-30 15:44:35   \n",
       "1                         2020-11-30 15:44:32        2020-11-30 15:44:32   \n",
       "2                         2020-11-30 15:44:35        2020-11-30 15:44:35   \n",
       "3                         2020-11-30 15:44:43        2020-11-30 15:44:43   \n",
       "4                         2020-11-30 15:44:38        2020-11-30 15:44:38   \n",
       "5                         2020-11-30 15:44:43        2020-11-30 15:44:43   \n",
       "6                         2020-11-30 15:44:35        2020-11-30 15:44:35   \n",
       "7                         2020-11-30 15:44:43        2020-11-30 15:44:43   \n",
       "8                         2020-11-30 15:44:38        2020-11-30 15:44:38   \n",
       "9                         2020-11-30 15:44:32        2020-11-30 15:44:32   \n",
       "\n",
       "  genomeLineage  \n",
       "0                \n",
       "1                \n",
       "2                \n",
       "3                \n",
       "4                \n",
       "5                \n",
       "6                \n",
       "7                \n",
       "8                \n",
       "9                "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Asssign unique identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://registry.identifiers.org/registry/insdc\n",
    "insdc_pattern = re.compile('^([A-Z]\\d{5}|[A-Z]{2}\\d{6}|[A-Z]{4}\\d{8}|[A-J][A-Z]{2}\\d{5})(\\.\\d+)?$')\n",
    "# https://registry.identifiers.org/registry/refseq\n",
    "refseq_pattern = re.compile('^(((AC|AP|NC|NG|NM|NP|NR|NT|NW|XM|XP|XR|YP|ZP)_\\d+)|(NZ\\_[A-Z]{2,4}\\d+))(\\.\\d+)?$')\n",
    "epi_pattern = re.compile('^EPI_ISL_\\d+$')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_curie(id):\n",
    "    id = id.strip()\n",
    "    if len(id) > 0:\n",
    "        if id.startswith('EPI'):\n",
    "            return 'https://www.gisaid.org/' + id\n",
    "        elif refseq_pattern.match(id) != None:\n",
    "            return 'refseq:' + id\n",
    "        elif insdc_pattern.match(id) != None:\n",
    "            return 'insdc:' + id\n",
    "        else:\n",
    "            # TODO are URIs available for these cases?\n",
    "            return id\n",
    "    else:\n",
    "        return id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_curies(ids):\n",
    "    return [assign_curie(id) for id in ids.split(',')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gisaid_id(ids):\n",
    "    for id in ids:\n",
    "        if id.startswith('https://www.gisaid.org/'):\n",
    "            return id\n",
    "        \n",
    "    return ''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rename and concatenate fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df['Accession ID'] = df['Accession ID'].str.strip()\n",
    "#df['Related ID'] = df['Related ID'].str.strip()\n",
    "\n",
    "# combine all ids into an accession column and assign curies\n",
    "df['accessions'] = df['accession'] + df['relatedAccession'].apply(lambda s: ',' + s if len(s) > 0 else s)\n",
    "df['accessions'] = df['accessions'].apply(assign_curies)\n",
    "df['gisaidId'] = df['accessions'].apply(get_gisaid_id)\n",
    "df['accessions'] = df['accessions'].apply(lambda x: ';'.join(x))\n",
    "\n",
    "df['accession'] = df['accession'].apply(lambda s: assign_curie(s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.rename(columns={'genomeQuality_virusLength': 'sequenceLength'}, inplace=True)\n",
    "#df.rename(columns={'Sequence Quality': 'sequenceQuality'}, inplace=True) # get from original spreadsheet\n",
    "#df.rename(columns={'Quality Assessment': 'qualityAssessment'}, inplace=True) # get from original spreadsheet\n",
    "df.rename(columns={'dataProvider': 'originatingLab'}, inplace=True)\n",
    "#df.rename(columns={'Virus Strain Name': 'name'}, inplace=True)\n",
    "df.rename(columns={'collectDate':'collectionDate'},inplace=True)\n",
    "df.rename(columns={'Location':'location'}, inplace=True)\n",
    "df.rename(columns={'genomeLineage_lineage': 'lineage'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove invalid collection date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>source</th>\n",
       "      <th>link</th>\n",
       "      <th>location</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>originatingLab</th>\n",
       "      <th>submitDate</th>\n",
       "      <th>dataSubmitter</th>\n",
       "      <th>host</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>relatedAccession</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>lastModified</th>\n",
       "      <th>createDate</th>\n",
       "      <th>genomeQuality_id</th>\n",
       "      <th>genomeQuality_completeness</th>\n",
       "      <th>sequenceLength</th>\n",
       "      <th>genomeQuality_nratio</th>\n",
       "      <th>genomeQuality_nbase</th>\n",
       "      <th>genomeQuality_mutation</th>\n",
       "      <th>genomeQuality_degenerateBase</th>\n",
       "      <th>genomeQuality_gap</th>\n",
       "      <th>genomeQuality_maxGapLength</th>\n",
       "      <th>genomeQuality_highDensity</th>\n",
       "      <th>genomeQuality_accession</th>\n",
       "      <th>genomeQuality_virusName</th>\n",
       "      <th>genomeQuality_createTime</th>\n",
       "      <th>genomeQuality_lastModified</th>\n",
       "      <th>genomeLineage_id</th>\n",
       "      <th>genomeLineage_accession</th>\n",
       "      <th>lineage</th>\n",
       "      <th>genomeLineage_lineagesVersion</th>\n",
       "      <th>genomeLineage_status</th>\n",
       "      <th>genomeLineage_note</th>\n",
       "      <th>genomeLineage_createTime</th>\n",
       "      <th>genomeLineage_lastModified</th>\n",
       "      <th>genomeLineage</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gisaidId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>960</th>\n",
       "      <td>104895</td>\n",
       "      <td>covid_hub_pl_ibch_0028</td>\n",
       "      <td>insdc:LR877414</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/LR877414</td>\n",
       "      <td>Poland</td>\n",
       "      <td>2020-00-00</td>\n",
       "      <td>WSSE</td>\n",
       "      <td>2020-08-17</td>\n",
       "      <td>COVID-HUB-PL, Institute of Bioorganic Chemistr...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Poland</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Partial</td>\n",
       "      <td>Female</td>\n",
       "      <td></td>\n",
       "      <td>2020-09-14 23:26:05</td>\n",
       "      <td>2020-09-09 18:34:25</td>\n",
       "      <td>104240</td>\n",
       "      <td>Partial</td>\n",
       "      <td>29903</td>\n",
       "      <td>91.98%</td>\n",
       "      <td>27050</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>LR877414</td>\n",
       "      <td>covid_hub_pl_ibch_0028</td>\n",
       "      <td>2020-09-09 18:35:26</td>\n",
       "      <td>2020-09-09 18:35:26</td>\n",
       "      <td>1.31674e+07</td>\n",
       "      <td>LR877414</td>\n",
       "      <td>NA</td>\n",
       "      <td>2020-10-30</td>\n",
       "      <td>fail</td>\n",
       "      <td>N_content:0.92</td>\n",
       "      <td>2020-11-24 15:50:36</td>\n",
       "      <td>2020-11-24 15:50:36</td>\n",
       "      <td></td>\n",
       "      <td>insdc:LR877414;</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>104905</td>\n",
       "      <td>covid_hub_pl_ibch_0044</td>\n",
       "      <td>insdc:LR877424</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/LR877424</td>\n",
       "      <td>Poland</td>\n",
       "      <td>2020-00-00</td>\n",
       "      <td>WSSE</td>\n",
       "      <td>2020-08-17</td>\n",
       "      <td>COVID-HUB-PL, Institute of Bioorganic Chemistr...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Poland</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td></td>\n",
       "      <td>2020-09-14 23:26:05</td>\n",
       "      <td>2020-09-09 18:34:25</td>\n",
       "      <td>104250</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>2.22%</td>\n",
       "      <td>654</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>LR877424</td>\n",
       "      <td>covid_hub_pl_ibch_0044</td>\n",
       "      <td>2020-09-09 18:35:27</td>\n",
       "      <td>2020-09-09 18:35:27</td>\n",
       "      <td>2.62474e+07</td>\n",
       "      <td>LR877424</td>\n",
       "      <td>B</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td></td>\n",
       "      <td>insdc:LR877424;</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id                    name       accession   source  \\\n",
       "960  104895  covid_hub_pl_ibch_0028  insdc:LR877414  GenBank   \n",
       "972  104905  covid_hub_pl_ibch_0044  insdc:LR877424  GenBank   \n",
       "\n",
       "                                              link location collectionDate  \\\n",
       "960  https://www.ncbi.nlm.nih.gov/nuccore/LR877414   Poland     2020-00-00   \n",
       "972  https://www.ncbi.nlm.nih.gov/nuccore/LR877424   Poland     2020-00-00   \n",
       "\n",
       "    originatingLab  submitDate  \\\n",
       "960           WSSE  2020-08-17   \n",
       "972           WSSE  2020-08-17   \n",
       "\n",
       "                                         dataSubmitter          host country  \\\n",
       "960  COVID-HUB-PL, Institute of Bioorganic Chemistr...  Homo sapiens  Poland   \n",
       "972  COVID-HUB-PL, Institute of Bioorganic Chemistr...  Homo sapiens  Poland   \n",
       "\n",
       "    province city relatedAccession completeness  gender age  \\\n",
       "960                                     Partial  Female       \n",
       "972                                    Complete  Female       \n",
       "\n",
       "            lastModified           createDate  genomeQuality_id  \\\n",
       "960  2020-09-14 23:26:05  2020-09-09 18:34:25            104240   \n",
       "972  2020-09-14 23:26:05  2020-09-09 18:34:25            104250   \n",
       "\n",
       "    genomeQuality_completeness  sequenceLength genomeQuality_nratio  \\\n",
       "960                    Partial           29903               91.98%   \n",
       "972                   Complete           29903                2.22%   \n",
       "\n",
       "     genomeQuality_nbase  genomeQuality_mutation  \\\n",
       "960                27050                      -1   \n",
       "972                  654                      -1   \n",
       "\n",
       "     genomeQuality_degenerateBase  genomeQuality_gap  \\\n",
       "960                             0                 -1   \n",
       "972                             0                 -1   \n",
       "\n",
       "     genomeQuality_maxGapLength genomeQuality_highDensity  \\\n",
       "960                          -1                        -1   \n",
       "972                          -1                        -1   \n",
       "\n",
       "    genomeQuality_accession genomeQuality_virusName genomeQuality_createTime  \\\n",
       "960                LR877414  covid_hub_pl_ibch_0028      2020-09-09 18:35:26   \n",
       "972                LR877424  covid_hub_pl_ibch_0044      2020-09-09 18:35:27   \n",
       "\n",
       "    genomeQuality_lastModified genomeLineage_id genomeLineage_accession  \\\n",
       "960        2020-09-09 18:35:26      1.31674e+07                LR877414   \n",
       "972        2020-09-09 18:35:27      2.62474e+07                LR877424   \n",
       "\n",
       "    lineage genomeLineage_lineagesVersion genomeLineage_status  \\\n",
       "960      NA                    2020-10-30                 fail   \n",
       "972       B                    2020/10/30            passed_qc   \n",
       "\n",
       "    genomeLineage_note genomeLineage_createTime genomeLineage_lastModified  \\\n",
       "960     N_content:0.92      2020-11-24 15:50:36        2020-11-24 15:50:36   \n",
       "972                         2020-11-30 15:44:32        2020-11-30 15:44:32   \n",
       "\n",
       "    genomeLineage       accessions gisaidId  \n",
       "960                insdc:LR877414;           \n",
       "972                insdc:LR877424;           "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.query(\"collectionDate == '2020-00-00'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['collectionDate'] = df['collectionDate'].apply(lambda d: '' if d == '2020-00-00' else d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['collectionDate'] = df['collectionDate'].apply(lambda d: dateutil.parser.parse(d) if len(d) > 0 else '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>source</th>\n",
       "      <th>link</th>\n",
       "      <th>location</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>originatingLab</th>\n",
       "      <th>submitDate</th>\n",
       "      <th>dataSubmitter</th>\n",
       "      <th>host</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>relatedAccession</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>lastModified</th>\n",
       "      <th>createDate</th>\n",
       "      <th>genomeQuality_id</th>\n",
       "      <th>genomeQuality_completeness</th>\n",
       "      <th>sequenceLength</th>\n",
       "      <th>genomeQuality_nratio</th>\n",
       "      <th>genomeQuality_nbase</th>\n",
       "      <th>genomeQuality_mutation</th>\n",
       "      <th>genomeQuality_degenerateBase</th>\n",
       "      <th>genomeQuality_gap</th>\n",
       "      <th>genomeQuality_maxGapLength</th>\n",
       "      <th>genomeQuality_highDensity</th>\n",
       "      <th>genomeQuality_accession</th>\n",
       "      <th>genomeQuality_virusName</th>\n",
       "      <th>genomeQuality_createTime</th>\n",
       "      <th>genomeQuality_lastModified</th>\n",
       "      <th>genomeLineage_id</th>\n",
       "      <th>genomeLineage_accession</th>\n",
       "      <th>lineage</th>\n",
       "      <th>genomeLineage_lineagesVersion</th>\n",
       "      <th>genomeLineage_status</th>\n",
       "      <th>genomeLineage_note</th>\n",
       "      <th>genomeLineage_createTime</th>\n",
       "      <th>genomeLineage_lastModified</th>\n",
       "      <th>genomeLineage</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gisaidId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51106</th>\n",
       "      <td>14</td>\n",
       "      <td>Wuhan-Hu-1</td>\n",
       "      <td>insdc:MN908947</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MN908947</td>\n",
       "      <td>China / Hubei / Wuhan</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>Shanghai Public Health Clinical Center &amp; Schoo...</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>Shanghai Public Health Clinical Center &amp; Schoo...</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>NC_045512,EPI_ISL_402125</td>\n",
       "      <td>Complete</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2020-05-20 11:14:12</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>629</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>NO</td>\n",
       "      <td>MN908947</td>\n",
       "      <td>Wuhan-Hu-1</td>\n",
       "      <td>2020-03-17 14:27:13</td>\n",
       "      <td>2020-03-17 14:27:13</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>insdc:MN908947;refseq:NC_045512;https://www.gi...</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_402125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id        name       accession   source  \\\n",
       "51106  14  Wuhan-Hu-1  insdc:MN908947  GenBank   \n",
       "\n",
       "                                                link               location  \\\n",
       "51106  https://www.ncbi.nlm.nih.gov/nuccore/MN908947  China / Hubei / Wuhan   \n",
       "\n",
       "      collectionDate                                     originatingLab  \\\n",
       "51106     2019-12-30  Shanghai Public Health Clinical Center & Schoo...   \n",
       "\n",
       "       submitDate                                      dataSubmitter  \\\n",
       "51106  2020-01-17  Shanghai Public Health Clinical Center & Schoo...   \n",
       "\n",
       "               host country province   city          relatedAccession  \\\n",
       "51106  Homo sapiens   China    Hubei  Wuhan  NC_045512,EPI_ISL_402125   \n",
       "\n",
       "      completeness gender age         lastModified           createDate  \\\n",
       "51106     Complete             2020-05-20 11:14:12  2020-01-20 20:04:48   \n",
       "\n",
       "       genomeQuality_id genomeQuality_completeness  sequenceLength  \\\n",
       "51106               629                   Complete           29903   \n",
       "\n",
       "      genomeQuality_nratio  genomeQuality_nbase  genomeQuality_mutation  \\\n",
       "51106                    0                    0                       0   \n",
       "\n",
       "       genomeQuality_degenerateBase  genomeQuality_gap  \\\n",
       "51106                             0                  0   \n",
       "\n",
       "       genomeQuality_maxGapLength genomeQuality_highDensity  \\\n",
       "51106                          -1                        NO   \n",
       "\n",
       "      genomeQuality_accession genomeQuality_virusName  \\\n",
       "51106                MN908947              Wuhan-Hu-1   \n",
       "\n",
       "      genomeQuality_createTime genomeQuality_lastModified genomeLineage_id  \\\n",
       "51106      2020-03-17 14:27:13        2020-03-17 14:27:13                    \n",
       "\n",
       "      genomeLineage_accession lineage genomeLineage_lineagesVersion  \\\n",
       "51106                                                                 \n",
       "\n",
       "      genomeLineage_status genomeLineage_note genomeLineage_createTime  \\\n",
       "51106                                                                    \n",
       "\n",
       "      genomeLineage_lastModified genomeLineage  \\\n",
       "51106                                            \n",
       "\n",
       "                                              accessions  \\\n",
       "51106  insdc:MN908947;refseq:NC_045512;https://www.gi...   \n",
       "\n",
       "                                    gisaidId  \n",
       "51106  https://www.gisaid.org/EPI_ISL_402125  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['accessions'].str.contains('refseq:NC_045512')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>source</th>\n",
       "      <th>link</th>\n",
       "      <th>location</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>originatingLab</th>\n",
       "      <th>submitDate</th>\n",
       "      <th>dataSubmitter</th>\n",
       "      <th>host</th>\n",
       "      <th>country</th>\n",
       "      <th>province</th>\n",
       "      <th>city</th>\n",
       "      <th>relatedAccession</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>lastModified</th>\n",
       "      <th>createDate</th>\n",
       "      <th>genomeQuality_id</th>\n",
       "      <th>genomeQuality_completeness</th>\n",
       "      <th>sequenceLength</th>\n",
       "      <th>genomeQuality_nratio</th>\n",
       "      <th>genomeQuality_nbase</th>\n",
       "      <th>genomeQuality_mutation</th>\n",
       "      <th>genomeQuality_degenerateBase</th>\n",
       "      <th>genomeQuality_gap</th>\n",
       "      <th>genomeQuality_maxGapLength</th>\n",
       "      <th>genomeQuality_highDensity</th>\n",
       "      <th>genomeQuality_accession</th>\n",
       "      <th>genomeQuality_virusName</th>\n",
       "      <th>genomeQuality_createTime</th>\n",
       "      <th>genomeQuality_lastModified</th>\n",
       "      <th>genomeLineage_id</th>\n",
       "      <th>genomeLineage_accession</th>\n",
       "      <th>lineage</th>\n",
       "      <th>genomeLineage_lineagesVersion</th>\n",
       "      <th>genomeLineage_status</th>\n",
       "      <th>genomeLineage_note</th>\n",
       "      <th>genomeLineage_createTime</th>\n",
       "      <th>genomeLineage_lastModified</th>\n",
       "      <th>genomeLineage</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gisaidId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>43580</td>\n",
       "      <td>hCoV-19/Brazil/AP162741-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458138</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>37</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41656</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458138</td>\n",
       "      <td>hCoV-19/Brazil/AP162741-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2.62809e+07</td>\n",
       "      <td>EPI_ISL_458138</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458138;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100669</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-164747/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524783</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>25</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>2020-04-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Para</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>25</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100014</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524783</td>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-164747/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2.62457e+07</td>\n",
       "      <td>EPI_ISL_524783</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td>2020-11-30 15:44:32</td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524783;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>43581</td>\n",
       "      <td>hCoV-19/Brazil/AC162535-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458139</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Acre</td>\n",
       "      <td>2020-03-18</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Acre</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>81</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41657</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458139</td>\n",
       "      <td>hCoV-19/Brazil/AC162535-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2.62809e+07</td>\n",
       "      <td>EPI_ISL_458139</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td>2020-11-30 15:44:35</td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458139;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100670</td>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-165513/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524784</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>49</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-08-28</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Amapa</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>49</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-08-29 18:49:44</td>\n",
       "      <td>100015</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_524784</td>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-165513/2020</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2020-08-29 18:52:11</td>\n",
       "      <td>2.63981e+07</td>\n",
       "      <td>EPI_ISL_524784</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td>2020-11-30 15:44:43</td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524784;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>43582</td>\n",
       "      <td>hCoV-19/Brazil/PA162802-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458140</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>https://gisaid.org/</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>2020-04-07</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>2020-06-03</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "      <td>Homo Sapiens</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Para</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>63</td>\n",
       "      <td>2020-09-09 11:31:17</td>\n",
       "      <td>2020-06-04 12:19:36</td>\n",
       "      <td>41658</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29903</td>\n",
       "      <td>0.00%</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...</td>\n",
       "      <td>EPI_ISL_458140</td>\n",
       "      <td>hCoV-19/Brazil/PA162802-IEC/2020</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2020-06-04 12:20:14</td>\n",
       "      <td>2.63127e+07</td>\n",
       "      <td>EPI_ISL_458140</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>2020/10/30</td>\n",
       "      <td>passed_qc</td>\n",
       "      <td></td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td>2020-11-30 15:44:38</td>\n",
       "      <td></td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458140;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       id                               name  \\\n",
       "0   43580   hCoV-19/Brazil/AP162741-IEC/2020   \n",
       "1  100669  hCoV-19/Brazil/PA-IEC-164747/2020   \n",
       "2   43581   hCoV-19/Brazil/AC162535-IEC/2020   \n",
       "3  100670  hCoV-19/Brazil/AP-IEC-165513/2020   \n",
       "4   43582   hCoV-19/Brazil/PA162802-IEC/2020   \n",
       "\n",
       "                               accession  source                 link  \\\n",
       "0  https://www.gisaid.org/EPI_ISL_458138  GISAID  https://gisaid.org/   \n",
       "1  https://www.gisaid.org/EPI_ISL_524783  GISAID                   25   \n",
       "2  https://www.gisaid.org/EPI_ISL_458139  GISAID  https://gisaid.org/   \n",
       "3  https://www.gisaid.org/EPI_ISL_524784  GISAID                   49   \n",
       "4  https://www.gisaid.org/EPI_ISL_458140  GISAID  https://gisaid.org/   \n",
       "\n",
       "         location collectionDate             originatingLab  submitDate  \\\n",
       "0  Brazil / Amapa     2020-04-03  ﻿Evandro Chagas Institute  2020-06-03   \n",
       "1   Brazil / Para     2020-04-28  ﻿Evandro Chagas Institute  2020-08-28   \n",
       "2   Brazil / Acre     2020-03-18  ﻿Evandro Chagas Institute  2020-06-03   \n",
       "3  Brazil / Amapa     2020-04-29  ﻿Evandro Chagas Institute  2020-08-28   \n",
       "4   Brazil / Para     2020-04-07  ﻿Evandro Chagas Institute  2020-06-03   \n",
       "\n",
       "               dataSubmitter          host country province city  \\\n",
       "0  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil    Amapa        \n",
       "1  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Para        \n",
       "2  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Acre        \n",
       "3  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil    Amapa        \n",
       "4  ﻿Evandro Chagas Institute  Homo Sapiens  Brazil     Para        \n",
       "\n",
       "  relatedAccession completeness  gender age         lastModified  \\\n",
       "0                      Complete  Female  37  2020-09-09 11:31:17   \n",
       "1                      Complete  Female  25  2020-09-09 11:31:17   \n",
       "2                      Complete    Male  81  2020-09-09 11:31:17   \n",
       "3                      Complete    Male  49  2020-09-09 11:31:17   \n",
       "4                      Complete    Male  63  2020-09-09 11:31:17   \n",
       "\n",
       "            createDate  genomeQuality_id genomeQuality_completeness  \\\n",
       "0  2020-06-04 12:19:36             41656                   Complete   \n",
       "1  2020-08-29 18:49:44            100014                   Complete   \n",
       "2  2020-06-04 12:19:36             41657                   Complete   \n",
       "3  2020-08-29 18:49:44            100015                   Complete   \n",
       "4  2020-06-04 12:19:36             41658                   Complete   \n",
       "\n",
       "   sequenceLength genomeQuality_nratio  genomeQuality_nbase  \\\n",
       "0           29903                0.00%                    0   \n",
       "1           29903                0.00%                    0   \n",
       "2           29903                0.00%                    0   \n",
       "3           29903                0.00%                    0   \n",
       "4           29903                0.00%                    0   \n",
       "\n",
       "   genomeQuality_mutation  genomeQuality_degenerateBase  genomeQuality_gap  \\\n",
       "0                      11                             0                  0   \n",
       "1                      11                             0                  0   \n",
       "2                      11                             0                  0   \n",
       "3                      13                             0                  0   \n",
       "4                      10                             0                  0   \n",
       "\n",
       "   genomeQuality_maxGapLength  \\\n",
       "0                          -1   \n",
       "1                          -1   \n",
       "2                          -1   \n",
       "3                          -1   \n",
       "4                          -1   \n",
       "\n",
       "                           genomeQuality_highDensity genomeQuality_accession  \\\n",
       "0  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458138   \n",
       "1  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524783   \n",
       "2  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458139   \n",
       "3  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_524784   \n",
       "4  28881~28883(3-3-1.00,SNP:28881; SNP:28882; SNP...          EPI_ISL_458140   \n",
       "\n",
       "             genomeQuality_virusName genomeQuality_createTime  \\\n",
       "0   hCoV-19/Brazil/AP162741-IEC/2020      2020-06-04 12:20:14   \n",
       "1  hCoV-19/Brazil/PA-IEC-164747/2020      2020-08-29 18:52:11   \n",
       "2   hCoV-19/Brazil/AC162535-IEC/2020      2020-06-04 12:20:14   \n",
       "3  hCoV-19/Brazil/AP-IEC-165513/2020      2020-08-29 18:52:11   \n",
       "4   hCoV-19/Brazil/PA162802-IEC/2020      2020-06-04 12:20:14   \n",
       "\n",
       "  genomeQuality_lastModified genomeLineage_id genomeLineage_accession  \\\n",
       "0        2020-06-04 12:20:14      2.62809e+07          EPI_ISL_458138   \n",
       "1        2020-08-29 18:52:11      2.62457e+07          EPI_ISL_524783   \n",
       "2        2020-06-04 12:20:14      2.62809e+07          EPI_ISL_458139   \n",
       "3        2020-08-29 18:52:11      2.63981e+07          EPI_ISL_524784   \n",
       "4        2020-06-04 12:20:14      2.63127e+07          EPI_ISL_458140   \n",
       "\n",
       "    lineage genomeLineage_lineagesVersion genomeLineage_status  \\\n",
       "0  B.1.1.33                    2020/10/30            passed_qc   \n",
       "1  B.1.1.28                    2020/10/30            passed_qc   \n",
       "2  B.1.1.33                    2020/10/30            passed_qc   \n",
       "3  B.1.1.33                    2020/10/30            passed_qc   \n",
       "4  B.1.1.28                    2020/10/30            passed_qc   \n",
       "\n",
       "  genomeLineage_note genomeLineage_createTime genomeLineage_lastModified  \\\n",
       "0                         2020-11-30 15:44:35        2020-11-30 15:44:35   \n",
       "1                         2020-11-30 15:44:32        2020-11-30 15:44:32   \n",
       "2                         2020-11-30 15:44:35        2020-11-30 15:44:35   \n",
       "3                         2020-11-30 15:44:43        2020-11-30 15:44:43   \n",
       "4                         2020-11-30 15:44:38        2020-11-30 15:44:38   \n",
       "\n",
       "  genomeLineage                              accessions  \\\n",
       "0                https://www.gisaid.org/EPI_ISL_458138;   \n",
       "1                https://www.gisaid.org/EPI_ISL_524783;   \n",
       "2                https://www.gisaid.org/EPI_ISL_458139;   \n",
       "3                https://www.gisaid.org/EPI_ISL_524784;   \n",
       "4                https://www.gisaid.org/EPI_ISL_458140;   \n",
       "\n",
       "                                gisaidId  \n",
       "0  https://www.gisaid.org/EPI_ISL_458138  \n",
       "1  https://www.gisaid.org/EPI_ISL_524783  \n",
       "2  https://www.gisaid.org/EPI_ISL_458139  \n",
       "3  https://www.gisaid.org/EPI_ISL_524784  \n",
       "4  https://www.gisaid.org/EPI_ISL_458140  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assign taxonomy ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read Organism reference dictionary\n",
    "organism_to_id = dict()\n",
    "data = pd.read_csv(\"../../reference_data/OrganismDictionary.csv\", comment='#')\n",
    "for index, row in data.iterrows():\n",
    "    organism_to_id[row['organism']] = row['taxonomyId']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'human': 'taxonomy:9606', 'homo sapiens': 'taxonomy:9606', 'mus musculus': 'taxonomy:10090', 'rhinolophus affinis': 'taxonomy:59477 ', 'rhinolophus malayanus': 'taxonomy:608659', 'mustela lutreola': 'taxonomy:9666', 'panthera tigris jacksoni': 'taxonomy:419130', 'rhinolophus sp. (bat)': 'taxonomy:49442', 'bat': 'taxonomy:49442', 'manis javanica': 'taxonomy:9974', 'manis pentadactyla': 'taxonomy:143292', 'palm civet': 'taxonomy:71116', 'canine': 'taxonomy:9608', 'canis lupus familiaris': 'taxonomy:9615', 'felis catus': 'taxonomy:9685', 'neovison vison': 'taxonomy:452646', 'mesocricetus auratus': 'taxonomy:10036', 'panthera leo': 'taxonomy:9689', 'panthera tigris': 'taxonomy:9694', 'environment': 'taxonomy:151659', 'environmental': 'taxonomy:151659'}\n"
     ]
    }
   ],
   "source": [
    "print(organism_to_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['taxonomy:9606', 'taxonomy:10090', 'taxonomy:151659',\n",
       "       'taxonomy:9666', 'taxonomy:9615', 'taxonomy:143292',\n",
       "       'taxonomy:9685', 'taxonomy:9974', 'unknown', 'taxonomy:608659',\n",
       "       'taxonomy:452646', 'taxonomy:419130', 'taxonomy:59477 ',\n",
       "       'Vero cell culture', 'taxonomy:9689', 'taxonomy:9694',\n",
       "       'taxonomy:10036'], dtype=object)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign taxonomy id to host\n",
    "df['host'] = df['host'].str.strip()\n",
    "df['hostTaxonomyId'] = df['host'].apply(lambda s: organism_to_id.get(s.lower(), s))\n",
    "df['hostTaxonomyId'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['taxonomyId'] = 'taxonomy:2697049' # SARS-CoV-2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Standardize location information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['loc0', 'loc1', 'loc2', 'loc3']] = df['location'].str.split('/', n=3, expand=True)\n",
    "# strip white space\n",
    "df = df.apply(lambda x: x.str.strip() if x.dtype == \"object\" else x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['origLocation'] = df[['loc0', 'loc1', 'loc2', 'loc3']].apply(\n",
    "    lambda x: ','.join(x.dropna().astype(str)),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save strain metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# strains = df[['name', 'accession', 'accessions', 'gisaidId', 'source', 'taxonomyId', 'hostTaxonomyId', \n",
    "#               'sequenceLength', 'sequenceQuality', 'qualityAssessment', 'collectionDate', 'location', \n",
    "#               'origLocation', 'originatingLab']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "strains = df[['name', 'accession', 'accessions', 'gisaidId', 'source', 'taxonomyId', 'hostTaxonomyId', 'lineage',\n",
    "              'sequenceLength', 'completeness', 'gender', 'age', 'collectionDate', 'location', \n",
    "              'origLocation', 'originatingLab']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>accession</th>\n",
       "      <th>accessions</th>\n",
       "      <th>gisaidId</th>\n",
       "      <th>source</th>\n",
       "      <th>taxonomyId</th>\n",
       "      <th>hostTaxonomyId</th>\n",
       "      <th>lineage</th>\n",
       "      <th>sequenceLength</th>\n",
       "      <th>completeness</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>location</th>\n",
       "      <th>origLocation</th>\n",
       "      <th>originatingLab</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCoV-19/Brazil/AP162741-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458138</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458138;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458138</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>37</td>\n",
       "      <td>2020-04-03</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>Brazil,Amapa</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Brazil/PA-IEC-164747/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524783</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524783;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524783</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Female</td>\n",
       "      <td>25</td>\n",
       "      <td>2020-04-28</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>Brazil,Para</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Brazil/AC162535-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458139</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458139;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458139</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>81</td>\n",
       "      <td>2020-03-18</td>\n",
       "      <td>Brazil / Acre</td>\n",
       "      <td>Brazil,Acre</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCoV-19/Brazil/AP-IEC-165513/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524784</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524784;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_524784</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.33</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>49</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>Brazil / Amapa</td>\n",
       "      <td>Brazil,Amapa</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCoV-19/Brazil/PA162802-IEC/2020</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458140</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458140;</td>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_458140</td>\n",
       "      <td>GISAID</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>B.1.1.28</td>\n",
       "      <td>29903</td>\n",
       "      <td>Complete</td>\n",
       "      <td>Male</td>\n",
       "      <td>63</td>\n",
       "      <td>2020-04-07</td>\n",
       "      <td>Brazil / Para</td>\n",
       "      <td>Brazil,Para</td>\n",
       "      <td>﻿Evandro Chagas Institute</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                name                              accession  \\\n",
       "0   hCoV-19/Brazil/AP162741-IEC/2020  https://www.gisaid.org/EPI_ISL_458138   \n",
       "1  hCoV-19/Brazil/PA-IEC-164747/2020  https://www.gisaid.org/EPI_ISL_524783   \n",
       "2   hCoV-19/Brazil/AC162535-IEC/2020  https://www.gisaid.org/EPI_ISL_458139   \n",
       "3  hCoV-19/Brazil/AP-IEC-165513/2020  https://www.gisaid.org/EPI_ISL_524784   \n",
       "4   hCoV-19/Brazil/PA162802-IEC/2020  https://www.gisaid.org/EPI_ISL_458140   \n",
       "\n",
       "                               accessions  \\\n",
       "0  https://www.gisaid.org/EPI_ISL_458138;   \n",
       "1  https://www.gisaid.org/EPI_ISL_524783;   \n",
       "2  https://www.gisaid.org/EPI_ISL_458139;   \n",
       "3  https://www.gisaid.org/EPI_ISL_524784;   \n",
       "4  https://www.gisaid.org/EPI_ISL_458140;   \n",
       "\n",
       "                                gisaidId  source        taxonomyId  \\\n",
       "0  https://www.gisaid.org/EPI_ISL_458138  GISAID  taxonomy:2697049   \n",
       "1  https://www.gisaid.org/EPI_ISL_524783  GISAID  taxonomy:2697049   \n",
       "2  https://www.gisaid.org/EPI_ISL_458139  GISAID  taxonomy:2697049   \n",
       "3  https://www.gisaid.org/EPI_ISL_524784  GISAID  taxonomy:2697049   \n",
       "4  https://www.gisaid.org/EPI_ISL_458140  GISAID  taxonomy:2697049   \n",
       "\n",
       "  hostTaxonomyId   lineage  sequenceLength completeness  gender age  \\\n",
       "0  taxonomy:9606  B.1.1.33           29903     Complete  Female  37   \n",
       "1  taxonomy:9606  B.1.1.28           29903     Complete  Female  25   \n",
       "2  taxonomy:9606  B.1.1.33           29903     Complete    Male  81   \n",
       "3  taxonomy:9606  B.1.1.33           29903     Complete    Male  49   \n",
       "4  taxonomy:9606  B.1.1.28           29903     Complete    Male  63   \n",
       "\n",
       "  collectionDate        location  origLocation             originatingLab  \n",
       "0     2020-04-03  Brazil / Amapa  Brazil,Amapa  ﻿Evandro Chagas Institute  \n",
       "1     2020-04-28   Brazil / Para   Brazil,Para  ﻿Evandro Chagas Institute  \n",
       "2     2020-03-18   Brazil / Acre   Brazil,Acre  ﻿Evandro Chagas Institute  \n",
       "3     2020-04-29  Brazil / Amapa  Brazil,Amapa  ﻿Evandro Chagas Institute  \n",
       "4     2020-04-07   Brazil / Para   Brazil,Para  ﻿Evandro Chagas Institute  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strains.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of strains: 294247\n"
     ]
    }
   ],
   "source": [
    "print('Number of strains:',strains.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "strains.to_csv(NEO4J_IMPORT / \"01c-CNCBStrainPre.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
