{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load SARS-CoV-2 Virus Strain Data from CNCB\n",
    "**[Work in progress]**\n",
    "\n",
    "This notebook downloads and standardizes viral strain and variation data from CNCB for ingestion into a Knowledge Graph.\n",
    "\n",
    "Data source: [China National Center for Bioinformation, 2019 Novel Coronavirus Resource (2019nCoVR)](https://bigd.big.ac.cn/ncov/release_genome)\n",
    "\n",
    "Author: Peter Rose (pwrose@ucsd.edu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import dateutil\n",
    "import re\n",
    "from pathlib import Path\n",
    "import glob\n",
    "import ftplib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_rows = None  # display all rows\n",
    "pd.options.display.max_columns = None  # display all columsns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/peter/Library/Application Support/Neo4j Desktop/Application/neo4jDatabases/database-b9d10363-6d59-4deb-9595-2cb904a99d1d/installation-4.1.0/import\n"
     ]
    }
   ],
   "source": [
    "# Path will take care of handling operating system differences.\n",
    "NEO4J_IMPORT = Path(os.getenv('NEO4J_IMPORT'))\n",
    "print(NEO4J_IMPORT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a directory to cache variation data\n",
    "CACHE = Path(NEO4J_IMPORT / 'cache')\n",
    "CACHE.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a directory to cache variation data that could not be parsed\n",
    "CACHE_FAILED = Path(NEO4J_IMPORT / 'cache_failed')\n",
    "CACHE_FAILED.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download SARS-CoV-2 Strain metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook will download > 20,000 files. To work on a small sample (50 files), set \n",
    "\n",
    "`run_small_sample_only = True`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_small_sample_only = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_url = \"https://bigd.big.ac.cn/ncov/genome/export/meta\"\n",
    "annotation_url = \"ftp://download.big.ac.cn/GVM/Coronavirus/gff3/\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(metadata_url, dtype='str')\n",
    "df.fillna('', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of strains: 96075\n"
     ]
    }
   ],
   "source": [
    "print(\"Total number of strains:\", df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.query(\"`Sequence Quality` == 'High'\")\n",
    "df = df.query(\"`Nuc.Completeness` == 'Complete'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of complete high quality strains 50907\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of complete high quality strains\", df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus Strain Name</th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Data Source</th>\n",
       "      <th>Related ID</th>\n",
       "      <th>Nuc.Completeness</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>Sequence Quality</th>\n",
       "      <th>Quality Assessment</th>\n",
       "      <th>Host</th>\n",
       "      <th>Sample Collection Date</th>\n",
       "      <th>Location</th>\n",
       "      <th>Originating Lab</th>\n",
       "      <th>Submission Date</th>\n",
       "      <th>Submitting Lab</th>\n",
       "      <th>Create Time</th>\n",
       "      <th>Last Update Time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BetaCoV/Wuhan/HBCDC-HB-01/2019</td>\n",
       "      <td>NMDC60013088-01</td>\n",
       "      <td>NMDC</td>\n",
       "      <td>EPI_ISL_402132</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29848</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/1/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>China / Hubei</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-05-07 23:03:25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Thailand/74/2020</td>\n",
       "      <td>EPI_ISL_403963</td>\n",
       "      <td>GISAID</td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>29859</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/0/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>Thailand/ Nonthaburi Province</td>\n",
       "      <td>Department of Medical Sciences, Ministry of Pu...</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>Department of Medical Sciences, Ministry of Pu...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-06-28 15:24:28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Thailand/61/2020</td>\n",
       "      <td>EPI_ISL_403962</td>\n",
       "      <td>GISAID</td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>29848</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/0/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2020-01-08</td>\n",
       "      <td>Thailand/ Nonthaburi Province</td>\n",
       "      <td>Department of Medical Sciences, Ministry of Pu...</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>Department of Medical Sciences, Ministry of Pu...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-06-28 15:24:28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BetaCoV/Wuhan/IVDC-HB-04/2020</td>\n",
       "      <td>NMDC60013085-01</td>\n",
       "      <td>NMDC</td>\n",
       "      <td>EPI_ISL_402120</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29896</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/2/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2020-01-01</td>\n",
       "      <td>China / Hubei / Wuhan</td>\n",
       "      <td>National Institute for Viral Disease Control a...</td>\n",
       "      <td>2020-01-11</td>\n",
       "      <td>National Institute for Viral Disease Control a...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-05-07 23:03:25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BetaCoV/Wuhan/IVDC-HB-01/2019</td>\n",
       "      <td>NMDC60013084-01</td>\n",
       "      <td>NMDC</td>\n",
       "      <td>EPI_ISL_402119</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29891</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/0/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>China / Hubei / Wuhan</td>\n",
       "      <td>National Institute for Viral Disease Control a...</td>\n",
       "      <td>2020-01-10</td>\n",
       "      <td>National Institute for Viral Disease Control a...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-05-07 23:03:25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Virus Strain Name     Accession ID Data Source  \\\n",
       "0  BetaCoV/Wuhan/HBCDC-HB-01/2019  NMDC60013088-01        NMDC   \n",
       "1        hCoV-19/Thailand/74/2020   EPI_ISL_403963      GISAID   \n",
       "2        hCoV-19/Thailand/61/2020   EPI_ISL_403962      GISAID   \n",
       "3   BetaCoV/Wuhan/IVDC-HB-04/2020  NMDC60013085-01        NMDC   \n",
       "4   BetaCoV/Wuhan/IVDC-HB-01/2019  NMDC60013084-01        NMDC   \n",
       "\n",
       "       Related ID Nuc.Completeness Sequence Length Sequence Quality  \\\n",
       "0  EPI_ISL_402132         Complete           29848             High   \n",
       "1                         Complete           29859             High   \n",
       "2                         Complete           29848             High   \n",
       "3  EPI_ISL_402120         Complete           29896             High   \n",
       "4  EPI_ISL_402119         Complete           29891             High   \n",
       "\n",
       "  Quality Assessment          Host Sample Collection Date  \\\n",
       "0         0/0/0/1/NO  Homo sapiens             2019-12-30   \n",
       "1         0/0/0/0/NO  Homo sapiens             2020-01-13   \n",
       "2         0/0/0/0/NO  Homo sapiens             2020-01-08   \n",
       "3         0/0/0/2/NO  Homo sapiens             2020-01-01   \n",
       "4         0/0/0/0/NO  Homo sapiens             2019-12-30   \n",
       "\n",
       "                        Location  \\\n",
       "0                  China / Hubei   \n",
       "1  Thailand/ Nonthaburi Province   \n",
       "2  Thailand/ Nonthaburi Province   \n",
       "3          China / Hubei / Wuhan   \n",
       "4          China / Hubei / Wuhan   \n",
       "\n",
       "                                     Originating Lab Submission Date  \\\n",
       "0  Hubei Provincial Center for Disease Control an...      2020-01-19   \n",
       "1  Department of Medical Sciences, Ministry of Pu...      2020-01-17   \n",
       "2  Department of Medical Sciences, Ministry of Pu...      2020-01-17   \n",
       "3  National Institute for Viral Disease Control a...      2020-01-11   \n",
       "4  National Institute for Viral Disease Control a...      2020-01-10   \n",
       "\n",
       "                                      Submitting Lab          Create Time  \\\n",
       "0  Hubei Provincial Center for Disease Control an...  2020-01-20 20:04:48   \n",
       "1  Department of Medical Sciences, Ministry of Pu...  2020-01-20 20:04:48   \n",
       "2  Department of Medical Sciences, Ministry of Pu...  2020-01-20 20:04:48   \n",
       "3  National Institute for Viral Disease Control a...  2020-01-20 20:04:48   \n",
       "4  National Institute for Viral Disease Control a...  2020-01-20 20:04:48   \n",
       "\n",
       "      Last Update Time  \n",
       "0  2020-05-07 23:03:25  \n",
       "1  2020-06-28 15:24:28  \n",
       "2  2020-06-28 15:24:28  \n",
       "3  2020-05-07 23:03:25  \n",
       "4  2020-05-07 23:03:25  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a separate row for each Accession and Related ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Accession ID'] = df['Accession ID'].str.strip()\n",
    "df['Related ID'] = df['Related ID'].str.strip()\n",
    "\n",
    "# combine all ids into a single column\n",
    "df['alias'] = df['Accession ID'] + df['Related ID'].apply(lambda s: ',' + s if len(s) > 0 else s)\n",
    "df['alias'] = df['alias'].str.replace(' ', '')\n",
    "\n",
    "# then \"explode\" ids into separate rows\n",
    "df['id'] = df['alias'].apply(lambda s: s.split(','))\n",
    "df = df.explode('id')\n",
    "df['id'] = df['id'].str.strip()\n",
    "df['alias'] = df['alias'].str.replace(',', ';')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assign taxonomy ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read Organism reference dictionary\n",
    "organism_to_id = dict()\n",
    "data = pd.read_csv(\"../../reference_data/OrganismDictionary.csv\", comment='#')\n",
    "for index, row in data.iterrows():\n",
    "    organism_to_id[row['organism']] = row['taxonomyId']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign taxonomy id to host\n",
    "df['Host'] = df['Host'].str.strip()\n",
    "df['hostTaxonomyId'] = df['Host'].apply(lambda s: organism_to_id.get(s.lower(), s))\n",
    "df['hostTaxonomyId'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['taxonomyId'] = 'taxonomy:2697049' # SARS-CoV-2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Standardize node property names (CURIEs and URIs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.rename(columns={'Virus Strain Name': 'name',\n",
    "                   'Sample Collection Date':'collectionDate',\n",
    "                   'Location':'location'}, \n",
    "          inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://registry.identifiers.org/registry/insdc\n",
    "insdc_pattern = re.compile('^([A-Z]\\d{5}|[A-Z]{2}\\d{6}|[A-Z]{4}\\d{8}|[A-J][A-Z]{2}\\d{5})(\\.\\d+)?$')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_curie(id):\n",
    "    id = id.strip()\n",
    "    # remove underscore to enable CURIE matching of NCBI reference sequences NC_...\n",
    "    id = id.replace('NC_', 'NC') \n",
    "    if len(id) > 0:\n",
    "        if id.startswith('EPI'):\n",
    "            return 'https://www.gisaid.org/' + id\n",
    "        elif id.startswith('NC_'):\n",
    "            # NCBI reference sequences resolve with ncbiprotein CURIE\n",
    "            return 'ncbiprotein:' + id\n",
    "        elif insdc_pattern.match(id) != None:\n",
    "            return 'insdc:' + id\n",
    "        else:\n",
    "            # TODO are URIs available for these cases?\n",
    "            return id\n",
    "    else:\n",
    "        return id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['loc0', 'loc1', 'loc2', 'loc3']] = df['location'].str.split('/', n=3, expand=True)\n",
    "# strip white space\n",
    "df = df.apply(lambda x: x.str.strip() if x.dtype == \"object\" else x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['origLocation'] = df[['loc0', 'loc1', 'loc2', 'loc3']].apply(\n",
    "    lambda x: ','.join(x.dropna().astype(str)),\n",
    "    axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>alias</th>\n",
       "      <th>taxonomyId</th>\n",
       "      <th>hostTaxonomyId</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>location</th>\n",
       "      <th>origLocation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NMDC60013088-01</td>\n",
       "      <td>BetaCoV/Wuhan/HBCDC-HB-01/2019</td>\n",
       "      <td>NMDC60013088-01;EPI_ISL_402132</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>China / Hubei</td>\n",
       "      <td>China,Hubei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_402132</td>\n",
       "      <td>BetaCoV/Wuhan/HBCDC-HB-01/2019</td>\n",
       "      <td>NMDC60013088-01;EPI_ISL_402132</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>China / Hubei</td>\n",
       "      <td>China,Hubei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_403963</td>\n",
       "      <td>hCoV-19/Thailand/74/2020</td>\n",
       "      <td>EPI_ISL_403963</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>Thailand/ Nonthaburi Province</td>\n",
       "      <td>Thailand,Nonthaburi Province</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_403962</td>\n",
       "      <td>hCoV-19/Thailand/61/2020</td>\n",
       "      <td>EPI_ISL_403962</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-08</td>\n",
       "      <td>Thailand/ Nonthaburi Province</td>\n",
       "      <td>Thailand,Nonthaburi Province</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NMDC60013085-01</td>\n",
       "      <td>BetaCoV/Wuhan/IVDC-HB-04/2020</td>\n",
       "      <td>NMDC60013085-01;EPI_ISL_402120</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-01</td>\n",
       "      <td>China / Hubei / Wuhan</td>\n",
       "      <td>China,Hubei,Wuhan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id                            name  \\\n",
       "0                        NMDC60013088-01  BetaCoV/Wuhan/HBCDC-HB-01/2019   \n",
       "0  https://www.gisaid.org/EPI_ISL_402132  BetaCoV/Wuhan/HBCDC-HB-01/2019   \n",
       "1  https://www.gisaid.org/EPI_ISL_403963        hCoV-19/Thailand/74/2020   \n",
       "2  https://www.gisaid.org/EPI_ISL_403962        hCoV-19/Thailand/61/2020   \n",
       "3                        NMDC60013085-01   BetaCoV/Wuhan/IVDC-HB-04/2020   \n",
       "\n",
       "                            alias        taxonomyId hostTaxonomyId  \\\n",
       "0  NMDC60013088-01;EPI_ISL_402132  taxonomy:2697049  taxonomy:9606   \n",
       "0  NMDC60013088-01;EPI_ISL_402132  taxonomy:2697049  taxonomy:9606   \n",
       "1                  EPI_ISL_403963  taxonomy:2697049  taxonomy:9606   \n",
       "2                  EPI_ISL_403962  taxonomy:2697049  taxonomy:9606   \n",
       "3  NMDC60013085-01;EPI_ISL_402120  taxonomy:2697049  taxonomy:9606   \n",
       "\n",
       "  collectionDate                       location                  origLocation  \n",
       "0     2019-12-30                  China / Hubei                   China,Hubei  \n",
       "0     2019-12-30                  China / Hubei                   China,Hubei  \n",
       "1     2020-01-13  Thailand/ Nonthaburi Province  Thailand,Nonthaburi Province  \n",
       "2     2020-01-08  Thailand/ Nonthaburi Province  Thailand,Nonthaburi Province  \n",
       "3     2020-01-01          China / Hubei / Wuhan             China,Hubei,Wuhan  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strains = df[['id', 'name', 'alias', 'taxonomyId', 'hostTaxonomyId','collectionDate', 'location', 'origLocation']].copy()\n",
    "strains['id'] = strains['id'].apply(assign_curie)\n",
    "strains.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>alias</th>\n",
       "      <th>taxonomyId</th>\n",
       "      <th>hostTaxonomyId</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>location</th>\n",
       "      <th>origLocation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_407896</td>\n",
       "      <td>hCoV-19/Australia/QLD02/2020</td>\n",
       "      <td>EPI_ISL_407896</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Australia / Queensland / Gold Coast</td>\n",
       "      <td>Australia,Queensland,Gold Coast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_412029</td>\n",
       "      <td>hCoV-19/Hong Kong/VB20024950-2/2020</td>\n",
       "      <td>EPI_ISL_412029</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Hong Kong</td>\n",
       "      <td>China,Hong Kong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_412870</td>\n",
       "      <td>hCoV-19/South Korea/KCDC06/2020</td>\n",
       "      <td>EPI_ISL_412870</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>South Korea/ Seoul</td>\n",
       "      <td>South Korea,Seoul</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_412869</td>\n",
       "      <td>hCoV-19/South Korea/KCDC05/2020</td>\n",
       "      <td>EPI_ISL_412869</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>South Korea /Seoul</td>\n",
       "      <td>South Korea,Seoul</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>414</th>\n",
       "      <td>insdc:MT066156</td>\n",
       "      <td>SARS-CoV-2/human/ITA/INMI1/2020</td>\n",
       "      <td>MT066156</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Italy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>NMDC60013027-01</td>\n",
       "      <td>hCoV-19/Guangdong/2020XN4273-P0036/2020</td>\n",
       "      <td>NMDC60013027-01;EPI_ISL_413860</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong</td>\n",
       "      <td>China,Guangdong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_413860</td>\n",
       "      <td>hCoV-19/Guangdong/2020XN4273-P0036/2020</td>\n",
       "      <td>NMDC60013027-01;EPI_ISL_413860</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong</td>\n",
       "      <td>China,Guangdong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>NMDC60013025-01</td>\n",
       "      <td>hCoV-19/Guangdong/2020XN4459-P0041/2020</td>\n",
       "      <td>NMDC60013025-01;EPI_ISL_413858</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong</td>\n",
       "      <td>China,Guangdong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_413858</td>\n",
       "      <td>hCoV-19/Guangdong/2020XN4459-P0041/2020</td>\n",
       "      <td>NMDC60013025-01;EPI_ISL_413858</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong</td>\n",
       "      <td>China,Guangdong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>NMDC60013021-01</td>\n",
       "      <td>hCoV-19/Guangdong/2020XN4475-P0042/2020</td>\n",
       "      <td>NMDC60013021-01;EPI_ISL_413854</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong</td>\n",
       "      <td>China,Guangdong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_413854</td>\n",
       "      <td>hCoV-19/Guangdong/2020XN4475-P0042/2020</td>\n",
       "      <td>NMDC60013021-01;EPI_ISL_413854</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong</td>\n",
       "      <td>China,Guangdong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>NMDC60013020-01</td>\n",
       "      <td>hCoV-19/Guangdong/2020XN4243-P0035/2020</td>\n",
       "      <td>NMDC60013020-01;EPI_ISL_413853</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong</td>\n",
       "      <td>China,Guangdong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_413853</td>\n",
       "      <td>hCoV-19/Guangdong/2020XN4243-P0035/2020</td>\n",
       "      <td>NMDC60013020-01;EPI_ISL_413853</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong</td>\n",
       "      <td>China,Guangdong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_414941</td>\n",
       "      <td>hCoV-19/Shandong/LY008-2/2020</td>\n",
       "      <td>EPI_ISL_414941</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Shandong</td>\n",
       "      <td>China,Shandong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_416326</td>\n",
       "      <td>hCoV-19/Shanghai/SH0013/2020</td>\n",
       "      <td>EPI_ISL_416326</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Shanghai</td>\n",
       "      <td>China,Shanghai</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_416332</td>\n",
       "      <td>hCoV-19/Shanghai/SH0023/2020</td>\n",
       "      <td>EPI_ISL_416332</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Shanghai</td>\n",
       "      <td>China,Shanghai</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_416333</td>\n",
       "      <td>hCoV-19/Shanghai/SH0024/2020</td>\n",
       "      <td>EPI_ISL_416333</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Shanghai</td>\n",
       "      <td>China,Shanghai</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1011</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_416352</td>\n",
       "      <td>hCoV-19/Shanghai/SH0043/2020</td>\n",
       "      <td>EPI_ISL_416352</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Shanghai</td>\n",
       "      <td>China,Shanghai</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_416365</td>\n",
       "      <td>hCoV-19/Shanghai/SH0058/2020</td>\n",
       "      <td>EPI_ISL_416365</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Shanghai</td>\n",
       "      <td>China,Shanghai</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1025</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_416366</td>\n",
       "      <td>hCoV-19/Shanghai/SH0059/2020</td>\n",
       "      <td>EPI_ISL_416366</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Shanghai</td>\n",
       "      <td>China,Shanghai</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4928</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_421241</td>\n",
       "      <td>hCoV-19/Nanchang/JX177/2020</td>\n",
       "      <td>EPI_ISL_421241</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Jiangxi / Nanchang</td>\n",
       "      <td>China,Jiangxi,Nanchang</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8408</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_425118</td>\n",
       "      <td>hCoV-19/Korea/KCDC2002/2020</td>\n",
       "      <td>EPI_ISL_425118</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>South Korea</td>\n",
       "      <td>South Korea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11076</th>\n",
       "      <td>NMDC60013105-01</td>\n",
       "      <td>hCoV-19/Guangdong/20SF1136/2020</td>\n",
       "      <td>NMDC60013105-01;EPI_ISL_428442</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong / Shanwei</td>\n",
       "      <td>China,Guangdong,Shanwei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11076</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_428442</td>\n",
       "      <td>hCoV-19/Guangdong/20SF1136/2020</td>\n",
       "      <td>NMDC60013105-01;EPI_ISL_428442</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong / Shanwei</td>\n",
       "      <td>China,Guangdong,Shanwei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11078</th>\n",
       "      <td>NMDC60013107-01</td>\n",
       "      <td>hCoV-19/Guangdong/20SF1152/2020</td>\n",
       "      <td>NMDC60013107-01;EPI_ISL_428444</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong / Huizhou</td>\n",
       "      <td>China,Guangdong,Huizhou</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11078</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_428444</td>\n",
       "      <td>hCoV-19/Guangdong/20SF1152/2020</td>\n",
       "      <td>NMDC60013107-01;EPI_ISL_428444</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong / Huizhou</td>\n",
       "      <td>China,Guangdong,Huizhou</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11079</th>\n",
       "      <td>NMDC60013108-01</td>\n",
       "      <td>hCoV-19/Guangdong/20SF1153/2020</td>\n",
       "      <td>NMDC60013108-01;EPI_ISL_428445</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong / Huizhou</td>\n",
       "      <td>China,Guangdong,Huizhou</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11079</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_428445</td>\n",
       "      <td>hCoV-19/Guangdong/20SF1153/2020</td>\n",
       "      <td>NMDC60013108-01;EPI_ISL_428445</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Guangdong / Huizhou</td>\n",
       "      <td>China,Guangdong,Huizhou</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16455</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_416331</td>\n",
       "      <td>hCoV-19/China/Shanghai/SH0022</td>\n",
       "      <td>EPI_ISL_416331</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Shanghai</td>\n",
       "      <td>China,Shanghai</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16902</th>\n",
       "      <td>insdc:MT365029</td>\n",
       "      <td>SARS-CoV-2/human/HKG/HKU-911a/2020</td>\n",
       "      <td>MT365029;EPI_ISL_450409</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Hong Kong</td>\n",
       "      <td>China,Hong Kong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16902</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_450409</td>\n",
       "      <td>SARS-CoV-2/human/HKG/HKU-911a/2020</td>\n",
       "      <td>MT365029;EPI_ISL_450409</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Hong Kong</td>\n",
       "      <td>China,Hong Kong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16905</th>\n",
       "      <td>insdc:MT365032</td>\n",
       "      <td>SARS-CoV-2/human/HKG/HKU-904a/2020</td>\n",
       "      <td>MT365032;EPI_ISL_450412</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Hong Kong</td>\n",
       "      <td>China,Hong Kong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16905</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_450412</td>\n",
       "      <td>SARS-CoV-2/human/HKG/HKU-904a/2020</td>\n",
       "      <td>MT365032;EPI_ISL_450412</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Hong Kong</td>\n",
       "      <td>China,Hong Kong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30551</th>\n",
       "      <td>insdc:MT270112</td>\n",
       "      <td>SARS-CoV-2/human/DEU/BavPat2-ChVir984-ChVir101...</td>\n",
       "      <td>MT270112;EPI_ISL_450209</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Germany / Bavaria</td>\n",
       "      <td>Germany,Bavaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30551</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_450209</td>\n",
       "      <td>SARS-CoV-2/human/DEU/BavPat2-ChVir984-ChVir101...</td>\n",
       "      <td>MT270112;EPI_ISL_450209</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Germany / Bavaria</td>\n",
       "      <td>Germany,Bavaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31806</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451302</td>\n",
       "      <td>hCoV-19/Italy/INMI2-N/2020</td>\n",
       "      <td>EPI_ISL_451302</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Italy / Lazio</td>\n",
       "      <td>Italy,Lazio</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31818</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451314</td>\n",
       "      <td>hCoV-19/China/SC-GA-063/2020</td>\n",
       "      <td>EPI_ISL_451314</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31819</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451315</td>\n",
       "      <td>hCoV-19/China/SC-GA-064/2020</td>\n",
       "      <td>EPI_ISL_451315</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31836</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451332</td>\n",
       "      <td>hCoV-19/China/SC-NC-073/2020</td>\n",
       "      <td>EPI_ISL_451332</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31838</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451334</td>\n",
       "      <td>hCoV-19/China/SC-NC-075/2020</td>\n",
       "      <td>EPI_ISL_451334</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31841</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451337</td>\n",
       "      <td>hCoV-19/China/SC-NC-078/2020</td>\n",
       "      <td>EPI_ISL_451337</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31847</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451343</td>\n",
       "      <td>hCoV-19/China/SC-NJ-094/2020</td>\n",
       "      <td>EPI_ISL_451343</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31892</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451389</td>\n",
       "      <td>hCoV-19/China/SC-WCH4-287/2020</td>\n",
       "      <td>EPI_ISL_451389</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31898</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451395</td>\n",
       "      <td>hCoV-19/China/SC-WCH4-301/2020</td>\n",
       "      <td>EPI_ISL_451395</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31899</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_451396</td>\n",
       "      <td>hCoV-19/China/SC-YA-087/2020</td>\n",
       "      <td>EPI_ISL_451396</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Sichuan</td>\n",
       "      <td>China,Sichuan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35951</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_455603</td>\n",
       "      <td>hCoV-19/Australia/VPRL014/2020</td>\n",
       "      <td>EPI_ISL_455603</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Australia / South Australia</td>\n",
       "      <td>Australia,South Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36099</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_455790</td>\n",
       "      <td>hCoV-19/Malaysia/IMR_WC119/2020</td>\n",
       "      <td>EPI_ISL_455790</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Malaysia / Selangor</td>\n",
       "      <td>Malaysia,Selangor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43499</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_463894</td>\n",
       "      <td>hCoV-19/Shaoxing/02/2020</td>\n",
       "      <td>EPI_ISL_463894</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Zhejiang / Shaoxing</td>\n",
       "      <td>China,Zhejiang,Shaoxing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43500</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_463895</td>\n",
       "      <td>hCoV-19/Shaoxing/03/2020</td>\n",
       "      <td>EPI_ISL_463895</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>China / Zhejiang / Shaoxing</td>\n",
       "      <td>China,Zhejiang,Shaoxing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47613</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_468032</td>\n",
       "      <td>hCoV-19/Australia/SAP214/2020</td>\n",
       "      <td>EPI_ISL_468032</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Australia / South Australia</td>\n",
       "      <td>Australia,South Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47618</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_468037</td>\n",
       "      <td>hCoV-19/Australia/SAP305/2020</td>\n",
       "      <td>EPI_ISL_468037</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Australia / South Australia</td>\n",
       "      <td>Australia,South Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59342</th>\n",
       "      <td>insdc:MT745747</td>\n",
       "      <td>SARS-CoV-2/human/AUS/SA02/2020</td>\n",
       "      <td>MT745747;EPI_ISL_480779</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Australia / South Australia</td>\n",
       "      <td>Australia,South Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59342</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_480779</td>\n",
       "      <td>SARS-CoV-2/human/AUS/SA02/2020</td>\n",
       "      <td>MT745747;EPI_ISL_480779</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Australia / South Australia</td>\n",
       "      <td>Australia,South Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59343</th>\n",
       "      <td>insdc:MT745748</td>\n",
       "      <td>SARS-CoV-2/human/AUS/SA03/2020</td>\n",
       "      <td>MT745748;EPI_ISL_480780</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Australia / South Australia</td>\n",
       "      <td>Australia,South Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59343</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_480780</td>\n",
       "      <td>SARS-CoV-2/human/AUS/SA03/2020</td>\n",
       "      <td>MT745748;EPI_ISL_480780</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>Australia / South Australia</td>\n",
       "      <td>Australia,South Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74110</th>\n",
       "      <td>https://www.gisaid.org/EPI_ISL_497962</td>\n",
       "      <td>hCoV-19/Korea/KCDC2061/2020</td>\n",
       "      <td>EPI_ISL_497962</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>2020-01-30</td>\n",
       "      <td>South Korea</td>\n",
       "      <td>South Korea</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          id  \\\n",
       "88     https://www.gisaid.org/EPI_ISL_407896   \n",
       "215    https://www.gisaid.org/EPI_ISL_412029   \n",
       "235    https://www.gisaid.org/EPI_ISL_412870   \n",
       "236    https://www.gisaid.org/EPI_ISL_412869   \n",
       "414                           insdc:MT066156   \n",
       "460                          NMDC60013027-01   \n",
       "460    https://www.gisaid.org/EPI_ISL_413860   \n",
       "462                          NMDC60013025-01   \n",
       "462    https://www.gisaid.org/EPI_ISL_413858   \n",
       "467                          NMDC60013021-01   \n",
       "467    https://www.gisaid.org/EPI_ISL_413854   \n",
       "468                          NMDC60013020-01   \n",
       "468    https://www.gisaid.org/EPI_ISL_413853   \n",
       "712    https://www.gisaid.org/EPI_ISL_414941   \n",
       "986    https://www.gisaid.org/EPI_ISL_416326   \n",
       "991    https://www.gisaid.org/EPI_ISL_416332   \n",
       "992    https://www.gisaid.org/EPI_ISL_416333   \n",
       "1011   https://www.gisaid.org/EPI_ISL_416352   \n",
       "1024   https://www.gisaid.org/EPI_ISL_416365   \n",
       "1025   https://www.gisaid.org/EPI_ISL_416366   \n",
       "4928   https://www.gisaid.org/EPI_ISL_421241   \n",
       "8408   https://www.gisaid.org/EPI_ISL_425118   \n",
       "11076                        NMDC60013105-01   \n",
       "11076  https://www.gisaid.org/EPI_ISL_428442   \n",
       "11078                        NMDC60013107-01   \n",
       "11078  https://www.gisaid.org/EPI_ISL_428444   \n",
       "11079                        NMDC60013108-01   \n",
       "11079  https://www.gisaid.org/EPI_ISL_428445   \n",
       "16455  https://www.gisaid.org/EPI_ISL_416331   \n",
       "16902                         insdc:MT365029   \n",
       "16902  https://www.gisaid.org/EPI_ISL_450409   \n",
       "16905                         insdc:MT365032   \n",
       "16905  https://www.gisaid.org/EPI_ISL_450412   \n",
       "30551                         insdc:MT270112   \n",
       "30551  https://www.gisaid.org/EPI_ISL_450209   \n",
       "31806  https://www.gisaid.org/EPI_ISL_451302   \n",
       "31818  https://www.gisaid.org/EPI_ISL_451314   \n",
       "31819  https://www.gisaid.org/EPI_ISL_451315   \n",
       "31836  https://www.gisaid.org/EPI_ISL_451332   \n",
       "31838  https://www.gisaid.org/EPI_ISL_451334   \n",
       "31841  https://www.gisaid.org/EPI_ISL_451337   \n",
       "31847  https://www.gisaid.org/EPI_ISL_451343   \n",
       "31892  https://www.gisaid.org/EPI_ISL_451389   \n",
       "31898  https://www.gisaid.org/EPI_ISL_451395   \n",
       "31899  https://www.gisaid.org/EPI_ISL_451396   \n",
       "35951  https://www.gisaid.org/EPI_ISL_455603   \n",
       "36099  https://www.gisaid.org/EPI_ISL_455790   \n",
       "43499  https://www.gisaid.org/EPI_ISL_463894   \n",
       "43500  https://www.gisaid.org/EPI_ISL_463895   \n",
       "47613  https://www.gisaid.org/EPI_ISL_468032   \n",
       "47618  https://www.gisaid.org/EPI_ISL_468037   \n",
       "59342                         insdc:MT745747   \n",
       "59342  https://www.gisaid.org/EPI_ISL_480779   \n",
       "59343                         insdc:MT745748   \n",
       "59343  https://www.gisaid.org/EPI_ISL_480780   \n",
       "74110  https://www.gisaid.org/EPI_ISL_497962   \n",
       "\n",
       "                                                    name  \\\n",
       "88                          hCoV-19/Australia/QLD02/2020   \n",
       "215                  hCoV-19/Hong Kong/VB20024950-2/2020   \n",
       "235                      hCoV-19/South Korea/KCDC06/2020   \n",
       "236                      hCoV-19/South Korea/KCDC05/2020   \n",
       "414                      SARS-CoV-2/human/ITA/INMI1/2020   \n",
       "460              hCoV-19/Guangdong/2020XN4273-P0036/2020   \n",
       "460              hCoV-19/Guangdong/2020XN4273-P0036/2020   \n",
       "462              hCoV-19/Guangdong/2020XN4459-P0041/2020   \n",
       "462              hCoV-19/Guangdong/2020XN4459-P0041/2020   \n",
       "467              hCoV-19/Guangdong/2020XN4475-P0042/2020   \n",
       "467              hCoV-19/Guangdong/2020XN4475-P0042/2020   \n",
       "468              hCoV-19/Guangdong/2020XN4243-P0035/2020   \n",
       "468              hCoV-19/Guangdong/2020XN4243-P0035/2020   \n",
       "712                        hCoV-19/Shandong/LY008-2/2020   \n",
       "986                         hCoV-19/Shanghai/SH0013/2020   \n",
       "991                         hCoV-19/Shanghai/SH0023/2020   \n",
       "992                         hCoV-19/Shanghai/SH0024/2020   \n",
       "1011                        hCoV-19/Shanghai/SH0043/2020   \n",
       "1024                        hCoV-19/Shanghai/SH0058/2020   \n",
       "1025                        hCoV-19/Shanghai/SH0059/2020   \n",
       "4928                         hCoV-19/Nanchang/JX177/2020   \n",
       "8408                         hCoV-19/Korea/KCDC2002/2020   \n",
       "11076                    hCoV-19/Guangdong/20SF1136/2020   \n",
       "11076                    hCoV-19/Guangdong/20SF1136/2020   \n",
       "11078                    hCoV-19/Guangdong/20SF1152/2020   \n",
       "11078                    hCoV-19/Guangdong/20SF1152/2020   \n",
       "11079                    hCoV-19/Guangdong/20SF1153/2020   \n",
       "11079                    hCoV-19/Guangdong/20SF1153/2020   \n",
       "16455                      hCoV-19/China/Shanghai/SH0022   \n",
       "16902                 SARS-CoV-2/human/HKG/HKU-911a/2020   \n",
       "16902                 SARS-CoV-2/human/HKG/HKU-911a/2020   \n",
       "16905                 SARS-CoV-2/human/HKG/HKU-904a/2020   \n",
       "16905                 SARS-CoV-2/human/HKG/HKU-904a/2020   \n",
       "30551  SARS-CoV-2/human/DEU/BavPat2-ChVir984-ChVir101...   \n",
       "30551  SARS-CoV-2/human/DEU/BavPat2-ChVir984-ChVir101...   \n",
       "31806                         hCoV-19/Italy/INMI2-N/2020   \n",
       "31818                       hCoV-19/China/SC-GA-063/2020   \n",
       "31819                       hCoV-19/China/SC-GA-064/2020   \n",
       "31836                       hCoV-19/China/SC-NC-073/2020   \n",
       "31838                       hCoV-19/China/SC-NC-075/2020   \n",
       "31841                       hCoV-19/China/SC-NC-078/2020   \n",
       "31847                       hCoV-19/China/SC-NJ-094/2020   \n",
       "31892                     hCoV-19/China/SC-WCH4-287/2020   \n",
       "31898                     hCoV-19/China/SC-WCH4-301/2020   \n",
       "31899                       hCoV-19/China/SC-YA-087/2020   \n",
       "35951                     hCoV-19/Australia/VPRL014/2020   \n",
       "36099                    hCoV-19/Malaysia/IMR_WC119/2020   \n",
       "43499                           hCoV-19/Shaoxing/02/2020   \n",
       "43500                           hCoV-19/Shaoxing/03/2020   \n",
       "47613                      hCoV-19/Australia/SAP214/2020   \n",
       "47618                      hCoV-19/Australia/SAP305/2020   \n",
       "59342                     SARS-CoV-2/human/AUS/SA02/2020   \n",
       "59342                     SARS-CoV-2/human/AUS/SA02/2020   \n",
       "59343                     SARS-CoV-2/human/AUS/SA03/2020   \n",
       "59343                     SARS-CoV-2/human/AUS/SA03/2020   \n",
       "74110                        hCoV-19/Korea/KCDC2061/2020   \n",
       "\n",
       "                                alias        taxonomyId hostTaxonomyId  \\\n",
       "88                     EPI_ISL_407896  taxonomy:2697049  taxonomy:9606   \n",
       "215                    EPI_ISL_412029  taxonomy:2697049  taxonomy:9606   \n",
       "235                    EPI_ISL_412870  taxonomy:2697049  taxonomy:9606   \n",
       "236                    EPI_ISL_412869  taxonomy:2697049  taxonomy:9606   \n",
       "414                          MT066156  taxonomy:2697049  taxonomy:9606   \n",
       "460    NMDC60013027-01;EPI_ISL_413860  taxonomy:2697049  taxonomy:9606   \n",
       "460    NMDC60013027-01;EPI_ISL_413860  taxonomy:2697049  taxonomy:9606   \n",
       "462    NMDC60013025-01;EPI_ISL_413858  taxonomy:2697049  taxonomy:9606   \n",
       "462    NMDC60013025-01;EPI_ISL_413858  taxonomy:2697049  taxonomy:9606   \n",
       "467    NMDC60013021-01;EPI_ISL_413854  taxonomy:2697049  taxonomy:9606   \n",
       "467    NMDC60013021-01;EPI_ISL_413854  taxonomy:2697049  taxonomy:9606   \n",
       "468    NMDC60013020-01;EPI_ISL_413853  taxonomy:2697049  taxonomy:9606   \n",
       "468    NMDC60013020-01;EPI_ISL_413853  taxonomy:2697049  taxonomy:9606   \n",
       "712                    EPI_ISL_414941  taxonomy:2697049  taxonomy:9606   \n",
       "986                    EPI_ISL_416326  taxonomy:2697049  taxonomy:9606   \n",
       "991                    EPI_ISL_416332  taxonomy:2697049  taxonomy:9606   \n",
       "992                    EPI_ISL_416333  taxonomy:2697049  taxonomy:9606   \n",
       "1011                   EPI_ISL_416352  taxonomy:2697049  taxonomy:9606   \n",
       "1024                   EPI_ISL_416365  taxonomy:2697049  taxonomy:9606   \n",
       "1025                   EPI_ISL_416366  taxonomy:2697049  taxonomy:9606   \n",
       "4928                   EPI_ISL_421241  taxonomy:2697049  taxonomy:9606   \n",
       "8408                   EPI_ISL_425118  taxonomy:2697049  taxonomy:9606   \n",
       "11076  NMDC60013105-01;EPI_ISL_428442  taxonomy:2697049  taxonomy:9606   \n",
       "11076  NMDC60013105-01;EPI_ISL_428442  taxonomy:2697049  taxonomy:9606   \n",
       "11078  NMDC60013107-01;EPI_ISL_428444  taxonomy:2697049  taxonomy:9606   \n",
       "11078  NMDC60013107-01;EPI_ISL_428444  taxonomy:2697049  taxonomy:9606   \n",
       "11079  NMDC60013108-01;EPI_ISL_428445  taxonomy:2697049  taxonomy:9606   \n",
       "11079  NMDC60013108-01;EPI_ISL_428445  taxonomy:2697049  taxonomy:9606   \n",
       "16455                  EPI_ISL_416331  taxonomy:2697049  taxonomy:9606   \n",
       "16902         MT365029;EPI_ISL_450409  taxonomy:2697049  taxonomy:9606   \n",
       "16902         MT365029;EPI_ISL_450409  taxonomy:2697049  taxonomy:9606   \n",
       "16905         MT365032;EPI_ISL_450412  taxonomy:2697049  taxonomy:9606   \n",
       "16905         MT365032;EPI_ISL_450412  taxonomy:2697049  taxonomy:9606   \n",
       "30551         MT270112;EPI_ISL_450209  taxonomy:2697049  taxonomy:9606   \n",
       "30551         MT270112;EPI_ISL_450209  taxonomy:2697049  taxonomy:9606   \n",
       "31806                  EPI_ISL_451302  taxonomy:2697049  taxonomy:9606   \n",
       "31818                  EPI_ISL_451314  taxonomy:2697049  taxonomy:9606   \n",
       "31819                  EPI_ISL_451315  taxonomy:2697049  taxonomy:9606   \n",
       "31836                  EPI_ISL_451332  taxonomy:2697049  taxonomy:9606   \n",
       "31838                  EPI_ISL_451334  taxonomy:2697049  taxonomy:9606   \n",
       "31841                  EPI_ISL_451337  taxonomy:2697049  taxonomy:9606   \n",
       "31847                  EPI_ISL_451343  taxonomy:2697049  taxonomy:9606   \n",
       "31892                  EPI_ISL_451389  taxonomy:2697049  taxonomy:9606   \n",
       "31898                  EPI_ISL_451395  taxonomy:2697049  taxonomy:9606   \n",
       "31899                  EPI_ISL_451396  taxonomy:2697049  taxonomy:9606   \n",
       "35951                  EPI_ISL_455603  taxonomy:2697049  taxonomy:9606   \n",
       "36099                  EPI_ISL_455790  taxonomy:2697049  taxonomy:9606   \n",
       "43499                  EPI_ISL_463894  taxonomy:2697049  taxonomy:9606   \n",
       "43500                  EPI_ISL_463895  taxonomy:2697049  taxonomy:9606   \n",
       "47613                  EPI_ISL_468032  taxonomy:2697049  taxonomy:9606   \n",
       "47618                  EPI_ISL_468037  taxonomy:2697049  taxonomy:9606   \n",
       "59342         MT745747;EPI_ISL_480779  taxonomy:2697049  taxonomy:9606   \n",
       "59342         MT745747;EPI_ISL_480779  taxonomy:2697049  taxonomy:9606   \n",
       "59343         MT745748;EPI_ISL_480780  taxonomy:2697049  taxonomy:9606   \n",
       "59343         MT745748;EPI_ISL_480780  taxonomy:2697049  taxonomy:9606   \n",
       "74110                  EPI_ISL_497962  taxonomy:2697049  taxonomy:9606   \n",
       "\n",
       "      collectionDate                             location  \\\n",
       "88        2020-01-30  Australia / Queensland / Gold Coast   \n",
       "215       2020-01-30                    China / Hong Kong   \n",
       "235       2020-01-30                   South Korea/ Seoul   \n",
       "236       2020-01-30                   South Korea /Seoul   \n",
       "414       2020-01-30                                Italy   \n",
       "460       2020-01-30                    China / Guangdong   \n",
       "460       2020-01-30                    China / Guangdong   \n",
       "462       2020-01-30                    China / Guangdong   \n",
       "462       2020-01-30                    China / Guangdong   \n",
       "467       2020-01-30                    China / Guangdong   \n",
       "467       2020-01-30                    China / Guangdong   \n",
       "468       2020-01-30                    China / Guangdong   \n",
       "468       2020-01-30                    China / Guangdong   \n",
       "712       2020-01-30                     China / Shandong   \n",
       "986       2020-01-30                     China / Shanghai   \n",
       "991       2020-01-30                     China / Shanghai   \n",
       "992       2020-01-30                     China / Shanghai   \n",
       "1011      2020-01-30                     China / Shanghai   \n",
       "1024      2020-01-30                     China / Shanghai   \n",
       "1025      2020-01-30                     China / Shanghai   \n",
       "4928      2020-01-30           China / Jiangxi / Nanchang   \n",
       "8408      2020-01-30                          South Korea   \n",
       "11076     2020-01-30          China / Guangdong / Shanwei   \n",
       "11076     2020-01-30          China / Guangdong / Shanwei   \n",
       "11078     2020-01-30          China / Guangdong / Huizhou   \n",
       "11078     2020-01-30          China / Guangdong / Huizhou   \n",
       "11079     2020-01-30          China / Guangdong / Huizhou   \n",
       "11079     2020-01-30          China / Guangdong / Huizhou   \n",
       "16455     2020-01-30                     China / Shanghai   \n",
       "16902     2020-01-30                    China / Hong Kong   \n",
       "16902     2020-01-30                    China / Hong Kong   \n",
       "16905     2020-01-30                    China / Hong Kong   \n",
       "16905     2020-01-30                    China / Hong Kong   \n",
       "30551     2020-01-30                    Germany / Bavaria   \n",
       "30551     2020-01-30                    Germany / Bavaria   \n",
       "31806     2020-01-30                        Italy / Lazio   \n",
       "31818     2020-01-30                      China / Sichuan   \n",
       "31819     2020-01-30                      China / Sichuan   \n",
       "31836     2020-01-30                      China / Sichuan   \n",
       "31838     2020-01-30                      China / Sichuan   \n",
       "31841     2020-01-30                      China / Sichuan   \n",
       "31847     2020-01-30                      China / Sichuan   \n",
       "31892     2020-01-30                      China / Sichuan   \n",
       "31898     2020-01-30                      China / Sichuan   \n",
       "31899     2020-01-30                      China / Sichuan   \n",
       "35951     2020-01-30          Australia / South Australia   \n",
       "36099     2020-01-30                  Malaysia / Selangor   \n",
       "43499     2020-01-30          China / Zhejiang / Shaoxing   \n",
       "43500     2020-01-30          China / Zhejiang / Shaoxing   \n",
       "47613     2020-01-30          Australia / South Australia   \n",
       "47618     2020-01-30          Australia / South Australia   \n",
       "59342     2020-01-30          Australia / South Australia   \n",
       "59342     2020-01-30          Australia / South Australia   \n",
       "59343     2020-01-30          Australia / South Australia   \n",
       "59343     2020-01-30          Australia / South Australia   \n",
       "74110     2020-01-30                          South Korea   \n",
       "\n",
       "                          origLocation  \n",
       "88     Australia,Queensland,Gold Coast  \n",
       "215                    China,Hong Kong  \n",
       "235                  South Korea,Seoul  \n",
       "236                  South Korea,Seoul  \n",
       "414                              Italy  \n",
       "460                    China,Guangdong  \n",
       "460                    China,Guangdong  \n",
       "462                    China,Guangdong  \n",
       "462                    China,Guangdong  \n",
       "467                    China,Guangdong  \n",
       "467                    China,Guangdong  \n",
       "468                    China,Guangdong  \n",
       "468                    China,Guangdong  \n",
       "712                     China,Shandong  \n",
       "986                     China,Shanghai  \n",
       "991                     China,Shanghai  \n",
       "992                     China,Shanghai  \n",
       "1011                    China,Shanghai  \n",
       "1024                    China,Shanghai  \n",
       "1025                    China,Shanghai  \n",
       "4928            China,Jiangxi,Nanchang  \n",
       "8408                       South Korea  \n",
       "11076          China,Guangdong,Shanwei  \n",
       "11076          China,Guangdong,Shanwei  \n",
       "11078          China,Guangdong,Huizhou  \n",
       "11078          China,Guangdong,Huizhou  \n",
       "11079          China,Guangdong,Huizhou  \n",
       "11079          China,Guangdong,Huizhou  \n",
       "16455                   China,Shanghai  \n",
       "16902                  China,Hong Kong  \n",
       "16902                  China,Hong Kong  \n",
       "16905                  China,Hong Kong  \n",
       "16905                  China,Hong Kong  \n",
       "30551                  Germany,Bavaria  \n",
       "30551                  Germany,Bavaria  \n",
       "31806                      Italy,Lazio  \n",
       "31818                    China,Sichuan  \n",
       "31819                    China,Sichuan  \n",
       "31836                    China,Sichuan  \n",
       "31838                    China,Sichuan  \n",
       "31841                    China,Sichuan  \n",
       "31847                    China,Sichuan  \n",
       "31892                    China,Sichuan  \n",
       "31898                    China,Sichuan  \n",
       "31899                    China,Sichuan  \n",
       "35951        Australia,South Australia  \n",
       "36099                Malaysia,Selangor  \n",
       "43499          China,Zhejiang,Shaoxing  \n",
       "43500          China,Zhejiang,Shaoxing  \n",
       "47613        Australia,South Australia  \n",
       "47618        Australia,South Australia  \n",
       "59342        Australia,South Australia  \n",
       "59342        Australia,South Australia  \n",
       "59343        Australia,South Australia  \n",
       "59343        Australia,South Australia  \n",
       "74110                      South Korea  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strains.query(\"collectionDate == '2020-01-30'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "strains.to_csv(NEO4J_IMPORT / \"01d-CNCBStrain.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge Metadata with Variation Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get list of file names from FTP site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'221 Goodbye.'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server = \"download.big.ac.cn\"\n",
    "user = \"anonymous\"\n",
    "password = \"anonymous\"\n",
    "source = \"/GVM/Coronavirus/gff3/\"\n",
    "\n",
    "ftp = ftplib.FTP(server)\n",
    "ftp.login(user, password)\n",
    "ftp.cwd(source) \n",
    "filelist=ftp.nlst()\n",
    "ftp.quit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_file = pd.DataFrame(filelist, columns=['filename'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract identifiers from file name\n",
    "\n",
    "Example: 2019-nCoV_CNA0013697_variants.gff3 -> CNA0013697"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_file['id'] = df_file['filename'].str[10:]\n",
    "df_file['id'] = df_file['id'].str.replace('_variants.gff3','')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_file = df_file.query(\"id != ''\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of available files: 43946\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of available files:\", df_file.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2019-nCoV_CNA0007332_variants.gff3</td>\n",
       "      <td>CNA0007332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2019-nCoV_CNA0007334_variants.gff3</td>\n",
       "      <td>CNA0007334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2019-nCoV_CNA0007335_variants.gff3</td>\n",
       "      <td>CNA0007335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2019-nCoV_CNA0013697_variants.gff3</td>\n",
       "      <td>CNA0013697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2019-nCoV_CNA0013698_variants.gff3</td>\n",
       "      <td>CNA0013698</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             filename          id\n",
       "0  2019-nCoV_CNA0007332_variants.gff3  CNA0007332\n",
       "1  2019-nCoV_CNA0007334_variants.gff3  CNA0007334\n",
       "2  2019-nCoV_CNA0007335_variants.gff3  CNA0007335\n",
       "3  2019-nCoV_CNA0013697_variants.gff3  CNA0013697\n",
       "4  2019-nCoV_CNA0013698_variants.gff3  CNA0013698"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_file.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.merge(df_file, on='id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Strains with a matching filename: 43804\n"
     ]
    }
   ],
   "source": [
    "print('Strains with a matching filename:', df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "if run_small_sample_only:\n",
    "    df = df.sample(n=50, random_state=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>Accession ID</th>\n",
       "      <th>Data Source</th>\n",
       "      <th>Related ID</th>\n",
       "      <th>Nuc.Completeness</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>Sequence Quality</th>\n",
       "      <th>Quality Assessment</th>\n",
       "      <th>Host</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>location</th>\n",
       "      <th>Originating Lab</th>\n",
       "      <th>Submission Date</th>\n",
       "      <th>Submitting Lab</th>\n",
       "      <th>Create Time</th>\n",
       "      <th>Last Update Time</th>\n",
       "      <th>alias</th>\n",
       "      <th>id</th>\n",
       "      <th>hostTaxonomyId</th>\n",
       "      <th>taxonomyId</th>\n",
       "      <th>loc1</th>\n",
       "      <th>loc2</th>\n",
       "      <th>loc3</th>\n",
       "      <th>loc4</th>\n",
       "      <th>origLocation</th>\n",
       "      <th>filename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BetaCoV/Wuhan/HBCDC-HB-01/2019</td>\n",
       "      <td>NMDC60013088-01</td>\n",
       "      <td>NMDC</td>\n",
       "      <td>EPI_ISL_402132</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29848</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/1/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>China / Hubei</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>2020-01-19</td>\n",
       "      <td>Hubei Provincial Center for Disease Control an...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-05-07 23:03:25</td>\n",
       "      <td>NMDC60013088-01;EPI_ISL_402132</td>\n",
       "      <td>EPI_ISL_402132</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>China,Hubei</td>\n",
       "      <td>2019-nCoV_EPI_ISL_402132_variants.gff3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCoV-19/Thailand/74/2020</td>\n",
       "      <td>EPI_ISL_403963</td>\n",
       "      <td>GISAID</td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>29859</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/0/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2020-01-13</td>\n",
       "      <td>Thailand/ Nonthaburi Province</td>\n",
       "      <td>Department of Medical Sciences, Ministry of Pu...</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>Department of Medical Sciences, Ministry of Pu...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-06-28 15:24:28</td>\n",
       "      <td>EPI_ISL_403963</td>\n",
       "      <td>EPI_ISL_403963</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>Thailand</td>\n",
       "      <td>Nonthaburi Province</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Thailand,Nonthaburi Province</td>\n",
       "      <td>2019-nCoV_EPI_ISL_403963_variants.gff3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCoV-19/Thailand/61/2020</td>\n",
       "      <td>EPI_ISL_403962</td>\n",
       "      <td>GISAID</td>\n",
       "      <td></td>\n",
       "      <td>Complete</td>\n",
       "      <td>29848</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/0/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2020-01-08</td>\n",
       "      <td>Thailand/ Nonthaburi Province</td>\n",
       "      <td>Department of Medical Sciences, Ministry of Pu...</td>\n",
       "      <td>2020-01-17</td>\n",
       "      <td>Department of Medical Sciences, Ministry of Pu...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-06-28 15:24:28</td>\n",
       "      <td>EPI_ISL_403962</td>\n",
       "      <td>EPI_ISL_403962</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>Thailand</td>\n",
       "      <td>Nonthaburi Province</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Thailand,Nonthaburi Province</td>\n",
       "      <td>2019-nCoV_EPI_ISL_403962_variants.gff3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BetaCoV/Wuhan/IVDC-HB-04/2020</td>\n",
       "      <td>NMDC60013085-01</td>\n",
       "      <td>NMDC</td>\n",
       "      <td>EPI_ISL_402120</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29896</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/2/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2020-01-01</td>\n",
       "      <td>China / Hubei / Wuhan</td>\n",
       "      <td>National Institute for Viral Disease Control a...</td>\n",
       "      <td>2020-01-11</td>\n",
       "      <td>National Institute for Viral Disease Control a...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-05-07 23:03:25</td>\n",
       "      <td>NMDC60013085-01;EPI_ISL_402120</td>\n",
       "      <td>EPI_ISL_402120</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>None</td>\n",
       "      <td>China,Hubei,Wuhan</td>\n",
       "      <td>2019-nCoV_EPI_ISL_402120_variants.gff3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BetaCoV/Wuhan/IVDC-HB-01/2019</td>\n",
       "      <td>NMDC60013084-01</td>\n",
       "      <td>NMDC</td>\n",
       "      <td>EPI_ISL_402119</td>\n",
       "      <td>Complete</td>\n",
       "      <td>29891</td>\n",
       "      <td>High</td>\n",
       "      <td>0/0/0/0/NO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>China / Hubei / Wuhan</td>\n",
       "      <td>National Institute for Viral Disease Control a...</td>\n",
       "      <td>2020-01-10</td>\n",
       "      <td>National Institute for Viral Disease Control a...</td>\n",
       "      <td>2020-01-20 20:04:48</td>\n",
       "      <td>2020-05-07 23:03:25</td>\n",
       "      <td>NMDC60013084-01;EPI_ISL_402119</td>\n",
       "      <td>EPI_ISL_402119</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>China</td>\n",
       "      <td>Hubei</td>\n",
       "      <td>Wuhan</td>\n",
       "      <td>None</td>\n",
       "      <td>China,Hubei,Wuhan</td>\n",
       "      <td>2019-nCoV_EPI_ISL_402119_variants.gff3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             name     Accession ID Data Source  \\\n",
       "0  BetaCoV/Wuhan/HBCDC-HB-01/2019  NMDC60013088-01        NMDC   \n",
       "1        hCoV-19/Thailand/74/2020   EPI_ISL_403963      GISAID   \n",
       "2        hCoV-19/Thailand/61/2020   EPI_ISL_403962      GISAID   \n",
       "3   BetaCoV/Wuhan/IVDC-HB-04/2020  NMDC60013085-01        NMDC   \n",
       "4   BetaCoV/Wuhan/IVDC-HB-01/2019  NMDC60013084-01        NMDC   \n",
       "\n",
       "       Related ID Nuc.Completeness Sequence Length Sequence Quality  \\\n",
       "0  EPI_ISL_402132         Complete           29848             High   \n",
       "1                         Complete           29859             High   \n",
       "2                         Complete           29848             High   \n",
       "3  EPI_ISL_402120         Complete           29896             High   \n",
       "4  EPI_ISL_402119         Complete           29891             High   \n",
       "\n",
       "  Quality Assessment          Host collectionDate  \\\n",
       "0         0/0/0/1/NO  Homo sapiens     2019-12-30   \n",
       "1         0/0/0/0/NO  Homo sapiens     2020-01-13   \n",
       "2         0/0/0/0/NO  Homo sapiens     2020-01-08   \n",
       "3         0/0/0/2/NO  Homo sapiens     2020-01-01   \n",
       "4         0/0/0/0/NO  Homo sapiens     2019-12-30   \n",
       "\n",
       "                        location  \\\n",
       "0                  China / Hubei   \n",
       "1  Thailand/ Nonthaburi Province   \n",
       "2  Thailand/ Nonthaburi Province   \n",
       "3          China / Hubei / Wuhan   \n",
       "4          China / Hubei / Wuhan   \n",
       "\n",
       "                                     Originating Lab Submission Date  \\\n",
       "0  Hubei Provincial Center for Disease Control an...      2020-01-19   \n",
       "1  Department of Medical Sciences, Ministry of Pu...      2020-01-17   \n",
       "2  Department of Medical Sciences, Ministry of Pu...      2020-01-17   \n",
       "3  National Institute for Viral Disease Control a...      2020-01-11   \n",
       "4  National Institute for Viral Disease Control a...      2020-01-10   \n",
       "\n",
       "                                      Submitting Lab          Create Time  \\\n",
       "0  Hubei Provincial Center for Disease Control an...  2020-01-20 20:04:48   \n",
       "1  Department of Medical Sciences, Ministry of Pu...  2020-01-20 20:04:48   \n",
       "2  Department of Medical Sciences, Ministry of Pu...  2020-01-20 20:04:48   \n",
       "3  National Institute for Viral Disease Control a...  2020-01-20 20:04:48   \n",
       "4  National Institute for Viral Disease Control a...  2020-01-20 20:04:48   \n",
       "\n",
       "      Last Update Time                           alias              id  \\\n",
       "0  2020-05-07 23:03:25  NMDC60013088-01;EPI_ISL_402132  EPI_ISL_402132   \n",
       "1  2020-06-28 15:24:28                  EPI_ISL_403963  EPI_ISL_403963   \n",
       "2  2020-06-28 15:24:28                  EPI_ISL_403962  EPI_ISL_403962   \n",
       "3  2020-05-07 23:03:25  NMDC60013085-01;EPI_ISL_402120  EPI_ISL_402120   \n",
       "4  2020-05-07 23:03:25  NMDC60013084-01;EPI_ISL_402119  EPI_ISL_402119   \n",
       "\n",
       "  hostTaxonomyId        taxonomyId      loc1                 loc2   loc3  \\\n",
       "0  taxonomy:9606  taxonomy:2697049     China                Hubei   None   \n",
       "1  taxonomy:9606  taxonomy:2697049  Thailand  Nonthaburi Province   None   \n",
       "2  taxonomy:9606  taxonomy:2697049  Thailand  Nonthaburi Province   None   \n",
       "3  taxonomy:9606  taxonomy:2697049     China                Hubei  Wuhan   \n",
       "4  taxonomy:9606  taxonomy:2697049     China                Hubei  Wuhan   \n",
       "\n",
       "   loc4                  origLocation                                filename  \n",
       "0  None                   China,Hubei  2019-nCoV_EPI_ISL_402132_variants.gff3  \n",
       "1  None  Thailand,Nonthaburi Province  2019-nCoV_EPI_ISL_403963_variants.gff3  \n",
       "2  None  Thailand,Nonthaburi Province  2019-nCoV_EPI_ISL_403962_variants.gff3  \n",
       "3  None             China,Hubei,Wuhan  2019-nCoV_EPI_ISL_402120_variants.gff3  \n",
       "4  None             China,Hubei,Wuhan  2019-nCoV_EPI_ISL_402119_variants.gff3  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Keep only unique entries (there are a few duplicate cases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop_duplicates(subset='name', inplace=True)\n",
    "df.drop_duplicates(subset='id', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Strains with a matching filename: 43743\n"
     ]
    }
   ],
   "source": [
    "print('Strains with a matching filename:', df.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Download variant annotations for each strain\n",
    "To avoid download the same files every time, they are cached, and newly downloaded files are added to the cache."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "names = ['taxon1', 'variantType', 'name', 'start', 'end','x1', 'x2', 'x3','taxon2', 'x4', 'strainStart', 'taxon3', 'x5', 'strainEnd', 'ref', 'alt', 'vepAnnotation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_gff3(filename, url):\n",
    "    gff3 = pd.read_csv(url, header=None, comment='#', sep='[\\t|;]', engine='python', names=names)\n",
    "    try:\n",
    "        gff3['ref'] = gff3['ref'].str.replace('REF=','')\n",
    "        gff3['alt'] = gff3['alt'].str.replace('ALT=','')\n",
    "        gff3['vepAnnotation'] = gff3['vepAnnotation'].str.replace('VEP=','')\n",
    "        # prepare for 3-way split (need at least two commas)\n",
    "        gff3['vepAnnotation'] = gff3['vepAnnotation'].apply(lambda s: s + ',,' if s.count(',') < 2 else s)\n",
    "        # 3-way split\n",
    "        gff3[['variantConsequence','proteinVariant','geneVariant']] = gff3['vepAnnotation'].str.split(',', n=2, expand=True)\n",
    "        gff3['geneVariant'] = gff3['geneVariant'].str.replace('gene-','')\n",
    "        gff3 = gff3[['name', 'variantType', 'start', 'end', 'ref', 'alt', 'variantConsequence', 'proteinVariant', 'geneVariant']]\n",
    "    \n",
    "        filename = row['filename'] + '.csv'\n",
    "        gff3.to_csv(CACHE / filename, index=False)\n",
    "    except:\n",
    "        print('Parsing failed for: ', row['filename'])\n",
    "        # cache files that failed to parse so we don't reprocess them next time\n",
    "        filename = row['filename'] + '.csv'\n",
    "        gff3 = pd.read_csv(url, header=None, comment='#', sep='[\\t|;]', engine='python', names=names)\n",
    "        gff3.to_csv(CACHE_FAILED / filename, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_EPI_ISL_411220_variants.gff3 2019-nCoV_EPI_ISL_415787_variants.gff3 2019-nCoV_EPI_ISL_416907_variants.gff3 2019-nCoV_EPI_ISL_418186_variants.gff3 2019-nCoV_EPI_ISL_419264_variants.gff3 2019-nCoV_EPI_ISL_419265_variants.gff3 2019-nCoV_EPI_ISL_419266_variants.gff3 2019-nCoV_EPI_ISL_419908_variants.gff3 2019-nCoV_EPI_ISL_419909_variants.gff3 2019-nCoV_EPI_ISL_419910_variants.gff3 2019-nCoV_EPI_ISL_419911_variants.gff3 2019-nCoV_EPI_ISL_419912_variants.gff3 2019-nCoV_EPI_ISL_419914_variants.gff3 2019-nCoV_EPI_ISL_419915_variants.gff3 2019-nCoV_EPI_ISL_419916_variants.gff3 2019-nCoV_EPI_ISL_419917_variants.gff3 2019-nCoV_EPI_ISL_419918_variants.gff3 2019-nCoV_EPI_ISL_419919_variants.gff3 2019-nCoV_EPI_ISL_419920_variants.gff3 2019-nCoV_EPI_ISL_419921_variants.gff3 2019-nCoV_EPI_ISL_419922_variants.gff3 2019-nCoV_EPI_ISL_419923_variants.gff3 2019-nCoV_EPI_ISL_419924_variants.gff3 2019-nCoV_EPI_ISL_419925_variants.gff3 2019-nCoV_EPI_ISL_419926_variants.gff3 2019-nCoV_EPI_ISL_419927_variants.gff3 2019-nCoV_EPI_ISL_419928_variants.gff3 2019-nCoV_EPI_ISL_419929_variants.gff3 2019-nCoV_EPI_ISL_419931_variants.gff3 2019-nCoV_EPI_ISL_419932_variants.gff3 2019-nCoV_EPI_ISL_419933_variants.gff3 2019-nCoV_EPI_ISL_419934_variants.gff3 2019-nCoV_EPI_ISL_419935_variants.gff3 2019-nCoV_EPI_ISL_419936_variants.gff3 2019-nCoV_EPI_ISL_419937_variants.gff3 2019-nCoV_EPI_ISL_419938_variants.gff3 2019-nCoV_EPI_ISL_419939_variants.gff3 2019-nCoV_EPI_ISL_419940_variants.gff3 2019-nCoV_EPI_ISL_419941_variants.gff3 2019-nCoV_EPI_ISL_419942_variants.gff3 2019-nCoV_EPI_ISL_419943_variants.gff3 2019-nCoV_EPI_ISL_419944_variants.gff3 2019-nCoV_EPI_ISL_419945_variants.gff3 2019-nCoV_EPI_ISL_419946_variants.gff3 2019-nCoV_EPI_ISL_419947_variants.gff3 2019-nCoV_EPI_ISL_419948_variants.gff3 2019-nCoV_EPI_ISL_419949_variants.gff3 2019-nCoV_EPI_ISL_419950_variants.gff3 2019-nCoV_EPI_ISL_419951_variants.gff3 2019-nCoV_EPI_ISL_419952_variants.gff3 2019-nCoV_EPI_ISL_419954_variants.gff3 2019-nCoV_EPI_ISL_419955_variants.gff3 2019-nCoV_EPI_ISL_419956_variants.gff3 2019-nCoV_EPI_ISL_419957_variants.gff3 2019-nCoV_EPI_ISL_419958_variants.gff3 2019-nCoV_EPI_ISL_419959_variants.gff3 2019-nCoV_EPI_ISL_419960_variants.gff3 2019-nCoV_EPI_ISL_419961_variants.gff3 2019-nCoV_EPI_ISL_419962_variants.gff3 2019-nCoV_EPI_ISL_419963_variants.gff3 2019-nCoV_EPI_ISL_419964_variants.gff3 2019-nCoV_EPI_ISL_419965_variants.gff3 2019-nCoV_EPI_ISL_419966_variants.gff3 2019-nCoV_EPI_ISL_419967_variants.gff3 2019-nCoV_EPI_ISL_419968_variants.gff3 2019-nCoV_EPI_ISL_419969_variants.gff3 2019-nCoV_EPI_ISL_419970_variants.gff3 2019-nCoV_EPI_ISL_419971_variants.gff3 2019-nCoV_EPI_ISL_419972_variants.gff3 2019-nCoV_EPI_ISL_419973_variants.gff3 2019-nCoV_EPI_ISL_419974_variants.gff3 2019-nCoV_EPI_ISL_419975_variants.gff3 2019-nCoV_EPI_ISL_419976_variants.gff3 2019-nCoV_EPI_ISL_419977_variants.gff3 2019-nCoV_EPI_ISL_419978_variants.gff3 2019-nCoV_EPI_ISL_419979_variants.gff3 2019-nCoV_EPI_ISL_419980_variants.gff3 2019-nCoV_EPI_ISL_419982_variants.gff3 2019-nCoV_EPI_ISL_419983_variants.gff3 2019-nCoV_EPI_ISL_419984_variants.gff3 2019-nCoV_EPI_ISL_419985_variants.gff3 2019-nCoV_EPI_ISL_419986_variants.gff3 2019-nCoV_EPI_ISL_419987_variants.gff3 2019-nCoV_EPI_ISL_419988_variants.gff3 2019-nCoV_EPI_ISL_419989_variants.gff3 2019-nCoV_EPI_ISL_419990_variants.gff3 2019-nCoV_EPI_ISL_419991_variants.gff3 2019-nCoV_EPI_ISL_419992_variants.gff3 2019-nCoV_EPI_ISL_419993_variants.gff3 2019-nCoV_EPI_ISL_419994_variants.gff3 2019-nCoV_EPI_ISL_419995_variants.gff3 2019-nCoV_EPI_ISL_419996_variants.gff3 2019-nCoV_EPI_ISL_419997_variants.gff3 2019-nCoV_EPI_ISL_419998_variants.gff3 2019-nCoV_EPI_ISL_419999_variants.gff3 2019-nCoV_EPI_ISL_420000_variants.gff3 2019-nCoV_EPI_ISL_420001_variants.gff3 2019-nCoV_EPI_ISL_420002_variants.gff3 2019-nCoV_EPI_ISL_420003_variants.gff3 2019-nCoV_EPI_ISL_420004_variants.gff3 2019-nCoV_EPI_ISL_420005_variants.gff3 2019-nCoV_EPI_ISL_420006_variants.gff3 2019-nCoV_EPI_ISL_420007_variants.gff3 2019-nCoV_EPI_ISL_420008_variants.gff3 2019-nCoV_EPI_ISL_420009_variants.gff3 2019-nCoV_EPI_ISL_420010_variants.gff3 2019-nCoV_EPI_ISL_420011_variants.gff3 2019-nCoV_EPI_ISL_420012_variants.gff3 2019-nCoV_EPI_ISL_420013_variants.gff3 2019-nCoV_EPI_ISL_420014_variants.gff3 2019-nCoV_EPI_ISL_420015_variants.gff3 2019-nCoV_EPI_ISL_420016_variants.gff3 2019-nCoV_EPI_ISL_420017_variants.gff3 2019-nCoV_EPI_ISL_420018_variants.gff3 2019-nCoV_EPI_ISL_420020_variants.gff3 2019-nCoV_EPI_ISL_420021_variants.gff3 2019-nCoV_EPI_ISL_420022_variants.gff3 2019-nCoV_EPI_ISL_420023_variants.gff3 2019-nCoV_EPI_ISL_420024_variants.gff3 2019-nCoV_EPI_ISL_420025_variants.gff3 2019-nCoV_EPI_ISL_420026_variants.gff3 2019-nCoV_EPI_ISL_420027_variants.gff3 2019-nCoV_EPI_ISL_420028_variants.gff3 2019-nCoV_EPI_ISL_420029_variants.gff3 2019-nCoV_MT276324_variants.gff3 2019-nCoV_MT276325_variants.gff3 2019-nCoV_MT276328_variants.gff3 2019-nCoV_MT276597_variants.gff3 2019-nCoV_GWHACEY01000001_variants.gff3 2019-nCoV_EPI_ISL_420030_variants.gff3 2019-nCoV_EPI_ISL_420032_variants.gff3 2019-nCoV_EPI_ISL_420033_variants.gff3 2019-nCoV_EPI_ISL_420035_variants.gff3 2019-nCoV_EPI_ISL_420036_variants.gff3 2019-nCoV_EPI_ISL_420037_variants.gff3 2019-nCoV_EPI_ISL_420038_variants.gff3 2019-nCoV_EPI_ISL_420039_variants.gff3 2019-nCoV_EPI_ISL_420040_variants.gff3 2019-nCoV_EPI_ISL_420041_variants.gff3 2019-nCoV_EPI_ISL_420042_variants.gff3 2019-nCoV_EPI_ISL_420043_variants.gff3 2019-nCoV_EPI_ISL_420044_variants.gff3 2019-nCoV_EPI_ISL_420045_variants.gff3 2019-nCoV_EPI_ISL_420046_variants.gff3 2019-nCoV_EPI_ISL_420048_variants.gff3 2019-nCoV_EPI_ISL_420049_variants.gff3 2019-nCoV_EPI_ISL_420050_variants.gff3 2019-nCoV_EPI_ISL_420051_variants.gff3 2019-nCoV_EPI_ISL_420052_variants.gff3 2019-nCoV_EPI_ISL_420055_variants.gff3 2019-nCoV_EPI_ISL_420056_variants.gff3 2019-nCoV_EPI_ISL_420057_variants.gff3 2019-nCoV_EPI_ISL_420058_variants.gff3 2019-nCoV_EPI_ISL_420059_variants.gff3 2019-nCoV_EPI_ISL_420060_variants.gff3 2019-nCoV_EPI_ISL_420061_variants.gff3 2019-nCoV_EPI_ISL_420063_variants.gff3 2019-nCoV_EPI_ISL_420064_variants.gff3 2019-nCoV_EPI_ISL_420065_variants.gff3 2019-nCoV_EPI_ISL_420066_variants.gff3 2019-nCoV_EPI_ISL_420067_variants.gff3 2019-nCoV_EPI_ISL_420069_variants.gff3 2019-nCoV_EPI_ISL_420070_variants.gff3 2019-nCoV_EPI_ISL_420073_variants.gff3 2019-nCoV_EPI_ISL_420074_variants.gff3 2019-nCoV_EPI_ISL_420076_variants.gff3 2019-nCoV_EPI_ISL_420077_variants.gff3 2019-nCoV_EPI_ISL_420080_variants.gff3 2019-nCoV_EPI_ISL_420081_variants.gff3 2019-nCoV_EPI_ISL_420082_variants.gff3 2019-nCoV_EPI_ISL_420083_variants.gff3 2019-nCoV_EPI_ISL_420084_variants.gff3 2019-nCoV_EPI_ISL_420085_variants.gff3 2019-nCoV_EPI_ISL_420099_variants.gff3 2019-nCoV_EPI_ISL_420100_variants.gff3 2019-nCoV_EPI_ISL_420101_variants.gff3 2019-nCoV_EPI_ISL_420102_variants.gff3 2019-nCoV_EPI_ISL_420103_variants.gff3 2019-nCoV_EPI_ISL_420104_variants.gff3 2019-nCoV_EPI_ISL_420105_variants.gff3 2019-nCoV_EPI_ISL_420107_variants.gff3 2019-nCoV_EPI_ISL_420108_variants.gff3 2019-nCoV_EPI_ISL_420109_variants.gff3 2019-nCoV_EPI_ISL_420110_variants.gff3 2019-nCoV_EPI_ISL_420111_variants.gff3 2019-nCoV_EPI_ISL_420112_variants.gff3 2019-nCoV_EPI_ISL_420113_variants.gff3 2019-nCoV_EPI_ISL_420115_variants.gff3 2019-nCoV_EPI_ISL_420117_variants.gff3 2019-nCoV_EPI_ISL_420118_variants.gff3 2019-nCoV_EPI_ISL_420119_variants.gff3 2019-nCoV_EPI_ISL_420120_variants.gff3 2019-nCoV_EPI_ISL_420121_variants.gff3 2019-nCoV_EPI_ISL_420122_variants.gff3 2019-nCoV_EPI_ISL_420123_variants.gff3 2019-nCoV_EPI_ISL_420124_variants.gff3 2019-nCoV_EPI_ISL_420125_variants.gff3 2019-nCoV_EPI_ISL_420126_variants.gff3 2019-nCoV_EPI_ISL_420127_variants.gff3 2019-nCoV_EPI_ISL_420128_variants.gff3 2019-nCoV_EPI_ISL_420129_variants.gff3 2019-nCoV_EPI_ISL_420130_variants.gff3 2019-nCoV_EPI_ISL_420131_variants.gff3 2019-nCoV_EPI_ISL_420132_variants.gff3 2019-nCoV_EPI_ISL_420135_variants.gff3 2019-nCoV_EPI_ISL_420136_variants.gff3 2019-nCoV_EPI_ISL_420139_variants.gff3 2019-nCoV_EPI_ISL_420140_variants.gff3 2019-nCoV_EPI_ISL_420141_variants.gff3 2019-nCoV_EPI_ISL_420142_variants.gff3 2019-nCoV_EPI_ISL_420143_variants.gff3 2019-nCoV_EPI_ISL_420144_variants.gff3 2019-nCoV_EPI_ISL_420145_variants.gff3 2019-nCoV_EPI_ISL_420146_variants.gff3 2019-nCoV_EPI_ISL_420147_variants.gff3 2019-nCoV_EPI_ISL_420150_variants.gff3 2019-nCoV_EPI_ISL_420151_variants.gff3 2019-nCoV_EPI_ISL_420152_variants.gff3 2019-nCoV_EPI_ISL_420157_variants.gff3 2019-nCoV_EPI_ISL_420158_variants.gff3 2019-nCoV_EPI_ISL_420159_variants.gff3 2019-nCoV_EPI_ISL_420160_variants.gff3 2019-nCoV_EPI_ISL_420162_variants.gff3 2019-nCoV_EPI_ISL_420163_variants.gff3 2019-nCoV_EPI_ISL_420165_variants.gff3 2019-nCoV_EPI_ISL_420166_variants.gff3 2019-nCoV_EPI_ISL_420167_variants.gff3 2019-nCoV_EPI_ISL_420171_variants.gff3 2019-nCoV_EPI_ISL_420173_variants.gff3 2019-nCoV_EPI_ISL_420175_variants.gff3 2019-nCoV_EPI_ISL_420177_variants.gff3 2019-nCoV_EPI_ISL_420178_variants.gff3 2019-nCoV_EPI_ISL_420179_variants.gff3 2019-nCoV_EPI_ISL_420181_variants.gff3 2019-nCoV_EPI_ISL_420184_variants.gff3 2019-nCoV_EPI_ISL_420185_variants.gff3 2019-nCoV_EPI_ISL_420188_variants.gff3 2019-nCoV_EPI_ISL_420189_variants.gff3 2019-nCoV_EPI_ISL_420190_variants.gff3 2019-nCoV_EPI_ISL_420192_variants.gff3 2019-nCoV_EPI_ISL_420193_variants.gff3 2019-nCoV_EPI_ISL_420194_variants.gff3 2019-nCoV_EPI_ISL_420196_variants.gff3 2019-nCoV_EPI_ISL_420197_variants.gff3 2019-nCoV_EPI_ISL_420199_variants.gff3 2019-nCoV_EPI_ISL_420200_variants.gff3 2019-nCoV_EPI_ISL_420202_variants.gff3 2019-nCoV_EPI_ISL_420203_variants.gff3 2019-nCoV_EPI_ISL_420205_variants.gff3 2019-nCoV_EPI_ISL_420208_variants.gff3 2019-nCoV_EPI_ISL_420210_variants.gff3 2019-nCoV_EPI_ISL_420211_variants.gff3 2019-nCoV_EPI_ISL_420213_variants.gff3 2019-nCoV_EPI_ISL_420215_variants.gff3 2019-nCoV_EPI_ISL_420216_variants.gff3 2019-nCoV_EPI_ISL_420217_variants.gff3 2019-nCoV_EPI_ISL_420219_variants.gff3 2019-nCoV_EPI_ISL_420220_variants.gff3 2019-nCoV_EPI_ISL_420221_variants.gff3 2019-nCoV_EPI_ISL_420222_variants.gff3 2019-nCoV_EPI_ISL_420223_variants.gff3 2019-nCoV_EPI_ISL_420224_variants.gff3 2019-nCoV_EPI_ISL_420225_variants.gff3 2019-nCoV_EPI_ISL_420229_variants.gff3 2019-nCoV_EPI_ISL_420231_variants.gff3 2019-nCoV_EPI_ISL_420232_variants.gff3 2019-nCoV_EPI_ISL_420233_variants.gff3 2019-nCoV_EPI_ISL_420235_variants.gff3 2019-nCoV_EPI_ISL_420236_variants.gff3 2019-nCoV_EPI_ISL_420237_variants.gff3 2019-nCoV_EPI_ISL_420238_variants.gff3 2019-nCoV_EPI_ISL_420239_variants.gff3 2019-nCoV_EPI_ISL_420240_variants.gff3 2019-nCoV_EPI_ISL_420241_variants.gff3 2019-nCoV_EPI_ISL_420242_variants.gff3 2019-nCoV_EPI_ISL_420243_variants.gff3 2019-nCoV_EPI_ISL_420244_variants.gff3 2019-nCoV_EPI_ISL_420247_variants.gff3 2019-nCoV_EPI_ISL_420248_variants.gff3 2019-nCoV_EPI_ISL_420249_variants.gff3 2019-nCoV_EPI_ISL_420250_variants.gff3 2019-nCoV_EPI_ISL_420251_variants.gff3 2019-nCoV_EPI_ISL_420252_variants.gff3 2019-nCoV_EPI_ISL_420254_variants.gff3 2019-nCoV_EPI_ISL_420259_variants.gff3 2019-nCoV_EPI_ISL_420260_variants.gff3 2019-nCoV_EPI_ISL_420261_variants.gff3 2019-nCoV_EPI_ISL_420262_variants.gff3 2019-nCoV_EPI_ISL_420263_variants.gff3 2019-nCoV_EPI_ISL_420264_variants.gff3 2019-nCoV_EPI_ISL_420265_variants.gff3 2019-nCoV_EPI_ISL_420267_variants.gff3 2019-nCoV_EPI_ISL_420268_variants.gff3 2019-nCoV_EPI_ISL_420269_variants.gff3 2019-nCoV_EPI_ISL_420270_variants.gff3 2019-nCoV_EPI_ISL_420276_variants.gff3 2019-nCoV_EPI_ISL_420277_variants.gff3 2019-nCoV_EPI_ISL_420278_variants.gff3 2019-nCoV_EPI_ISL_420279_variants.gff3 2019-nCoV_EPI_ISL_420280_variants.gff3 2019-nCoV_EPI_ISL_420282_variants.gff3 2019-nCoV_EPI_ISL_420284_variants.gff3 2019-nCoV_EPI_ISL_420285_variants.gff3 2019-nCoV_EPI_ISL_420289_variants.gff3 2019-nCoV_EPI_ISL_420292_variants.gff3 2019-nCoV_GWHABKZ01000001_variants.gff3 2019-nCoV_GWHACAT01000001_variants.gff3 2019-nCoV_GWHACAU01000001_variants.gff3 2019-nCoV_GWHACAV01000001_variants.gff3 2019-nCoV_GWHACAW01000001_variants.gff3 2019-nCoV_GWHACAX01000001_variants.gff3 2019-nCoV_GWHACAY01000001_variants.gff3 2019-nCoV_GWHACAZ01000001_variants.gff3 2019-nCoV_GWHACBA01000001_variants.gff3 2019-nCoV_GWHACBD01000001_variants.gff3 2019-nCoV_EPI_ISL_420294_variants.gff3 2019-nCoV_EPI_ISL_420295_variants.gff3 2019-nCoV_EPI_ISL_420296_variants.gff3 2019-nCoV_EPI_ISL_420297_variants.gff3 2019-nCoV_EPI_ISL_420298_variants.gff3 2019-nCoV_EPI_ISL_420299_variants.gff3 2019-nCoV_EPI_ISL_420300_variants.gff3 2019-nCoV_EPI_ISL_420301_variants.gff3 2019-nCoV_EPI_ISL_420302_variants.gff3 2019-nCoV_EPI_ISL_420303_variants.gff3 2019-nCoV_EPI_ISL_420304_variants.gff3 2019-nCoV_EPI_ISL_420305_variants.gff3 2019-nCoV_EPI_ISL_420306_variants.gff3 2019-nCoV_EPI_ISL_420307_variants.gff3 2019-nCoV_EPI_ISL_420308_variants.gff3 2019-nCoV_EPI_ISL_420309_variants.gff3 2019-nCoV_EPI_ISL_420312_variants.gff3 2019-nCoV_EPI_ISL_420313_variants.gff3 2019-nCoV_EPI_ISL_420314_variants.gff3 2019-nCoV_EPI_ISL_420315_variants.gff3 2019-nCoV_EPI_ISL_420319_variants.gff3 2019-nCoV_EPI_ISL_420321_variants.gff3 2019-nCoV_EPI_ISL_420322_variants.gff3 2019-nCoV_EPI_ISL_420323_variants.gff3 2019-nCoV_EPI_ISL_420324_variants.gff3 2019-nCoV_EPI_ISL_420325_variants.gff3 2019-nCoV_EPI_ISL_420326_variants.gff3 2019-nCoV_EPI_ISL_420327_variants.gff3 2019-nCoV_EPI_ISL_420328_variants.gff3 2019-nCoV_EPI_ISL_420329_variants.gff3 2019-nCoV_EPI_ISL_420330_variants.gff3 2019-nCoV_EPI_ISL_420331_variants.gff3 2019-nCoV_EPI_ISL_420332_variants.gff3 2019-nCoV_EPI_ISL_420333_variants.gff3 2019-nCoV_EPI_ISL_420334_variants.gff3 2019-nCoV_EPI_ISL_420335_variants.gff3 2019-nCoV_EPI_ISL_420336_variants.gff3 2019-nCoV_EPI_ISL_420337_variants.gff3 2019-nCoV_EPI_ISL_420338_variants.gff3 2019-nCoV_EPI_ISL_420339_variants.gff3 2019-nCoV_EPI_ISL_420340_variants.gff3 2019-nCoV_EPI_ISL_420341_variants.gff3 2019-nCoV_EPI_ISL_420342_variants.gff3 2019-nCoV_EPI_ISL_420343_variants.gff3 2019-nCoV_EPI_ISL_420344_variants.gff3 2019-nCoV_EPI_ISL_420345_variants.gff3 2019-nCoV_EPI_ISL_420347_variants.gff3 2019-nCoV_EPI_ISL_420349_variants.gff3 2019-nCoV_EPI_ISL_420350_variants.gff3 2019-nCoV_EPI_ISL_420351_variants.gff3 2019-nCoV_EPI_ISL_420352_variants.gff3 2019-nCoV_EPI_ISL_420353_variants.gff3 2019-nCoV_EPI_ISL_420354_variants.gff3 2019-nCoV_EPI_ISL_420355_variants.gff3 2019-nCoV_EPI_ISL_420356_variants.gff3 2019-nCoV_EPI_ISL_420357_variants.gff3 2019-nCoV_EPI_ISL_420358_variants.gff3 2019-nCoV_EPI_ISL_420359_variants.gff3 2019-nCoV_EPI_ISL_420360_variants.gff3 2019-nCoV_EPI_ISL_420361_variants.gff3 2019-nCoV_EPI_ISL_420362_variants.gff3 2019-nCoV_EPI_ISL_420363_variants.gff3 2019-nCoV_EPI_ISL_420364_variants.gff3 2019-nCoV_EPI_ISL_420365_variants.gff3 2019-nCoV_EPI_ISL_420366_variants.gff3 2019-nCoV_EPI_ISL_420367_variants.gff3 2019-nCoV_EPI_ISL_420368_variants.gff3 2019-nCoV_EPI_ISL_420370_variants.gff3 2019-nCoV_EPI_ISL_420371_variants.gff3 2019-nCoV_EPI_ISL_420372_variants.gff3 2019-nCoV_EPI_ISL_420373_variants.gff3 2019-nCoV_EPI_ISL_420374_variants.gff3 2019-nCoV_EPI_ISL_420375_variants.gff3 2019-nCoV_EPI_ISL_420376_variants.gff3 2019-nCoV_EPI_ISL_420377_variants.gff3 2019-nCoV_EPI_ISL_420378_variants.gff3 2019-nCoV_EPI_ISL_420379_variants.gff3 2019-nCoV_EPI_ISL_420380_variants.gff3 2019-nCoV_EPI_ISL_420381_variants.gff3 2019-nCoV_EPI_ISL_420382_variants.gff3 2019-nCoV_EPI_ISL_420383_variants.gff3 2019-nCoV_EPI_ISL_420384_variants.gff3 2019-nCoV_EPI_ISL_420385_variants.gff3 2019-nCoV_EPI_ISL_420386_variants.gff3 2019-nCoV_EPI_ISL_420387_variants.gff3 2019-nCoV_EPI_ISL_420389_variants.gff3 2019-nCoV_EPI_ISL_420390_variants.gff3 2019-nCoV_EPI_ISL_420391_variants.gff3 2019-nCoV_EPI_ISL_420392_variants.gff3 2019-nCoV_EPI_ISL_420393_variants.gff3 2019-nCoV_EPI_ISL_420395_variants.gff3 2019-nCoV_EPI_ISL_420396_variants.gff3 2019-nCoV_EPI_ISL_420397_variants.gff3 2019-nCoV_EPI_ISL_420398_variants.gff3 2019-nCoV_EPI_ISL_420399_variants.gff3 2019-nCoV_EPI_ISL_420400_variants.gff3 2019-nCoV_EPI_ISL_420402_variants.gff3 2019-nCoV_EPI_ISL_420403_variants.gff3 2019-nCoV_EPI_ISL_420404_variants.gff3 2019-nCoV_EPI_ISL_420405_variants.gff3 2019-nCoV_EPI_ISL_420407_variants.gff3 2019-nCoV_EPI_ISL_420409_variants.gff3 2019-nCoV_EPI_ISL_420410_variants.gff3 2019-nCoV_EPI_ISL_420411_variants.gff3 2019-nCoV_EPI_ISL_420412_variants.gff3 2019-nCoV_EPI_ISL_420413_variants.gff3 2019-nCoV_EPI_ISL_420414_variants.gff3 2019-nCoV_EPI_ISL_420415_variants.gff3 2019-nCoV_EPI_ISL_420416_variants.gff3 2019-nCoV_EPI_ISL_420417_variants.gff3 2019-nCoV_EPI_ISL_420418_variants.gff3 2019-nCoV_EPI_ISL_420419_variants.gff3 2019-nCoV_EPI_ISL_420420_variants.gff3 2019-nCoV_EPI_ISL_420421_variants.gff3 2019-nCoV_EPI_ISL_420422_variants.gff3 2019-nCoV_EPI_ISL_420423_variants.gff3 2019-nCoV_EPI_ISL_420424_variants.gff3 2019-nCoV_EPI_ISL_420426_variants.gff3 2019-nCoV_EPI_ISL_420427_variants.gff3 2019-nCoV_EPI_ISL_420428_variants.gff3 2019-nCoV_EPI_ISL_420429_variants.gff3 2019-nCoV_EPI_ISL_420430_variants.gff3 2019-nCoV_EPI_ISL_420431_variants.gff3 2019-nCoV_EPI_ISL_420432_variants.gff3 2019-nCoV_EPI_ISL_420433_variants.gff3 2019-nCoV_EPI_ISL_420434_variants.gff3 2019-nCoV_EPI_ISL_420435_variants.gff3 2019-nCoV_EPI_ISL_420436_variants.gff3 2019-nCoV_EPI_ISL_420437_variants.gff3 2019-nCoV_EPI_ISL_420438_variants.gff3 2019-nCoV_EPI_ISL_420439_variants.gff3 2019-nCoV_EPI_ISL_420440_variants.gff3 2019-nCoV_EPI_ISL_420441_variants.gff3 2019-nCoV_EPI_ISL_420442_variants.gff3 2019-nCoV_EPI_ISL_420443_variants.gff3 2019-nCoV_EPI_ISL_420444_variants.gff3 2019-nCoV_EPI_ISL_420445_variants.gff3 2019-nCoV_EPI_ISL_420446_variants.gff3 2019-nCoV_EPI_ISL_420448_variants.gff3 2019-nCoV_EPI_ISL_420452_variants.gff3 2019-nCoV_EPI_ISL_420456_variants.gff3 2019-nCoV_EPI_ISL_420508_variants.gff3 2019-nCoV_EPI_ISL_420531_variants.gff3 2019-nCoV_EPI_ISL_420532_variants.gff3 2019-nCoV_EPI_ISL_420533_variants.gff3 2019-nCoV_EPI_ISL_420534_variants.gff3 2019-nCoV_EPI_ISL_420536_variants.gff3 2019-nCoV_EPI_ISL_420537_variants.gff3 2019-nCoV_EPI_ISL_420538_variants.gff3 2019-nCoV_EPI_ISL_420539_variants.gff3 2019-nCoV_EPI_ISL_420540_variants.gff3 2019-nCoV_EPI_ISL_420541_variants.gff3 2019-nCoV_EPI_ISL_417443_variants.gff3 2019-nCoV_EPI_ISL_420543_variants.gff3 2019-nCoV_EPI_ISL_420544_variants.gff3 2019-nCoV_EPI_ISL_420545_variants.gff3 2019-nCoV_EPI_ISL_420546_variants.gff3 2019-nCoV_EPI_ISL_420547_variants.gff3 2019-nCoV_EPI_ISL_420548_variants.gff3 2019-nCoV_EPI_ISL_420549_variants.gff3 2019-nCoV_EPI_ISL_420550_variants.gff3 2019-nCoV_EPI_ISL_420551_variants.gff3 2019-nCoV_EPI_ISL_420552_variants.gff3 2019-nCoV_EPI_ISL_420553_variants.gff3 2019-nCoV_EPI_ISL_420554_variants.gff3 2019-nCoV_EPI_ISL_420555_variants.gff3 2019-nCoV_EPI_ISL_420556_variants.gff3 2019-nCoV_EPI_ISL_420563_variants.gff3 2019-nCoV_EPI_ISL_420564_variants.gff3 2019-nCoV_EPI_ISL_420565_variants.gff3 2019-nCoV_EPI_ISL_420567_variants.gff3 2019-nCoV_EPI_ISL_420568_variants.gff3 2019-nCoV_EPI_ISL_420569_variants.gff3 2019-nCoV_EPI_ISL_420570_variants.gff3 2019-nCoV_EPI_ISL_420571_variants.gff3 2019-nCoV_EPI_ISL_420572_variants.gff3 2019-nCoV_EPI_ISL_420573_variants.gff3 2019-nCoV_EPI_ISL_420575_variants.gff3 2019-nCoV_EPI_ISL_420577_variants.gff3 2019-nCoV_EPI_ISL_420578_variants.gff3 2019-nCoV_EPI_ISL_420579_variants.gff3 2019-nCoV_EPI_ISL_420580_variants.gff3 2019-nCoV_EPI_ISL_420581_variants.gff3 2019-nCoV_EPI_ISL_420583_variants.gff3 2019-nCoV_EPI_ISL_420585_variants.gff3 2019-nCoV_EPI_ISL_420586_variants.gff3 2019-nCoV_EPI_ISL_420587_variants.gff3 2019-nCoV_EPI_ISL_420588_variants.gff3 2019-nCoV_EPI_ISL_420589_variants.gff3 2019-nCoV_EPI_ISL_420590_variants.gff3 2019-nCoV_EPI_ISL_420591_variants.gff3 2019-nCoV_EPI_ISL_420598_variants.gff3 2019-nCoV_EPI_ISL_420599_variants.gff3 2019-nCoV_EPI_ISL_420600_variants.gff3 2019-nCoV_EPI_ISL_420604_variants.gff3 2019-nCoV_EPI_ISL_420605_variants.gff3 2019-nCoV_EPI_ISL_420606_variants.gff3 2019-nCoV_EPI_ISL_420607_variants.gff3 2019-nCoV_EPI_ISL_420608_variants.gff3 2019-nCoV_EPI_ISL_420609_variants.gff3 2019-nCoV_EPI_ISL_420610_variants.gff3 2019-nCoV_EPI_ISL_420611_variants.gff3 2019-nCoV_EPI_ISL_420613_variants.gff3 2019-nCoV_EPI_ISL_420614_variants.gff3 2019-nCoV_EPI_ISL_420615_variants.gff3 2019-nCoV_EPI_ISL_420616_variants.gff3 2019-nCoV_EPI_ISL_420617_variants.gff3 2019-nCoV_EPI_ISL_420618_variants.gff3 2019-nCoV_EPI_ISL_420619_variants.gff3 2019-nCoV_EPI_ISL_420620_variants.gff3 2019-nCoV_EPI_ISL_420621_variants.gff3 2019-nCoV_EPI_ISL_420623_variants.gff3 2019-nCoV_EPI_ISL_420624_variants.gff3 2019-nCoV_EPI_ISL_420625_variants.gff3 2019-nCoV_EPI_ISL_420628_variants.gff3 2019-nCoV_EPI_ISL_420629_variants.gff3 2019-nCoV_EPI_ISL_420630_variants.gff3 2019-nCoV_EPI_ISL_420636_variants.gff3 2019-nCoV_EPI_ISL_420640_variants.gff3 2019-nCoV_EPI_ISL_420689_variants.gff3 2019-nCoV_EPI_ISL_420690_variants.gff3 2019-nCoV_EPI_ISL_420692_variants.gff3 2019-nCoV_EPI_ISL_420693_variants.gff3 2019-nCoV_EPI_ISL_420694_variants.gff3 2019-nCoV_EPI_ISL_420697_variants.gff3 2019-nCoV_EPI_ISL_420716_variants.gff3 2019-nCoV_EPI_ISL_420718_variants.gff3 2019-nCoV_EPI_ISL_420719_variants.gff3 2019-nCoV_EPI_ISL_420721_variants.gff3 2019-nCoV_EPI_ISL_420722_variants.gff3 2019-nCoV_EPI_ISL_420723_variants.gff3 2019-nCoV_EPI_ISL_420724_variants.gff3 2019-nCoV_EPI_ISL_420727_variants.gff3 2019-nCoV_EPI_ISL_420728_variants.gff3 2019-nCoV_EPI_ISL_420731_variants.gff3 2019-nCoV_EPI_ISL_420732_variants.gff3 2019-nCoV_EPI_ISL_420733_variants.gff3 2019-nCoV_EPI_ISL_420734_variants.gff3 2019-nCoV_EPI_ISL_420742_variants.gff3 2019-nCoV_EPI_ISL_420743_variants.gff3 2019-nCoV_EPI_ISL_420745_variants.gff3 2019-nCoV_EPI_ISL_420747_variants.gff3 2019-nCoV_EPI_ISL_420751_variants.gff3 2019-nCoV_EPI_ISL_420755_variants.gff3 2019-nCoV_EPI_ISL_420766_variants.gff3 2019-nCoV_EPI_ISL_420767_variants.gff3 2019-nCoV_EPI_ISL_420772_variants.gff3 2019-nCoV_EPI_ISL_420774_variants.gff3 2019-nCoV_EPI_ISL_420784_variants.gff3 2019-nCoV_EPI_ISL_420785_variants.gff3 2019-nCoV_EPI_ISL_420786_variants.gff3 2019-nCoV_EPI_ISL_420787_variants.gff3 2019-nCoV_EPI_ISL_420788_variants.gff3 2019-nCoV_EPI_ISL_420790_variants.gff3 2019-nCoV_EPI_ISL_420791_variants.gff3 2019-nCoV_EPI_ISL_420792_variants.gff3 2019-nCoV_EPI_ISL_420793_variants.gff3 2019-nCoV_EPI_ISL_420794_variants.gff3 2019-nCoV_EPI_ISL_420795_variants.gff3 2019-nCoV_EPI_ISL_420796_variants.gff3 2019-nCoV_EPI_ISL_420797_variants.gff3 2019-nCoV_EPI_ISL_420798_variants.gff3 2019-nCoV_EPI_ISL_420799_variants.gff3 2019-nCoV_EPI_ISL_420800_variants.gff3 2019-nCoV_EPI_ISL_420801_variants.gff3 2019-nCoV_EPI_ISL_420847_variants.gff3 2019-nCoV_EPI_ISL_420849_variants.gff3 2019-nCoV_MT292569_variants.gff3 2019-nCoV_MT292570_variants.gff3 2019-nCoV_MT292571_variants.gff3 2019-nCoV_MT292572_variants.gff3 2019-nCoV_MT292573_variants.gff3 2019-nCoV_MT292574_variants.gff3 2019-nCoV_MT292575_variants.gff3 2019-nCoV_MT293156_variants.gff3 2019-nCoV_MT293158_variants.gff3 2019-nCoV_MT293159_variants.gff3 2019-nCoV_MT293160_variants.gff3 2019-nCoV_MT293161_variants.gff3 2019-nCoV_MT293162_variants.gff3 2019-nCoV_MT293163_variants.gff3 2019-nCoV_MT293164_variants.gff3 2019-nCoV_MT293165_variants.gff3 2019-nCoV_MT293166_variants.gff3 2019-nCoV_MT293167_variants.gff3 2019-nCoV_MT293169_variants.gff3 2019-nCoV_MT293170_variants.gff3 2019-nCoV_MT293171_variants.gff3 2019-nCoV_MT293172_variants.gff3 2019-nCoV_MT293173_variants.gff3 2019-nCoV_MT293174_variants.gff3 2019-nCoV_MT293175_variants.gff3 2019-nCoV_MT293176_variants.gff3 2019-nCoV_MT293177_variants.gff3 2019-nCoV_MT293178_variants.gff3 2019-nCoV_MT293179_variants.gff3 2019-nCoV_MT293181_variants.gff3 2019-nCoV_MT293182_variants.gff3 2019-nCoV_MT293183_variants.gff3 2019-nCoV_MT293184_variants.gff3 2019-nCoV_MT293185_variants.gff3 2019-nCoV_MT293186_variants.gff3 2019-nCoV_MT293187_variants.gff3 2019-nCoV_MT293188_variants.gff3 2019-nCoV_MT293189_variants.gff3 2019-nCoV_MT293190_variants.gff3 2019-nCoV_MT293191_variants.gff3 2019-nCoV_MT293192_variants.gff3 2019-nCoV_MT293194_variants.gff3 2019-nCoV_MT293195_variants.gff3 2019-nCoV_MT293196_variants.gff3 2019-nCoV_MT293197_variants.gff3 2019-nCoV_MT293198_variants.gff3 2019-nCoV_MT293199_variants.gff3 2019-nCoV_MT293200_variants.gff3 2019-nCoV_MT293201_variants.gff3 2019-nCoV_MT293202_variants.gff3 2019-nCoV_MT293204_variants.gff3 2019-nCoV_MT293205_variants.gff3 2019-nCoV_MT293206_variants.gff3 2019-nCoV_MT293207_variants.gff3 2019-nCoV_MT293208_variants.gff3 2019-nCoV_MT293209_variants.gff3 2019-nCoV_MT293210_variants.gff3 2019-nCoV_MT293211_variants.gff3 2019-nCoV_MT293212_variants.gff3 2019-nCoV_MT293213_variants.gff3 2019-nCoV_MT293214_variants.gff3 2019-nCoV_MT293215_variants.gff3 2019-nCoV_MT293216_variants.gff3 2019-nCoV_MT293218_variants.gff3 2019-nCoV_MT293219_variants.gff3 2019-nCoV_MT293220_variants.gff3 2019-nCoV_MT293222_variants.gff3 2019-nCoV_MT293224_variants.gff3 2019-nCoV_MT293225_variants.gff3 2019-nCoV_MT295464_variants.gff3 2019-nCoV_MT295465_variants.gff3 2019-nCoV_EPI_ISL_420876_variants.gff3 2019-nCoV_EPI_ISL_420877_variants.gff3 2019-nCoV_EPI_ISL_420878_variants.gff3 2019-nCoV_EPI_ISL_420879_variants.gff3 2019-nCoV_EPI_ISL_420899_variants.gff3 Download failed for:  2019-nCoV_EPI_ISL_420901_variants.gff3\n",
      "2019-nCoV_EPI_ISL_420903_variants.gff3 2019-nCoV_EPI_ISL_420912_variants.gff3 2019-nCoV_EPI_ISL_420929_variants.gff3 "
     ]
    }
   ],
   "source": [
    "for index, row in df.iterrows():\n",
    "    url = annotation_url + row['filename']\n",
    "    filename = row['filename'] + '.csv'\n",
    "    # skip files that have been processed in previous runs\n",
    "    if not (Path.exists(CACHE / filename) or Path.exists(CACHE_FAILED / filename)):\n",
    "        try:\n",
    "            download_gff3(row['filename'], url)\n",
    "            print(row['filename'], end=' ')\n",
    "        except:\n",
    "            print('Download failed for: ', row['filename'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concatenate all variation data into a single dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use all cached data files\n",
    "path = str(CACHE / '*.gff3.csv')\n",
    "filenames = glob.glob(path)\n",
    "\n",
    "variations = pd.concat((pd.read_csv(f, index_col=None, header=0) for f in filenames))\n",
    "variations.fillna('', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "List of variant types and consequences:\n",
    "\n",
    "https://uswest.ensembl.org/info/genome/variation/prediction/classification.html\n",
    "\n",
    "https://uswest.ensembl.org/info/genome/variation/prediction/predicted_data.html#consequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract protein position and protein id from proteinVariant string\n",
    "\n",
    "Example: QHD43415.1:p.5828P>L\n",
    "\n",
    "proteinPosition: 5828\n",
    "proteinId: QHD43415"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "position_pattern = re.compile(':p\\.(.*?)[A-Z|\\-]+')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_protein_position(s):\n",
    "    if s == '':\n",
    "        return s\n",
    "    else:\n",
    "        groups = position_pattern.search(s)\n",
    "        if groups == None:\n",
    "            return ''\n",
    "        else:\n",
    "            return groups.group(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variations['proteinPosition'] = variations['proteinVariant'].apply(extract_protein_position)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variations['proteinAccession'] = variations['proteinVariant'].apply(lambda s: s.split('.')[0] if '.' in s else '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variations['proteinAccession'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assign SARS-CoV-2 taxonomy id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variations['taxonomyId'] = 'taxonomy:2697049'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assign Reference genome\n",
    "\n",
    "The first SARS-CoV-2 genome sequence is the reference for the variant annotation below.\n",
    "\n",
    "[Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1](https://www.ncbi.nlm.nih.gov/nuccore/MN908947)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variations['referenceGenome'] = 'insdc:MN908947' -> same as NCBI reference sequence NC_045512\n",
    "variations['referenceGenome'] = 'ncbiprotein:NC_045512'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variations['proteinAccession'] = variations['proteinAccession'].apply(lambda s: 'ncbiprotein:' + s if s != '' else s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('variantType:', variations['variantType'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"variantConsequence:\", variations['variantConsequence'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Number of variants:\", variations.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variations.to_csv(NEO4J_IMPORT / \"01d-CNCBVariant.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variations.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
