{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example queries for biological entities on COVID-19-Net Knowledge Graph\n",
    "[Work in progress]\n",
    "\n",
    "This notebook demonstrates how to run Cypher queries to retrieve data about biological entities in the Knowledge Graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import pandas as pd\n",
    "from py2neo import Graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_rows = None  # display all rows\n",
    "pd.options.display.max_columns = None  # display all columsns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Connect to COVID-19-Net Knowledge Graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = Graph(\"bolt://132.249.238.185:7687\", user=\"reader\", password=\"demo\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What kind of entities, including bioentities, does the KG contain?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (n:MetaNode)\n",
    "RETURN n.name, n.shortDescription, n.description, n.example, n.definitionSource, n.dataSources\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fuzzy full-text search for bioentities\n",
    "Results are ordered by match score. \n",
    "\n",
    "Note, proteins may have one or more protein names. Proteins that have been cleaved have the fullLength flag set to `False`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>fullLength</th>\n",
       "      <th>labels(node)</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Spike</td>\n",
       "      <td>True</td>\n",
       "      <td>[Protein]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Spike</td>\n",
       "      <td>True</td>\n",
       "      <td>[Protein]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Spike</td>\n",
       "      <td>None</td>\n",
       "      <td>[ProteinName]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Spike</td>\n",
       "      <td>True</td>\n",
       "      <td>[Protein]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Spike</td>\n",
       "      <td>True</td>\n",
       "      <td>[Protein]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Spike</td>\n",
       "      <td>True</td>\n",
       "      <td>[Protein]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Spike</td>\n",
       "      <td>True</td>\n",
       "      <td>[Protein]</td>\n",
       "      <td>4.653281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pdb:5ZHY.A</td>\n",
       "      <td>None</td>\n",
       "      <td>[Chain]</td>\n",
       "      <td>4.592937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>pdb:5ZHY.B</td>\n",
       "      <td>None</td>\n",
       "      <td>[Chain]</td>\n",
       "      <td>4.592937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>pdb:5ZHY.C</td>\n",
       "      <td>None</td>\n",
       "      <td>[Chain]</td>\n",
       "      <td>4.592937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>pdb:5ZHY.D</td>\n",
       "      <td>None</td>\n",
       "      <td>[Chain]</td>\n",
       "      <td>4.592937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>pdb:5ZHY.E</td>\n",
       "      <td>None</td>\n",
       "      <td>[Chain]</td>\n",
       "      <td>4.592937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>pdb:5ZHY.F</td>\n",
       "      <td>None</td>\n",
       "      <td>[Chain]</td>\n",
       "      <td>4.592937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pdb:5ZHY</td>\n",
       "      <td>None</td>\n",
       "      <td>[Structure]</td>\n",
       "      <td>4.592937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Chersomanes albofasciata</td>\n",
       "      <td>None</td>\n",
       "      <td>[Organism]</td>\n",
       "      <td>4.592161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>pdb:5ZUV.A</td>\n",
       "      <td>None</td>\n",
       "      <td>[Chain]</td>\n",
       "      <td>4.220284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>pdb:5ZUV.B</td>\n",
       "      <td>None</td>\n",
       "      <td>[Chain]</td>\n",
       "      <td>4.220284</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name fullLength   labels(node)     score\n",
       "0                      Spike       True      [Protein]  4.653281\n",
       "1                      Spike       True      [Protein]  4.653281\n",
       "2                      Spike       None  [ProteinName]  4.653281\n",
       "3                      Spike       None  [ProteinName]  4.653281\n",
       "4                      Spike       None  [ProteinName]  4.653281\n",
       "5                      Spike       None  [ProteinName]  4.653281\n",
       "6                      Spike       None  [ProteinName]  4.653281\n",
       "7                      Spike       None  [ProteinName]  4.653281\n",
       "8                      Spike       None  [ProteinName]  4.653281\n",
       "9                      Spike       None  [ProteinName]  4.653281\n",
       "10                     Spike       None  [ProteinName]  4.653281\n",
       "11                     Spike       True      [Protein]  4.653281\n",
       "12                     Spike       True      [Protein]  4.653281\n",
       "13                     Spike       True      [Protein]  4.653281\n",
       "14                     Spike       True      [Protein]  4.653281\n",
       "15                pdb:5ZHY.A       None        [Chain]  4.592937\n",
       "16                pdb:5ZHY.B       None        [Chain]  4.592937\n",
       "17                pdb:5ZHY.C       None        [Chain]  4.592937\n",
       "18                pdb:5ZHY.D       None        [Chain]  4.592937\n",
       "19                pdb:5ZHY.E       None        [Chain]  4.592937\n",
       "20                pdb:5ZHY.F       None        [Chain]  4.592937\n",
       "21                  pdb:5ZHY       None    [Structure]  4.592937\n",
       "22  Chersomanes albofasciata       None     [Organism]  4.592161\n",
       "23                pdb:5ZUV.A       None        [Chain]  4.220284\n",
       "24                pdb:5ZUV.B       None        [Chain]  4.220284"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "CALL db.index.fulltext.queryNodes(\"bioentities\", \"Spike\") YIELD node, score\n",
    "RETURN node.name AS name, node.fullLength AS fullLength, labels(node), score\n",
    "\"\"\"\n",
    "df = graph.run(query).to_data_frame()\n",
    "df.head(25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exact full-text search for bioentities\n",
    "For exact matches, enclose the phrase with `\\\"`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>fullLength</th>\n",
       "      <th>accession</th>\n",
       "      <th>proId</th>\n",
       "      <th>taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1173</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:H1AG29</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q1HVK9</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1171</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:A0A384RBT3</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:A0A384QXY4</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1171</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:A0A384RHP9</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>567</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q6TUL3</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>353</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:A0A0A0QCI3</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>567</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q6TUL1</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:A0A384RNQ9</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q1HVL9</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1173</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:P15423</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>542</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:E2DNU8</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q1HVL3</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>567</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q6TUM0</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:S5YNN4</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>546</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:E2DNU6</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>578</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q6TUK8</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>559</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q6TUL0</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q1HVM3</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:A0A384R2L9</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:A0A384RK44</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q1HVM5</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>596</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q1HVK2</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:A0A384RBT6</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>1170</td>\n",
       "      <td>True</td>\n",
       "      <td>uniprot:Q1HVL8</td>\n",
       "      <td>None</td>\n",
       "      <td>taxonomy:11137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  name  start   end fullLength           accession proId  \\\n",
       "0   Spike glycoprotein      1  1173       True      uniprot:H1AG29  None   \n",
       "1   Spike glycoprotein      1  1170       True      uniprot:Q1HVK9  None   \n",
       "2   Spike glycoprotein      1  1171       True  uniprot:A0A384RBT3  None   \n",
       "3   Spike glycoprotein      1  1170       True  uniprot:A0A384QXY4  None   \n",
       "4   Spike glycoprotein      1  1171       True  uniprot:A0A384RHP9  None   \n",
       "5   Spike glycoprotein      1   567       True      uniprot:Q6TUL3  None   \n",
       "6   Spike glycoprotein      1   353       True  uniprot:A0A0A0QCI3  None   \n",
       "7   Spike glycoprotein      1   567       True      uniprot:Q6TUL1  None   \n",
       "8   Spike glycoprotein      1  1170       True  uniprot:A0A384RNQ9  None   \n",
       "9   Spike glycoprotein      1  1170       True      uniprot:Q1HVL9  None   \n",
       "10  Spike glycoprotein      1  1173       True      uniprot:P15423  None   \n",
       "11  Spike glycoprotein      1   542       True      uniprot:E2DNU8  None   \n",
       "12  Spike glycoprotein      1  1170       True      uniprot:Q1HVL3  None   \n",
       "13  Spike glycoprotein      1   567       True      uniprot:Q6TUM0  None   \n",
       "14  Spike glycoprotein      1  1170       True      uniprot:S5YNN4  None   \n",
       "15  Spike glycoprotein      1   546       True      uniprot:E2DNU6  None   \n",
       "16  Spike glycoprotein      1   578       True      uniprot:Q6TUK8  None   \n",
       "17  Spike glycoprotein      1   559       True      uniprot:Q6TUL0  None   \n",
       "18  Spike glycoprotein      1  1170       True      uniprot:Q1HVM3  None   \n",
       "19  Spike glycoprotein      1  1170       True  uniprot:A0A384R2L9  None   \n",
       "20  Spike glycoprotein      1  1170       True  uniprot:A0A384RK44  None   \n",
       "21  Spike glycoprotein      1  1170       True      uniprot:Q1HVM5  None   \n",
       "22  Spike glycoprotein      1   596       True      uniprot:Q1HVK2  None   \n",
       "23  Spike glycoprotein      1  1170       True  uniprot:A0A384RBT6  None   \n",
       "24  Spike glycoprotein      1  1170       True      uniprot:Q1HVL8  None   \n",
       "\n",
       "          taxonomy  \n",
       "0   taxonomy:11137  \n",
       "1   taxonomy:11137  \n",
       "2   taxonomy:11137  \n",
       "3   taxonomy:11137  \n",
       "4   taxonomy:11137  \n",
       "5   taxonomy:11137  \n",
       "6   taxonomy:11137  \n",
       "7   taxonomy:11137  \n",
       "8   taxonomy:11137  \n",
       "9   taxonomy:11137  \n",
       "10  taxonomy:11137  \n",
       "11  taxonomy:11137  \n",
       "12  taxonomy:11137  \n",
       "13  taxonomy:11137  \n",
       "14  taxonomy:11137  \n",
       "15  taxonomy:11137  \n",
       "16  taxonomy:11137  \n",
       "17  taxonomy:11137  \n",
       "18  taxonomy:11137  \n",
       "19  taxonomy:11137  \n",
       "20  taxonomy:11137  \n",
       "21  taxonomy:11137  \n",
       "22  taxonomy:11137  \n",
       "23  taxonomy:11137  \n",
       "24  taxonomy:11137  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "CALL db.index.fulltext.queryNodes(\"bioentities\", '\\\"Spike glycoprotein\\\"') YIELD node\n",
    "WHERE 'Protein' IN labels(node)\n",
    "RETURN node.name AS name, node.start as start, node.end as end, node.fullLength as fullLength, node.accession as accession, node.proId as proId, node.taxonomyId as taxonomy\n",
    "ORDER BY taxonomy, start\n",
    "\"\"\"\n",
    "df = graph.run(query).to_data_frame()\n",
    "df.head(25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Which organisms have been identified as hosts for Coronaviruses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>taxonomyId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>human</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>carnivores</td>\n",
       "      <td>Canidae</td>\n",
       "      <td>taxonomy:9608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dog</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>taxonomy:9615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>European mink</td>\n",
       "      <td>Mustela lutreola</td>\n",
       "      <td>taxonomy:9666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cat</td>\n",
       "      <td>Felis catus</td>\n",
       "      <td>taxonomy:9685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lion</td>\n",
       "      <td>Panthera leo</td>\n",
       "      <td>taxonomy:9689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tiger</td>\n",
       "      <td>Panthera tigris</td>\n",
       "      <td>taxonomy:9694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Malayan pangolin</td>\n",
       "      <td>Manis javanica</td>\n",
       "      <td>taxonomy:9974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>golden hamster</td>\n",
       "      <td>Mesocricetus auratus</td>\n",
       "      <td>taxonomy:10036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>house mouse</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>taxonomy:10090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>horseshoe bat</td>\n",
       "      <td>Rhinolophus</td>\n",
       "      <td>taxonomy:49442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>intermediate horseshoe bat</td>\n",
       "      <td>Rhinolophus affinis</td>\n",
       "      <td>taxonomy:59477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>greater horseshoe bat</td>\n",
       "      <td>Rhinolophus ferrumequinum</td>\n",
       "      <td>taxonomy:59479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>palm civet</td>\n",
       "      <td>Paradoxurus</td>\n",
       "      <td>taxonomy:71116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Chinese pangolin</td>\n",
       "      <td>Manis pentadactyla</td>\n",
       "      <td>taxonomy:143292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>environmental samples</td>\n",
       "      <td>environmental samples</td>\n",
       "      <td>taxonomy:151659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Malayan tiger</td>\n",
       "      <td>Panthera tigris jacksoni</td>\n",
       "      <td>taxonomy:419130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>American mink</td>\n",
       "      <td>Neovison vison</td>\n",
       "      <td>taxonomy:452646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Malayan horseshoe bat</td>\n",
       "      <td>Rhinolophus malayanus</td>\n",
       "      <td>taxonomy:608659</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          name             scientificName       taxonomyId\n",
       "0                        human               Homo sapiens    taxonomy:9606\n",
       "1                   carnivores                    Canidae    taxonomy:9608\n",
       "2                          dog     Canis lupus familiaris    taxonomy:9615\n",
       "3                European mink           Mustela lutreola    taxonomy:9666\n",
       "4                          cat                Felis catus    taxonomy:9685\n",
       "5                         lion               Panthera leo    taxonomy:9689\n",
       "6                        tiger            Panthera tigris    taxonomy:9694\n",
       "7             Malayan pangolin             Manis javanica    taxonomy:9974\n",
       "8               golden hamster       Mesocricetus auratus   taxonomy:10036\n",
       "9                  house mouse               Mus musculus   taxonomy:10090\n",
       "10               horseshoe bat                Rhinolophus   taxonomy:49442\n",
       "11  intermediate horseshoe bat        Rhinolophus affinis   taxonomy:59477\n",
       "12       greater horseshoe bat  Rhinolophus ferrumequinum   taxonomy:59479\n",
       "13                  palm civet                Paradoxurus   taxonomy:71116\n",
       "14            Chinese pangolin         Manis pentadactyla  taxonomy:143292\n",
       "15       environmental samples      environmental samples  taxonomy:151659\n",
       "16               Malayan tiger   Panthera tigris jacksoni  taxonomy:419130\n",
       "17               American mink             Neovison vison  taxonomy:452646\n",
       "18       Malayan horseshoe bat      Rhinolophus malayanus  taxonomy:608659"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (h:Organism{type: 'Host'})\n",
    "RETURN h.name AS name, h.scientificName AS scientificName, h.id AS taxonomyId\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Which Pathogens have caused Coronavirus Outbreaks?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>scientificName</th>\n",
       "      <th>taxonomyId</th>\n",
       "      <th>outbreak</th>\n",
       "      <th>startDate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>taxonomy:2697049</td>\n",
       "      <td>COVID-19</td>\n",
       "      <td>2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MERS-CoV</td>\n",
       "      <td>Middle East respiratory syndrome-related coron...</td>\n",
       "      <td>taxonomy:1263720</td>\n",
       "      <td>MERS</td>\n",
       "      <td>2012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>Severe acute respiratory syndrome-related coro...</td>\n",
       "      <td>taxonomy:694009</td>\n",
       "      <td>SARS</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         name                                     scientificName  \\\n",
       "0  SARS-CoV-2    Severe acute respiratory syndrome coronavirus 2   \n",
       "1    MERS-CoV  Middle East respiratory syndrome-related coron...   \n",
       "2    SARS-CoV  Severe acute respiratory syndrome-related coro...   \n",
       "\n",
       "         taxonomyId  outbreak startDate  \n",
       "0  taxonomy:2697049  COVID-19      2019  \n",
       "1  taxonomy:1263720      MERS      2012  \n",
       "2   taxonomy:694009      SARS      2003  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (p:Organism)-[:CAUSES]->(o:Outbreak)\n",
    "RETURN p.name AS name, p.scientificName AS scientificName, p.id AS taxonomyId, o.id AS outbreak, o.startDate AS startDate\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Which PubMedCentral articles mention strain datasets about different hosts?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmc</th>\n",
       "      <th>name</th>\n",
       "      <th>collectionDate</th>\n",
       "      <th>host</th>\n",
       "      <th>hostTaxonomyId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1626</th>\n",
       "      <td>https://europepmc.org/article/PPR/PPR204749</td>\n",
       "      <td>hCoV-19/USA/NV-NSPHL-A0132/2020</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2477</th>\n",
       "      <td>https://europepmc.org/article/PPR/PPR180499</td>\n",
       "      <td>hCoV-19/Canada/ON_QGLO-011/2020</td>\n",
       "      <td>2020-03-26</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2181</th>\n",
       "      <td>https://europepmc.org/article/PPR/PPR195212</td>\n",
       "      <td>hCoV-19/Turkey/HSGM-10232/2020</td>\n",
       "      <td>2020-03-24</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3410</th>\n",
       "      <td>pubmed:32543353</td>\n",
       "      <td>SARS-CoV-2/human/TWN/CGMH-CGU-18/2020</td>\n",
       "      <td>2020-03-18</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2402</th>\n",
       "      <td>https://europepmc.org/article/PPR/PPR210955</td>\n",
       "      <td>hCoV-19/Peru/INS-024/2020</td>\n",
       "      <td>2020-03-25</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3511</th>\n",
       "      <td>https://europepmc.org/article/PPR/PPR214566</td>\n",
       "      <td>hCoV-19/India/CCMB_J375/2020</td>\n",
       "      <td>2020-04-02</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3696</th>\n",
       "      <td>https://europepmc.org/article/PPR/PPR210888</td>\n",
       "      <td>hCoV-19/South Korea/KCDC2655/2020</td>\n",
       "      <td>2020-07-21</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3300</th>\n",
       "      <td>https://europepmc.org/article/PPR/PPR224372</td>\n",
       "      <td>hCoV-19/Cyprus/005/2020</td>\n",
       "      <td>2020-04-01</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>804</th>\n",
       "      <td>pubmed:32937193</td>\n",
       "      <td>Wuhan-Hu-1</td>\n",
       "      <td>2019-12-30</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>pubmed:32489698</td>\n",
       "      <td>hCoV-19/Italy/SPL1/2020</td>\n",
       "      <td>2020-01-29</td>\n",
       "      <td>human</td>\n",
       "      <td>taxonomy:9606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              pmc  \\\n",
       "1626  https://europepmc.org/article/PPR/PPR204749   \n",
       "2477  https://europepmc.org/article/PPR/PPR180499   \n",
       "2181  https://europepmc.org/article/PPR/PPR195212   \n",
       "3410                              pubmed:32543353   \n",
       "2402  https://europepmc.org/article/PPR/PPR210955   \n",
       "3511  https://europepmc.org/article/PPR/PPR214566   \n",
       "3696  https://europepmc.org/article/PPR/PPR210888   \n",
       "3300  https://europepmc.org/article/PPR/PPR224372   \n",
       "804                               pubmed:32937193   \n",
       "122                               pubmed:32489698   \n",
       "\n",
       "                                       name collectionDate   host  \\\n",
       "1626        hCoV-19/USA/NV-NSPHL-A0132/2020     2020-04-29  human   \n",
       "2477        hCoV-19/Canada/ON_QGLO-011/2020     2020-03-26  human   \n",
       "2181         hCoV-19/Turkey/HSGM-10232/2020     2020-03-24  human   \n",
       "3410  SARS-CoV-2/human/TWN/CGMH-CGU-18/2020     2020-03-18  human   \n",
       "2402              hCoV-19/Peru/INS-024/2020     2020-03-25  human   \n",
       "3511           hCoV-19/India/CCMB_J375/2020     2020-04-02  human   \n",
       "3696      hCoV-19/South Korea/KCDC2655/2020     2020-07-21  human   \n",
       "3300                hCoV-19/Cyprus/005/2020     2020-04-01  human   \n",
       "804                              Wuhan-Hu-1     2019-12-30  human   \n",
       "122                 hCoV-19/Italy/SPL1/2020     2020-01-29  human   \n",
       "\n",
       "     hostTaxonomyId  \n",
       "1626  taxonomy:9606  \n",
       "2477  taxonomy:9606  \n",
       "2181  taxonomy:9606  \n",
       "3410  taxonomy:9606  \n",
       "2402  taxonomy:9606  \n",
       "3511  taxonomy:9606  \n",
       "3696  taxonomy:9606  \n",
       "3300  taxonomy:9606  \n",
       "804   taxonomy:9606  \n",
       "122   taxonomy:9606  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (p:Publication)-[:MENTIONS]->(s:Strain)<-[:CARRIES]-(h:Organism)\n",
    "RETURN p.id AS pmc, s.name AS name, s.collectionDate  AS collectionDate, h.name AS host, h.id AS hostTaxonomyId\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame().sample(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the genes and gene products of the SARS-CoV-2 Virus?\n",
    "This query lists the genes and proteins encoded by the SARS-CoV-2 NC_045512 reference genome. MN908947 represents the original un-annotated dataset submitted to NCBI. This is the first sequenced genome of SARS-CoV-2 collected in Wuhan."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>protein</th>\n",
       "      <th>accession</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>M</td>\n",
       "      <td>Membrane protein</td>\n",
       "      <td>uniprot:P0DTC5</td>\n",
       "      <td>MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>3a</td>\n",
       "      <td>ORF3a protein</td>\n",
       "      <td>uniprot:P0DTC3</td>\n",
       "      <td>MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Replicase polyprotein 1ab</td>\n",
       "      <td>uniprot:P0DTD1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>N</td>\n",
       "      <td>Nucleoprotein</td>\n",
       "      <td>uniprot:P0DTC9</td>\n",
       "      <td>MSDNGPQNQRNAPRITFGGPSDSTGSNQNGERSGARSKQRRPQGLP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>6</td>\n",
       "      <td>ORF6 protein</td>\n",
       "      <td>uniprot:P0DTC6</td>\n",
       "      <td>MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>ORF14</td>\n",
       "      <td>Uncharacterized protein 14</td>\n",
       "      <td>uniprot:P0DTD3</td>\n",
       "      <td>MLQSCYNFLKEQHCQKASTQKGAEAAVKPLLVPHHVVATVQEIQLQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>E</td>\n",
       "      <td>Envelope small membrane protein</td>\n",
       "      <td>uniprot:P0DTC4</td>\n",
       "      <td>MYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>9b</td>\n",
       "      <td>ORF9b protein</td>\n",
       "      <td>uniprot:P0DTD2</td>\n",
       "      <td>MDPKISEMHPALRLVDPQIQLAVTRMENAVGRDQNNVGPKVYPIIL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Replicase polyprotein 1a</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>S</td>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>uniprot:P0DTC2</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>7b</td>\n",
       "      <td>ORF7b protein</td>\n",
       "      <td>uniprot:P0DTD8</td>\n",
       "      <td>MIELSLIDFYLCFLAFLLFLVLIMLIIFWFSLELQDHNETCHA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>8</td>\n",
       "      <td>ORF8 protein</td>\n",
       "      <td>uniprot:P0DTC8</td>\n",
       "      <td>MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>7a</td>\n",
       "      <td>ORF7a protein</td>\n",
       "      <td>uniprot:P0DTC7</td>\n",
       "      <td>MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          name   gene                          protein       accession  \\\n",
       "0   SARS-CoV-2      M                 Membrane protein  uniprot:P0DTC5   \n",
       "1   SARS-CoV-2     3a                    ORF3a protein  uniprot:P0DTC3   \n",
       "2   SARS-CoV-2    rep        Replicase polyprotein 1ab  uniprot:P0DTD1   \n",
       "3   SARS-CoV-2      N                    Nucleoprotein  uniprot:P0DTC9   \n",
       "4   SARS-CoV-2      6                     ORF6 protein  uniprot:P0DTC6   \n",
       "5   SARS-CoV-2  ORF14       Uncharacterized protein 14  uniprot:P0DTD3   \n",
       "6   SARS-CoV-2      E  Envelope small membrane protein  uniprot:P0DTC4   \n",
       "7   SARS-CoV-2     9b                    ORF9b protein  uniprot:P0DTD2   \n",
       "8   SARS-CoV-2     1a         Replicase polyprotein 1a  uniprot:P0DTC1   \n",
       "9   SARS-CoV-2      S               Spike glycoprotein  uniprot:P0DTC2   \n",
       "10  SARS-CoV-2     7b                    ORF7b protein  uniprot:P0DTD8   \n",
       "11  SARS-CoV-2      8                     ORF8 protein  uniprot:P0DTC8   \n",
       "12  SARS-CoV-2     7a                    ORF7a protein  uniprot:P0DTC7   \n",
       "\n",
       "                                             sequence  \n",
       "0   MADSNGTITVEELKKLLEQWNLVIGFLFLTWICLLQFAYANRNRFL...  \n",
       "1   MDLFMRIFTIGTVTLKQGEIKDATPSDFVRATATIPIQASLPFGWL...  \n",
       "2   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "3   MSDNGPQNQRNAPRITFGGPSDSTGSNQNGERSGARSKQRRPQGLP...  \n",
       "4   MFHLVDFQVTIAEILLIIMRTFKVSIWNLDYIINLIIKNLSKSLTE...  \n",
       "5   MLQSCYNFLKEQHCQKASTQKGAEAAVKPLLVPHHVVATVQEIQLQ...  \n",
       "6   MYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNI...  \n",
       "7   MDPKISEMHPALRLVDPQIQLAVTRMENAVGRDQNNVGPKVYPIIL...  \n",
       "8   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "9   MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...  \n",
       "10        MIELSLIDFYLCFLAFLLFLVLIMLIIFWFSLELQDHNETCHA  \n",
       "11  MKFLVFLGIITTVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...  \n",
       "12  MKIILFLALITLATCELYHYQECVRGTTVLLKEPCSSGTYEGNSPF...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (o:Organism{name: 'SARS-CoV-2'})-[:HAS_GENOME]-(n:Genome)-[:HAS_CHROMOSOME]-(Chromosome)-[:HAS_GENE]-(g:Gene)-[:ENCODES]->(p:Protein{reviewed: True})\n",
    "RETURN o.name as name, g.name as gene, p.name as protein, p.accession as accession, p.sequence as sequence\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The following proteins are protolytic cleavage products of expressed proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>protein</th>\n",
       "      <th>accession</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Non-structural protein 4</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>KIVNNWLKQLIKVTLVFLFVAAIFYLITPVHVMSKHTDFSSEIIGY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Non-structural protein 2</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>AYTRYVDNNFCGPDGYPLECIKDLLARAGKASCTLSEQLDFIDTKR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>RNA-directed RNA polymerase</td>\n",
       "      <td>uniprot:P0DTD1</td>\n",
       "      <td>SADAQSFLNRVCGVSAARLTPCGTGTSTDVVYRAFDIYNDKVAGFA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Helicase</td>\n",
       "      <td>uniprot:P0DTD1</td>\n",
       "      <td>AVGACVLCNSQTSLRCGACIRRPFLCCKCCYDHVISTSHKLVLSVN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Non-structural protein 7</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>SKMSDVKCTSVVLLSVLQQLRVESSSKLWAQCVQLHNDILLAKDTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>3C-like proteinase</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Non-structural protein 3</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>APTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVEL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Host translation inhibitor nsp1</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Non-structural protein 8</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>AIASEFSSLPSYAAFATAQEAYEQAVANGDSEVVLKKLKKSLNVAK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Uridylate-specific endoribonuclease</td>\n",
       "      <td>uniprot:P0DTD1</td>\n",
       "      <td>SLENVAFNVVNKGHFDGQQGEVPVSIINNTVYTKVDGVDVELFENK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Non-structural protein 10</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>AGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>2'-O-methyltransferase</td>\n",
       "      <td>uniprot:P0DTD1</td>\n",
       "      <td>SSQAWQPGVAMPNLYKMQRMLLEKCDLQNYGDSATLPKGIMMNVAK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Non-structural protein 6</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>SAVKRTIKGTHHWLLLTILTSLLVLVQSTQWSLFFFLYENAFLPFA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Proofreading exoribonuclease</td>\n",
       "      <td>uniprot:P0DTD1</td>\n",
       "      <td>AENVTGLFKDCSKVITGLHPTQAPTHLSVDTKFKTEGLCVDIPGIP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>rep</td>\n",
       "      <td>Non-structural protein 9</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>NNELSPVALRQMSCAAGTTQTACTDDNALAYYNTTKGGRFVLALLS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 2</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>AYTRYVDNNFCGPDGYPLECIKDLLARAGKASCTLSEQLDFIDTKR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 11</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>SADAQSFLNGFAV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 3</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>APTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVEL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 8</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>AIASEFSSLPSYAAFATAQEAYEQAVANGDSEVVLKKLKKSLNVAK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Host translation inhibitor nsp1</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 6</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>SAVKRTIKGTHHWLLLTILTSLLVLVQSTQWSLFFFLYENAFLPFA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 7</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>SKMSDVKCTSVVLLSVLQQLRVESSSKLWAQCVQLHNDILLAKDTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 4</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>KIVNNWLKQLIKVTLVFLFVAAIFYLITPVHVMSKHTDFSSEIIGY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 9</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>NNELSPVALRQMSCAAGTTQTACTDDNALAYYNTTKGGRFVLALLS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>3C-like proteinase</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>1a</td>\n",
       "      <td>Non-structural protein 10</td>\n",
       "      <td>uniprot:P0DTC1</td>\n",
       "      <td>AGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>S</td>\n",
       "      <td>Spike protein S2'</td>\n",
       "      <td>uniprot:P0DTC2</td>\n",
       "      <td>SFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>S</td>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>uniprot:P0DTC2</td>\n",
       "      <td>SQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>S</td>\n",
       "      <td>Spike protein S1</td>\n",
       "      <td>uniprot:P0DTC2</td>\n",
       "      <td>SQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>S</td>\n",
       "      <td>Spike protein S2</td>\n",
       "      <td>uniprot:P0DTC2</td>\n",
       "      <td>SVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>8</td>\n",
       "      <td>ORF8 protein</td>\n",
       "      <td>uniprot:P0DTC8</td>\n",
       "      <td>FHQECSLQSCTQHQPYVVDDPCPIHFYSKWYIRVGARKSAPLIELC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>7a</td>\n",
       "      <td>ORF7a protein</td>\n",
       "      <td>uniprot:P0DTC7</td>\n",
       "      <td>ELYHYQECVRGTTVLLKEPCSSGTYEGNSPFHPLADNKFALTCFST...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          name gene                              protein       accession  \\\n",
       "0   SARS-CoV-2  rep             Non-structural protein 4  uniprot:P0DTC1   \n",
       "1   SARS-CoV-2  rep             Non-structural protein 2  uniprot:P0DTC1   \n",
       "2   SARS-CoV-2  rep          RNA-directed RNA polymerase  uniprot:P0DTD1   \n",
       "3   SARS-CoV-2  rep                             Helicase  uniprot:P0DTD1   \n",
       "4   SARS-CoV-2  rep             Non-structural protein 7  uniprot:P0DTC1   \n",
       "5   SARS-CoV-2  rep                   3C-like proteinase  uniprot:P0DTC1   \n",
       "6   SARS-CoV-2  rep             Non-structural protein 3  uniprot:P0DTC1   \n",
       "7   SARS-CoV-2  rep      Host translation inhibitor nsp1  uniprot:P0DTC1   \n",
       "8   SARS-CoV-2  rep             Non-structural protein 8  uniprot:P0DTC1   \n",
       "9   SARS-CoV-2  rep  Uridylate-specific endoribonuclease  uniprot:P0DTD1   \n",
       "10  SARS-CoV-2  rep            Non-structural protein 10  uniprot:P0DTC1   \n",
       "11  SARS-CoV-2  rep               2'-O-methyltransferase  uniprot:P0DTD1   \n",
       "12  SARS-CoV-2  rep             Non-structural protein 6  uniprot:P0DTC1   \n",
       "13  SARS-CoV-2  rep         Proofreading exoribonuclease  uniprot:P0DTD1   \n",
       "14  SARS-CoV-2  rep             Non-structural protein 9  uniprot:P0DTC1   \n",
       "15  SARS-CoV-2   1a             Non-structural protein 2  uniprot:P0DTC1   \n",
       "16  SARS-CoV-2   1a            Non-structural protein 11  uniprot:P0DTC1   \n",
       "17  SARS-CoV-2   1a             Non-structural protein 3  uniprot:P0DTC1   \n",
       "18  SARS-CoV-2   1a             Non-structural protein 8  uniprot:P0DTC1   \n",
       "19  SARS-CoV-2   1a      Host translation inhibitor nsp1  uniprot:P0DTC1   \n",
       "20  SARS-CoV-2   1a             Non-structural protein 6  uniprot:P0DTC1   \n",
       "21  SARS-CoV-2   1a             Non-structural protein 7  uniprot:P0DTC1   \n",
       "22  SARS-CoV-2   1a             Non-structural protein 4  uniprot:P0DTC1   \n",
       "23  SARS-CoV-2   1a             Non-structural protein 9  uniprot:P0DTC1   \n",
       "24  SARS-CoV-2   1a                   3C-like proteinase  uniprot:P0DTC1   \n",
       "25  SARS-CoV-2   1a            Non-structural protein 10  uniprot:P0DTC1   \n",
       "26  SARS-CoV-2    S                    Spike protein S2'  uniprot:P0DTC2   \n",
       "27  SARS-CoV-2    S                   Spike glycoprotein  uniprot:P0DTC2   \n",
       "28  SARS-CoV-2    S                     Spike protein S1  uniprot:P0DTC2   \n",
       "29  SARS-CoV-2    S                     Spike protein S2  uniprot:P0DTC2   \n",
       "30  SARS-CoV-2    8                         ORF8 protein  uniprot:P0DTC8   \n",
       "31  SARS-CoV-2   7a                        ORF7a protein  uniprot:P0DTC7   \n",
       "\n",
       "                                             sequence  \n",
       "0   KIVNNWLKQLIKVTLVFLFVAAIFYLITPVHVMSKHTDFSSEIIGY...  \n",
       "1   AYTRYVDNNFCGPDGYPLECIKDLLARAGKASCTLSEQLDFIDTKR...  \n",
       "2   SADAQSFLNRVCGVSAARLTPCGTGTSTDVVYRAFDIYNDKVAGFA...  \n",
       "3   AVGACVLCNSQTSLRCGACIRRPFLCCKCCYDHVISTSHKLVLSVN...  \n",
       "4   SKMSDVKCTSVVLLSVLQQLRVESSSKLWAQCVQLHNDILLAKDTT...  \n",
       "5   SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTS...  \n",
       "6   APTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVEL...  \n",
       "7   MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "8   AIASEFSSLPSYAAFATAQEAYEQAVANGDSEVVLKKLKKSLNVAK...  \n",
       "9   SLENVAFNVVNKGHFDGQQGEVPVSIINNTVYTKVDGVDVELFENK...  \n",
       "10  AGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLC...  \n",
       "11  SSQAWQPGVAMPNLYKMQRMLLEKCDLQNYGDSATLPKGIMMNVAK...  \n",
       "12  SAVKRTIKGTHHWLLLTILTSLLVLVQSTQWSLFFFLYENAFLPFA...  \n",
       "13  AENVTGLFKDCSKVITGLHPTQAPTHLSVDTKFKTEGLCVDIPGIP...  \n",
       "14  NNELSPVALRQMSCAAGTTQTACTDDNALAYYNTTKGGRFVLALLS...  \n",
       "15  AYTRYVDNNFCGPDGYPLECIKDLLARAGKASCTLSEQLDFIDTKR...  \n",
       "16                                      SADAQSFLNGFAV  \n",
       "17  APTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVEL...  \n",
       "18  AIASEFSSLPSYAAFATAQEAYEQAVANGDSEVVLKKLKKSLNVAK...  \n",
       "19  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...  \n",
       "20  SAVKRTIKGTHHWLLLTILTSLLVLVQSTQWSLFFFLYENAFLPFA...  \n",
       "21  SKMSDVKCTSVVLLSVLQQLRVESSSKLWAQCVQLHNDILLAKDTT...  \n",
       "22  KIVNNWLKQLIKVTLVFLFVAAIFYLITPVHVMSKHTDFSSEIIGY...  \n",
       "23  NNELSPVALRQMSCAAGTTQTACTDDNALAYYNTTKGGRFVLALLS...  \n",
       "24  SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTS...  \n",
       "25  AGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLC...  \n",
       "26  SFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVL...  \n",
       "27  SQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPF...  \n",
       "28  SQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPF...  \n",
       "29  SVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSM...  \n",
       "30  FHQECSLQSCTQHQPYVVDDPCPIHFYSKWYIRVGARKSAPLIELC...  \n",
       "31  ELYHYQECVRGTTVLLKEPCSSGTYEGNSPFHPLADNKFALTCFST...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (o:Organism{name: 'SARS-CoV-2'})-[:HAS_GENOME]-(n:Genome)-[:HAS_CHROMOSOME]-(Chromosome)-[:HAS_GENE]-(g:Gene)-[:ENCODES]->(:Protein{reviewed: True})-[:CLEAVED_TO]->(p:Protein)\n",
    "RETURN o.name as name, g.name as gene, p.name as protein, p.accession as accession, p.sequence as sequence\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many complete SARS-CoV-2 genomes have been sequenced in each country?\n",
    "Here we aggregate the number of genomes over up to 3 hops to the UN Region level:\n",
    "\n",
    "`City -> Admin2(county) -> Admin1(state, province) -> Country`\n",
    "\n",
    "using the variable-length relationship `[:IN*0..3]`.\n",
    "\n",
    "This number includes only complete high-quality genomes as determined by the [China National Center for Bioinformation, 2019 Novel Coronavirus Resource (2019nCoVR)](https://bigd.big.ac.cn/ncov/release_genome).\n",
    "\n",
    "Note, some strains have been deposited under different names in multiple repositories. Therefore, the numbers below includes some duplicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>90014</td>\n",
       "      <td>United Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>67228</td>\n",
       "      <td>United States</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14343</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9008</td>\n",
       "      <td>Denmark</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4500</td>\n",
       "      <td>Spain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3496</td>\n",
       "      <td>Netherlands</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3350</td>\n",
       "      <td>Switzerland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3261</td>\n",
       "      <td>Canada</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3256</td>\n",
       "      <td>India</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1986</td>\n",
       "      <td>South Africa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1660</td>\n",
       "      <td>Portugal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1631</td>\n",
       "      <td>France</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1444</td>\n",
       "      <td>Belgium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1230</td>\n",
       "      <td>China</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1180</td>\n",
       "      <td>Singapore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1179</td>\n",
       "      <td>Brazil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>973</td>\n",
       "      <td>New Zealand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>910</td>\n",
       "      <td>Italy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>879</td>\n",
       "      <td>Russia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>846</td>\n",
       "      <td>South Korea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>837</td>\n",
       "      <td>Japan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>808</td>\n",
       "      <td>Sweden</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>756</td>\n",
       "      <td>Austria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>755</td>\n",
       "      <td>Germany</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>630</td>\n",
       "      <td>Bangladesh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>600</td>\n",
       "      <td>Iceland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>575</td>\n",
       "      <td>Saudi Arabia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>573</td>\n",
       "      <td>Norway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>534</td>\n",
       "      <td>Hong Kong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>533</td>\n",
       "      <td>Thailand</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>510</td>\n",
       "      <td>Ireland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>419</td>\n",
       "      <td>Israel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>385</td>\n",
       "      <td>Peru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>360</td>\n",
       "      <td>Gambia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>352</td>\n",
       "      <td>Democratic Republic of the Congo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>314</td>\n",
       "      <td>Panama</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>305</td>\n",
       "      <td>Mexico</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>290</td>\n",
       "      <td>Kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>276</td>\n",
       "      <td>Egypt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>271</td>\n",
       "      <td>Luxembourg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>268</td>\n",
       "      <td>Finland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>245</td>\n",
       "      <td>United Arab Emirates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>239</td>\n",
       "      <td>Chile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>230</td>\n",
       "      <td>Iran</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>229</td>\n",
       "      <td>Jordan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>226</td>\n",
       "      <td>Poland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>215</td>\n",
       "      <td>Turkey</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>201</td>\n",
       "      <td>Oman</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>196</td>\n",
       "      <td>Gibraltar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>177</td>\n",
       "      <td>Colombia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>177</td>\n",
       "      <td>Czechia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>167</td>\n",
       "      <td>Greece</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>136</td>\n",
       "      <td>Senegal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>135</td>\n",
       "      <td>Latvia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>135</td>\n",
       "      <td>Taiwan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>123</td>\n",
       "      <td>Malaysia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>120</td>\n",
       "      <td>Indonesia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>112</td>\n",
       "      <td>Romania</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>108</td>\n",
       "      <td>Slovenia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>92</td>\n",
       "      <td>Bahrain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>92</td>\n",
       "      <td>Lithuania</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>91</td>\n",
       "      <td>Costa Rica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>90</td>\n",
       "      <td>Ecuador</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>90</td>\n",
       "      <td>Vietnam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>88</td>\n",
       "      <td>Nigeria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>85</td>\n",
       "      <td>Hungary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>78</td>\n",
       "      <td>Iraq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>75</td>\n",
       "      <td>Uruguay</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>72</td>\n",
       "      <td>Serbia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>71</td>\n",
       "      <td>Suriname</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>70</td>\n",
       "      <td>Ghana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>58</td>\n",
       "      <td>Ukraine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>56</td>\n",
       "      <td>Kazakhstan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>56</td>\n",
       "      <td>Morocco</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>53</td>\n",
       "      <td>Philippines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>39</td>\n",
       "      <td>Argentina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>37</td>\n",
       "      <td>Faroe Islands</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>32</td>\n",
       "      <td>North Macedonia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>31</td>\n",
       "      <td>Bosnia and Herzegovina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>31</td>\n",
       "      <td>Tunisia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>27</td>\n",
       "      <td>Sri Lanka</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>25</td>\n",
       "      <td>Guadeloupe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>25</td>\n",
       "      <td>Slovakia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>23</td>\n",
       "      <td>Venezuela</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>22</td>\n",
       "      <td>Georgia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>22</td>\n",
       "      <td>Croatia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>22</td>\n",
       "      <td>Pakistan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>21</td>\n",
       "      <td>Mali</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>20</td>\n",
       "      <td>Estonia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>20</td>\n",
       "      <td>Malta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>20</td>\n",
       "      <td>Uganda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>19</td>\n",
       "      <td>Timor Leste</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>17</td>\n",
       "      <td>Lebanon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>16</td>\n",
       "      <td>Qatar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>15</td>\n",
       "      <td>Bulgaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>12</td>\n",
       "      <td>Benin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>12</td>\n",
       "      <td>Montenegro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>11</td>\n",
       "      <td>Aruba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>11</td>\n",
       "      <td>Republic of the Congo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>11</td>\n",
       "      <td>Sierra Leone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>10</td>\n",
       "      <td>Guatemala</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>10</td>\n",
       "      <td>Rwanda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>9</td>\n",
       "      <td>Curacao</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>9</td>\n",
       "      <td>Gabon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>9</td>\n",
       "      <td>Moldova</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>8</td>\n",
       "      <td>Cyprus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>8</td>\n",
       "      <td>Dominican Republic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>8</td>\n",
       "      <td>Jamaica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>8</td>\n",
       "      <td>Kuwait</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>7</td>\n",
       "      <td>Mongolia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>6</td>\n",
       "      <td>Botswana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>6</td>\n",
       "      <td>Guam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>6</td>\n",
       "      <td>U.S. Virgin Islands</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>5</td>\n",
       "      <td>Brunei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>4</td>\n",
       "      <td>Belize</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>4</td>\n",
       "      <td>Reunion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>3</td>\n",
       "      <td>Belarus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>3</td>\n",
       "      <td>Algeria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>3</td>\n",
       "      <td>Madagascar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>3</td>\n",
       "      <td>Puerto Rico</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>2</td>\n",
       "      <td>Cuba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>2</td>\n",
       "      <td>Myanmar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>2</td>\n",
       "      <td>Nepal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>2</td>\n",
       "      <td>Uzbekistan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1</td>\n",
       "      <td>Andorra</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>1</td>\n",
       "      <td>Cameroon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>1</td>\n",
       "      <td>Cambodia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>1</td>\n",
       "      <td>Zambia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     count                           country\n",
       "0    90014                    United Kingdom\n",
       "1    67228                     United States\n",
       "2    14343                         Australia\n",
       "3     9008                           Denmark\n",
       "4     4500                             Spain\n",
       "5     3496                       Netherlands\n",
       "6     3350                       Switzerland\n",
       "7     3261                            Canada\n",
       "8     3256                             India\n",
       "9     1986                      South Africa\n",
       "10    1660                          Portugal\n",
       "11    1631                            France\n",
       "12    1444                           Belgium\n",
       "13    1230                             China\n",
       "14    1180                         Singapore\n",
       "15    1179                            Brazil\n",
       "16     973                       New Zealand\n",
       "17     910                             Italy\n",
       "18     879                            Russia\n",
       "19     846                       South Korea\n",
       "20     837                             Japan\n",
       "21     808                            Sweden\n",
       "22     756                           Austria\n",
       "23     755                           Germany\n",
       "24     630                        Bangladesh\n",
       "25     600                           Iceland\n",
       "26     575                      Saudi Arabia\n",
       "27     573                            Norway\n",
       "28     534                         Hong Kong\n",
       "29     533                          Thailand\n",
       "30     510                           Ireland\n",
       "31     419                            Israel\n",
       "32     385                              Peru\n",
       "33     360                            Gambia\n",
       "34     352  Democratic Republic of the Congo\n",
       "35     314                            Panama\n",
       "36     305                            Mexico\n",
       "37     290                             Kenya\n",
       "38     276                             Egypt\n",
       "39     271                        Luxembourg\n",
       "40     268                           Finland\n",
       "41     245              United Arab Emirates\n",
       "42     239                             Chile\n",
       "43     230                              Iran\n",
       "44     229                            Jordan\n",
       "45     226                            Poland\n",
       "46     215                            Turkey\n",
       "47     201                              Oman\n",
       "48     196                         Gibraltar\n",
       "49     177                          Colombia\n",
       "50     177                           Czechia\n",
       "51     167                            Greece\n",
       "52     136                           Senegal\n",
       "53     135                            Latvia\n",
       "54     135                            Taiwan\n",
       "55     123                          Malaysia\n",
       "56     120                         Indonesia\n",
       "57     112                           Romania\n",
       "58     108                          Slovenia\n",
       "59      92                           Bahrain\n",
       "60      92                         Lithuania\n",
       "61      91                        Costa Rica\n",
       "62      90                           Ecuador\n",
       "63      90                           Vietnam\n",
       "64      88                           Nigeria\n",
       "65      85                           Hungary\n",
       "66      78                              Iraq\n",
       "67      75                           Uruguay\n",
       "68      72                            Serbia\n",
       "69      71                          Suriname\n",
       "70      70                             Ghana\n",
       "71      58                           Ukraine\n",
       "72      56                        Kazakhstan\n",
       "73      56                           Morocco\n",
       "74      53                       Philippines\n",
       "75      39                         Argentina\n",
       "76      37                     Faroe Islands\n",
       "77      32                   North Macedonia\n",
       "78      31            Bosnia and Herzegovina\n",
       "79      31                           Tunisia\n",
       "80      27                         Sri Lanka\n",
       "81      25                        Guadeloupe\n",
       "82      25                          Slovakia\n",
       "83      23                         Venezuela\n",
       "84      22                           Georgia\n",
       "85      22                           Croatia\n",
       "86      22                          Pakistan\n",
       "87      21                              Mali\n",
       "88      20                           Estonia\n",
       "89      20                             Malta\n",
       "90      20                            Uganda\n",
       "91      19                       Timor Leste\n",
       "92      17                           Lebanon\n",
       "93      16                             Qatar\n",
       "94      15                          Bulgaria\n",
       "95      12                             Benin\n",
       "96      12                        Montenegro\n",
       "97      11                             Aruba\n",
       "98      11             Republic of the Congo\n",
       "99      11                      Sierra Leone\n",
       "100     10                         Guatemala\n",
       "101     10                            Rwanda\n",
       "102      9                           Curacao\n",
       "103      9                             Gabon\n",
       "104      9                           Moldova\n",
       "105      8                            Cyprus\n",
       "106      8                Dominican Republic\n",
       "107      8                           Jamaica\n",
       "108      8                            Kuwait\n",
       "109      7                          Mongolia\n",
       "110      6                          Botswana\n",
       "111      6                              Guam\n",
       "112      6               U.S. Virgin Islands\n",
       "113      5                            Brunei\n",
       "114      4                            Belize\n",
       "115      4                           Reunion\n",
       "116      3                           Belarus\n",
       "117      3                           Algeria\n",
       "118      3                        Madagascar\n",
       "119      3                       Puerto Rico\n",
       "120      2                              Cuba\n",
       "121      2                           Myanmar\n",
       "122      2                             Nepal\n",
       "123      2                        Uzbekistan\n",
       "124      1                           Andorra\n",
       "125      1                          Cameroon\n",
       "126      1                          Cambodia\n",
       "127      1                            Zambia"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (s:Strain)-[:FOUND_IN]->(l:Location)-[:IN*0..3]->(c:Country)\n",
    "RETURN count(s) AS count, c.name AS country\n",
    "ORDER by count DESC\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How many complete SARS-CoV-2 genome have been sequenced in each UN Region?\n",
    "Here we aggregate the number of genomes over up to 6 hops to the UN Region level:\n",
    "\n",
    "`City -> Admin2(county) -> Admin1(state, province) -> Country -> UNSubRegion -> UNIntermediateRegion -> UNRegion`\n",
    "\n",
    "using the variable-length relationship `[:IN*0..6]`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>u.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>123088</td>\n",
       "      <td>Europe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>83625</td>\n",
       "      <td>Americas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15322</td>\n",
       "      <td>Oceania</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12065</td>\n",
       "      <td>Asia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3757</td>\n",
       "      <td>Africa</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    count    u.name\n",
       "0  123088    Europe\n",
       "1   83625  Americas\n",
       "2   15322   Oceania\n",
       "3   12065      Asia\n",
       "4    3757    Africa"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "MATCH (s:Strain)-[:FOUND_IN]->(l:Location)-[:IN*0..6]->(u:UNRegion)\n",
    "RETURN count(s) AS count, u.name\n",
    "ORDER by count DESC\n",
    "\"\"\"\n",
    "graph.run(query).to_data_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
